{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cancer Integration with NCGA < https://portal.gdc.cancer.gov/\n",
    "\n",
    "There is an API, however since we are only dealing with 31 GAIN domains / 32 receptors, I grabbed all data in JSON. https://docs.gdc.cancer.gov/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import json, glob, re, math\n",
    "\n",
    "from matplotlib.ticker import (MultipleLocator, AutoMinorLocator, FixedLocator)\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sse_func\n",
    "from gain_classes import *\n",
    "jsons = glob.glob(\"../repo_data/gain_json/*.json\")\n",
    "csvs = glob.glob(\"/home/hildilab/projects/agpcr_nom/snp_mane/*csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The loaded JSON will contain a list of dictionaries, where every dictionary corresponds to a single mutation of the respective receptor. With the pre-defined window, we have a list of mutations, where the keys correspond to properties of respective mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''print(j.split(\"/\")[-1].split(\".\")[0].upper(), \"CANCER GENOME ATLAS MUTATIONS\\n\",\"_\"*70)\n",
    "positions = []\n",
    "for mutation in mutation_data['mutations']:\n",
    "    for key in mutation.keys():\n",
    "        print(key.ljust(30), mutation[key])\n",
    "    print(\"_\"*70)\n",
    "    resid = int(mutation[\"x\"]) # X is the residue ID\n",
    "    positions.append(resid)\n",
    "    \n",
    "def totext(f, stuff):\n",
    "    with  open(f, 'a') as txt:\n",
    "        txt.write(stuff)\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to construct a new dictionary where the indexing itself locates to. This can be a new attribute of the class\n",
    "Let's construct the human collection first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the human GAIN collection\n",
    "def filter_by_list(sequences, selection): # selection list\n",
    "    new_list = []\n",
    "    for seq_tup in sequences:\n",
    "        for it in selection:\n",
    "            if it in seq_tup[0]:\n",
    "                new_list.append(seq_tup)\n",
    "    return new_list\n",
    "\n",
    "quality_file = \"/home/hildilab/projects/agpcr_nom/app_gain_gain.mafft.jal\"\n",
    "alignment_file = \"/home/hildilab/projects/agpcr_nom/app_gain_gain.mafft.fa\"\n",
    "stride_folder = \"/home/hildilab/projects/agpcr_nom/all_gps_stride\"\n",
    "stride_files = glob.glob(\"/home/hildilab/projects/agpcr_nom/sigmas/sigma_2/*\")\n",
    "quality = sse_func.read_quality(quality_file)\n",
    "gps_minus_one = 6781 # -1 of the ACTUAL COLUMN (6782) in JALVIEW since there is is ONE-INDEXED\n",
    "aln_cutoff = 6826 # \n",
    "\n",
    "alignment_dict = sse_func.read_alignment(alignment_file, aln_cutoff)\n",
    "valid_seqs = sse_func.read_multi_seq(\"/home/hildilab/projects/agpcr_nom/app_gain_gain.fa\")\n",
    "human_seqs = [\"Q9HBW9\",\"O60241\",\"Q6QNK2\",\"Q9UHX3\",\"Q5T601\",\"Q96PE1\",\"O60242\",\"Q86SQ4\",\n",
    "                \"O94910\",\"Q8IWK6\",\"Q8IZP9\",\"Q8WXG9\",\"Q86Y34\",\"O95490\",\"Q14246\",\"Q9BY15\",\n",
    "                \"Q8IZF2\",\"Q86SQ3\",\"Q8IZF6\",\"Q96K78\",\"Q8IZF3\",\"Q8IZF7\",\"Q8IZF5\",\"Q7Z7M1\",\n",
    "                \"Q8IZF4\",\"Q9HCU4\",\"Q9NYQ6\",\"Q9NYQ7\",\"Q9HAR2\",\"O14514\",\"P48960\",\n",
    "                \"Q9Y653\"]\n",
    "sigma_2_strides = glob.glob(\"/home/hildilab/projects/agpcr_nom/sigma_2*/*.stride\")\n",
    "list_32 = filter_by_list(valid_seqs, human_seqs)\n",
    "\n",
    "human_collection = GainCollection( alignment_file = alignment_file,\n",
    "                                        aln_cutoff = aln_cutoff,\n",
    "                                        quality = quality,\n",
    "                                        gps_index = gps_minus_one,\n",
    "                                        stride_files =  sigma_2_strides, #stride_files,\n",
    "                                        sequence_files=None,\n",
    "                                        sequences=list_32,\n",
    "                                        alignment_dict = alignment_dict,\n",
    "                                        coil_weight=0.00, # won't work with altered boundary detection\n",
    "                                        is_truncated = True,\n",
    "                                        stride_outlier_mode = True\n",
    "                                         )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each GAIN domain within this collection, map the mutations by residue on the corresponding element. Then, merge all that to construct a global mutation map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VALID DATA FROM THE MSA AND THE ANALYSIS PIPELINE\n",
    "sdb = 3425\n",
    "# ADJUSTED ANCHORS!\n",
    "anchors = [ 662, 1194, 1912, 2490, 2848, 3011, 3073, 3260, 3455, 3607, 3998, 4279, 4850, 5339,\n",
    " 5413, 5813, 6337, 6659, 6696, 6765, 6808] # removed S7 @  5341\n",
    "anchor_occupation = [ 4594.,  6539., 11392., 13658.,  8862., 5092.,  3228., 14189.,  9413., 12760.,\n",
    "  9420., 11201., 12283.,  3676.,   13992., 12575., 13999., 14051., 14353., 9760., 14215.] # removed S7 @ 4562.,\n",
    "\n",
    "anchor_dict = sse_func.make_anchor_dict(anchors, 3425)\n",
    "\n",
    "gain_adresses = {}\n",
    "named_dirs = []\n",
    "for gain_idx, gain in enumerate(human_collection.collection):\n",
    "    _,_, named_dir, _ = sse_func.create_indexing(gain, anchors, anchor_occupation, anchor_dict)\n",
    "    named_dirs.append(named_dir)\n",
    "    gain.idx_dir = named_dir\n",
    "    gain.rev_idx_dir = {v: k for k, v in named_dir.items()}\n",
    "    gain_adresses[gain_idx] = gain.name\n",
    "\n",
    "rev_named_dirs = []\n",
    "for named_dir in named_dirs:\n",
    "    rev_dir = {v: k for k, v in named_dir.items()}\n",
    "    rev_named_dirs.append(rev_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### FUNCTION BLOCK FOR VARIANT FILE PARSING AND LOOP CONSTRUCTION.\n",
    "\n",
    "def retrieve_json_vars(name,jsons):\n",
    "    # A function retrieving variants from a JSON File as exported by TCGA/NCGA, returning them as a dictionary\n",
    "    # The list of {jsons} is filtered to find {name}\n",
    "    # Returns a loaded dict from the json.\n",
    "    try: \n",
    "        identifier = name.split(\"AGR\")[1][:2] # When its a normal aGPCR\n",
    "    except: \n",
    "        identifier = \"C\"+name.split(\"CELR\")[1][0]\n",
    "\n",
    "    json_identifier = identifier.lower()\n",
    "    json_file = [j for j in jsons if json_identifier in j.lower()]\n",
    "    #print(f\"DEBUG : retrieve_mutation_json: {json_file = }\" )\n",
    "    if len(json_file) > 1 : print(\"WARNING: Multiple matching JSON Files detected:\", json_file)\n",
    "    with open(json_file[0]) as j_open:\n",
    "        j_data = j_open.read()\n",
    "    mutation_data = json.loads(j_data) # Will contain a list of dicts\n",
    "    return mutation_data\n",
    "\n",
    "def retrieve_csv_vars(name, csv_files, filter_str=None, with_resid=True):\n",
    "    # A function retrieving variants from a CSV File as exported by gnomAD, returning them as a dictionary\n",
    "    # The list of {csv_files} is filtered to find {name}, filter_str is used to find specific mutations, i.e. missense\n",
    "    # With_resid also adds the residue and the standard AA at this position to the mutation entries.\n",
    "    # Returns a list of dict, where each dict corresponds to one mutation\n",
    "    try: \n",
    "        identifier = name.split(\"AGR\")[1][:2] # When its a normal aGPCR\n",
    "    except: \n",
    "        identifier = \"C\"+name.split(\"CELR\")[1][0]\n",
    "\n",
    "    csv_identifier = identifier.lower()\n",
    "    csv_file = [c for c in csv_files if csv_identifier in c.lower()]\n",
    "    print(f\"DEBUG : parse_snp_csv: {csv_file = }\" )\n",
    "    \n",
    "    # the target CSV files have 53 items per row. RETURNS columns: dict where key points to axis 1 index of np.array; val_arr: value array \n",
    "    with open(csv_file[0]) as cc:\n",
    "        data = cc.readlines()\n",
    "\n",
    "    # Pre-filter the data if specified.\n",
    "    if filter_str is not None:\n",
    "        newdata = [l for l in data[1:] if filter_str in l]\n",
    "        data = [data[0]]\n",
    "        for l in newdata: \n",
    "            data.append(l)\n",
    "\n",
    "    oneletter = {'Ala':'A','Arg':'R','Asn':'N','Asp':'D','Cys':'C','Gln':'Q','Glu':'E','Gly':'G','His':'H',\n",
    "                 'Ile':'I','Leu':'L','Lys':'K','Met':'M','Phe':'F','Pro':'P','Pyl':'O','Ser':'S','Thr':'T',\n",
    "                'Trp':'W','Tyr':'Y','Val':'V'}\n",
    "\n",
    "    variant_list = []\n",
    "    variant_keys = data[0].strip().split(\",\")\n",
    "\n",
    "    for i,row in enumerate(data[1:]):\n",
    "        row = re.sub(r'\\\".*\\\"', 'REMOVED', row)   # Removes combined irrelevant entries that contain a \",\" within string denotation. This confuses the split routine.\n",
    "        var_dict = {variant_keys[i]:val for i,val in enumerate(row.strip().split(\",\"))}\n",
    "        #print(\"DEBUG\", var_dict)\n",
    "        if with_resid:\n",
    "            var_string = var_dict['Protein Consequence']\n",
    "            #print(f\"DEBUG: {var_string = }\")\n",
    "            if len(var_string) < 2:\n",
    "                var_dict['resid'] = None\n",
    "                var_dict['resname'] = None\n",
    "            else:\n",
    "                var_dict['resname'] = oneletter[re.search(r'[A-z]{3}', var_string).group(0)]\n",
    "                var_dict['resid'] = int(re.search(r'[\\d]+', var_string).group(0))\n",
    "        variant_list.append(var_dict)\n",
    "\n",
    "    return variant_list\n",
    "\n",
    "def extract_variants(gain, files, resid_key):\n",
    "    # Wrapper function for getting variants / mutations in the correct format.\n",
    "    # Draws the list of vars/muts from the respective list of jsons/csvs abd gets the individual positions either at entry \n",
    "    # resid_key : 'x' (json from NCGA) / 'resid' (csv from gnomAD)\n",
    "    gain_name = gain.name\n",
    "\n",
    "    if files[0][-3:].lower() == 'csv':\n",
    "        print(\"List of CSV files detected.\")\n",
    "        vars = retrieve_csv_vars(gain_name, csvs, filter_str='missense_variant',with_resid=True)\n",
    "    else:\n",
    "        print(\"This should be a list of JSON Files.\")\n",
    "        v_dict = retrieve_json_vars(gain_name, files,)\n",
    "        vars = [mut for mut in v_dict['mutations']]\n",
    "        # returns a dictionary with a list of dicts, each dict corresponding to a var.\n",
    "    variant_dict = {}\n",
    "    positions = []\n",
    "    for var in vars:\n",
    "        # The \"x\" key is the identifying position!\n",
    "        resid = int(var[resid_key]) \n",
    "        #\n",
    "        # Here, one can define criteria for skipping said var. For now, use all.\n",
    "        #\n",
    "        # Add the receptor name to the var dictionary\n",
    "        var[\"receptor\"] = gain_name\n",
    "\n",
    "        if resid in variant_dict.keys():\n",
    "            variant_dict[resid].append(var)\n",
    "        else:\n",
    "            variant_dict[resid] = [var]\n",
    "        positions.append(resid)\n",
    "    return positions, variant_dict\n",
    "\n",
    "def get_loop_stats(indexing_dir, sequence):\n",
    "    # Returns a named dict with loop lengths, i.e. {\"H1-H2\":13, \"H8-S1\":12}\n",
    "    inverted_dir = {sse[0] : (sse[1],ki) for ki, sse in indexing_dir.items() if \"GPS\" not in ki} # The begin of each sse is here {0:(13, \"H2\")}\n",
    "    loop_loc = {}\n",
    "    loop_dir = {}\n",
    "    ordered_starts = sorted(inverted_dir.keys())\n",
    "    for i, sse_start in enumerate(ordered_starts):\n",
    "        if i == 0: \n",
    "            continue # Skip the first and go from the second SSE onwards, looking in N-terminal direction.\n",
    "        c_label = inverted_dir[sse_start][1]\n",
    "        n_end, n_label = inverted_dir[ordered_starts[i-1]]\n",
    "        loop_loc[f\"{n_label}-{c_label}\"] = (n_end, sse_start-1)\n",
    "        loop_dir[f\"{n_label}-{c_label}\"] = sequence[n_end+1:sse_start] # The one-letter-coded seqeuence. Will be a list of lists\n",
    "    return loop_loc, loop_dir\n",
    "\n",
    "def compose_vars(aGainCollection, files, resid_key, aa_key, fasta_offsets):\n",
    "    # Take the Gain Collection and retrieve variant/mutation info from the files list via extract_variants()\n",
    "    # Collect these for each GAIN in the aGainCollection.collection and compose them together to enable addressing each individual labeled position.\n",
    "    # Returns the generalized (aligned) variants/mutations and their respective counts\n",
    "    mismatch_flag = False\n",
    "    generalized_vars = {}\n",
    "    valid = 0\n",
    "    invalid = 0\n",
    "\n",
    "    for gain_ndx, gain in enumerate(aGainCollection.collection):\n",
    "        gain_valid = 0\n",
    "        gain_invalid = 0\n",
    "        fasta_offset = fasta_offsets[gain_ndx]\n",
    "        # Retrieve vars/mutations for respective receptor\n",
    "        positions, gain_pos_dict = extract_variants(gain, files, resid_key)\n",
    "        # Find all vars/mutations whose RESID (fasta) matches the corrected GAIN domain INTERVAL(for fasta resids)\n",
    "        within = [pos for pos in positions if pos in range(fasta_offset, fasta_offset + 1 + gain.end - gain.start)]\n",
    "        #print(\"DEBUG: vars/mutations within GAIN space\\n\", sorted(within))\n",
    "        # Evaluate the vars/mutations. If the residue in question (corrected index) is named (i.e. \"H6.50\"), get its resepective label.\n",
    "        for p_resid in gain_pos_dict.keys():\n",
    "            # Since fasta_offset maps to the first gain residue (i.e. 2027 with gain.start being 459), an offset needs to be set by the difference\n",
    "            corrected_resid = p_resid - fasta_offset + gain.start -1\n",
    "            #print(f\"DEBUG {corrected_resid = }, {p_resid = }\n",
    "            #   TRUE GAIN INTERVAL {fasta_offset+gain.start} - {fasta_offset+gain.end}\\nCORRECTED GAIN INTERVAL {gain.start}-{gain.end}\")\n",
    "            # map it to its corresponding label. If there is none, continue\n",
    "            try:\n",
    "                p_label = gain.rev_idx_dir[corrected_resid]\n",
    "                gain_valid +=1\n",
    "            except KeyError:\n",
    "                gain_invalid += 1\n",
    "                continue\n",
    "            # SANITY CHECK:\n",
    "            if gain_pos_dict[p_resid][0][aa_key][0] != gain.sequence[corrected_resid-gain.start]:\n",
    "                print(f\"{gain_pos_dict[p_resid][0][aa_key][0] = }\", {gain.sequence[corrected_resid-gain.start]})\n",
    "                print(f\"MISMATCH! {p_resid} -> {corrected_resid}\")\n",
    "                print(f\"{gain_pos_dict[p_resid][0][aa_key][0]} : {gain.sequence[corrected_resid-gain.start]} in GAIN\",\n",
    "                    f\"{gain.sequence[corrected_resid-gain.start-2:corrected_resid-gain.start+3]}\")\n",
    "                mismatch_flag = True\n",
    "            #else:\n",
    "                #print(f\"MATCH {p_resid} -> {corrected_resid} | {gain_mutation_dict[p_resid][0][aa_key][0]} : {gain.sequence[corrected_resid-gain.start]}\")\n",
    "            \n",
    "            # with the Label, map into generalized dictionary\n",
    "            if p_label not in generalized_vars.keys():\n",
    "                generalized_vars[p_label] = gain_pos_dict[p_resid]\n",
    "            else:\n",
    "                [generalized_vars[p_label].append(pos) for pos in gain_pos_dict[p_resid]]\n",
    "        valid += gain_valid\n",
    "        invalid += gain_invalid\n",
    "        print(\"VALID MAPPED POSITIONS: \", gain_valid, \"\\nINVALID POSITIONS     : \", gain_invalid, \"\\nTOTAL POSITIONS       : \", len(gain_pos_dict.keys()))\n",
    "    if mismatch_flag: \n",
    "        print(\"[WARNING]: MISMATCHES HAVE BEEN FOUND. PLEASE CHECK THE OUTPUT.\")\n",
    "    else: \n",
    "        print(\"[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\")\n",
    "    print(f\"TOTAL VARS WITHIN GAIN:\", valid, \"\\nTOTAL VARS OUTSIDE GAIN:\", invalid, \"\\nTOTAL VARS:\", valid+invalid)\n",
    "    \n",
    "    generalized_counts = {k:len(v) for k,v in generalized_vars.items()}\n",
    "\n",
    "    return generalized_vars, generalized_counts\n",
    "\n",
    "def loop2fasta(outfile, itemlist):\n",
    "    # Write the collected loop sequences to a FASTA file for later alignment.\n",
    "    with open(outfile, 'w') as out:\n",
    "        for subdict in itemlist:\n",
    "            out.write(f\">{subdict['name']}\\n{subdict['sequence']}\\n\")\n",
    "    print(\"Done with\", outfile)\n",
    "\n",
    "def compose_loop_vars(aGainCollection, files, resid_key, aa_key, fasta_offsets):\n",
    "    # Take the Gain Collection and retrieve variant/mutation info from the files list via extract_variants()\n",
    "    # Collect these for each GAIN in the aGainCollection.collection and compose them together to enable addressing each individual labeled position.\n",
    "    # Returns the generalized (aligned) variants/mutations and their respective counts\n",
    "    mismatch_flag = False\n",
    "    generalized_vars = {}\n",
    "    valid = 0\n",
    "    invalid = 0\n",
    "\n",
    "    loop_valid = 0\n",
    "    loop_invalid = 0\n",
    "    for idx, gain in enumerate(aGainCollection.collection):\n",
    "    # Retrieve vars/mutations for respective receptor\n",
    "        positions, gain_pos_dict = extract_variants(gain, files, resid_key)\n",
    "        fasta_offset = fasta_offsets[idx]\n",
    "        # Find all vars/mutations whose RESID (fasta) matches the corrected GAIN domain INTERVAL(for fasta resids)\n",
    "        within = [pos for pos in positions if pos in range(fasta_offset, fasta_offset + 1 + gain.end - gain.start)]\n",
    "        #print(\"DEBUG: vars/mutations within GAIN space\\n\", sorted(within))\n",
    "        idx_dir, _, _, _ = sse_func.create_indexing(gain, anchors, anchor_occupation, anchor_dict, outdir=None, offset=fasta_offsets[idx], silent=True, split_mode='double')\n",
    "        # i.e. 'H2-H3': (13, 20), 'H3-H4': (36, 42), 'H4-H5': (61, 75), 'H5-H6': (87, 89), ...., 'S11-S12': (316, 319), 'S12-S13': (327, 329)}\n",
    "        i_loc, _ = get_loop_stats(idx_dir, gain.sequence)\n",
    "        # Modify this to a position --> loop dict\n",
    "        loop_locations = {}\n",
    "        for loop, interval in i_loc.items():\n",
    "            for i in range(interval[0], interval[1]+1):\n",
    "                loop_locations[i+gain.start] = loop\n",
    "        # Evaluate the vars/mutations. If the residue in question (corrected index) is named (i.e. \"H6.50\"), get its resepective label.\n",
    "        print(\"DEBUG\")\n",
    "        #[print(i,k) for i,k in loop_locations.items()]\n",
    "\n",
    "        for p_resid in gain_pos_dict.keys():\n",
    "            # Since fasta_offset maps to the first gain residue (i.e. 2027 with gain.start being 459), an offset needs to be set by the difference\n",
    "            corrected_resid = p_resid - fasta_offset + gain.start -1\n",
    "            # Check in which loop this is located and collect it in a dictionary.\n",
    "            try:\n",
    "                loop_label = loop_locations[corrected_resid]\n",
    "                loop_valid +=1\n",
    "            except KeyError:\n",
    "                loop_invalid += 1\n",
    "                continue\n",
    "            # SANITY CHECK:\n",
    "            if gain_pos_dict[p_resid][0][aa_key][0] != gain.sequence[corrected_resid-gain.start]:\n",
    "                print(f\"{gain_pos_dict[p_resid][0][aa_key][0] = }\", {gain.sequence[corrected_resid-gain.start]})\n",
    "                print(f\"MISMATCH! {p_resid} -> {corrected_resid}\")\n",
    "                print(f\"{gain_pos_dict[p_resid][0][aa_key][0]} : {gain.sequence[corrected_resid-gain.start]} in GAIN\",\n",
    "                    f\"{gain.sequence[corrected_resid-gain.start-2:corrected_resid-gain.start+3]}\")\n",
    "                mismatch_flag = True\n",
    "            # with the Label, map into generalized dictionary\n",
    "            if loop_label not in generalized_vars.keys():\n",
    "                generalized_vars[loop_label] = gain_pos_dict[p_resid]\n",
    "            else:\n",
    "                [generalized_vars[loop_label].append(pos) for pos in gain_pos_dict[p_resid]]\n",
    "        valid += loop_valid\n",
    "        invalid += loop_invalid\n",
    "        print(\"VALID LOOP POSITIONS: \", loop_valid, \"\\nINVALID POSITIONS     : \", loop_invalid, \"\\nTOTAL POSITIONS       : \", len(gain_pos_dict.keys()))\n",
    "    if mismatch_flag: \n",
    "        print(\"[WARNING]: MISMATCHES HAVE BEEN FOUND. PLEASE CHECK THE OUTPUT.\")\n",
    "    else: \n",
    "        print(\"[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\")\n",
    "    print(f\"TOTAL VARS WITHIN GAIN:\", valid, \"\\nTOTAL VARS OUTSIDE GAIN:\", invalid, \"\\nTOTAL VARS:\", valid+invalid)\n",
    "    \n",
    "    generalized_counts = {k:len(v) for k,v in generalized_vars.items()}\n",
    "\n",
    "    return generalized_vars, generalized_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construct a Dictionary with info to all mutations at respective RESID. Also just make a list with the resids to summarize all entry occurences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c2.csv']\n",
      "VALID MAPPED POSITIONS:  24 \n",
      "INVALID POSITIONS     :  346 \n",
      "TOTAL POSITIONS       :  370\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c1.csv']\n",
      "VALID MAPPED POSITIONS:  23 \n",
      "INVALID POSITIONS     :  358 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c3.csv']\n",
      "VALID MAPPED POSITIONS:  15 \n",
      "INVALID POSITIONS     :  387 \n",
      "TOTAL POSITIONS       :  402\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f3.csv']\n",
      "VALID MAPPED POSITIONS:  12 \n",
      "INVALID POSITIONS     :  81 \n",
      "TOTAL POSITIONS       :  93\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b1.csv']\n",
      "VALID MAPPED POSITIONS:  28 \n",
      "INVALID POSITIONS     :  212 \n",
      "TOTAL POSITIONS       :  240\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b2.csv']\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  183 \n",
      "TOTAL POSITIONS       :  201\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b3.csv']\n",
      "VALID MAPPED POSITIONS:  52 \n",
      "INVALID POSITIONS     :  465 \n",
      "TOTAL POSITIONS       :  517\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l1.csv']\n",
      "VALID MAPPED POSITIONS:  24 \n",
      "INVALID POSITIONS     :  158 \n",
      "TOTAL POSITIONS       :  182\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l2.csv']\n",
      "VALID MAPPED POSITIONS:  45 \n",
      "INVALID POSITIONS     :  328 \n",
      "TOTAL POSITIONS       :  373\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g6.csv']\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  157 \n",
      "TOTAL POSITIONS       :  175\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f5.csv']\n",
      "VALID MAPPED POSITIONS:  27 \n",
      "INVALID POSITIONS     :  215 \n",
      "TOTAL POSITIONS       :  242\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g4.csv']\n",
      "VALID MAPPED POSITIONS:  27 \n",
      "INVALID POSITIONS     :  714 \n",
      "TOTAL POSITIONS       :  741\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/v1.csv']\n",
      "VALID MAPPED POSITIONS:  17 \n",
      "INVALID POSITIONS     :  1102 \n",
      "TOTAL POSITIONS       :  1119\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l3.csv']\n",
      "VALID MAPPED POSITIONS:  47 \n",
      "INVALID POSITIONS     :  334 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g5.csv']\n",
      "VALID MAPPED POSITIONS:  22 \n",
      "INVALID POSITIONS     :  54 \n",
      "TOTAL POSITIONS       :  76\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f1.csv']\n",
      "VALID MAPPED POSITIONS:  20 \n",
      "INVALID POSITIONS     :  134 \n",
      "TOTAL POSITIONS       :  154\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g3.csv']\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  67 \n",
      "TOTAL POSITIONS       :  85\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g7.csv']\n",
      "VALID MAPPED POSITIONS:  29 \n",
      "INVALID POSITIONS     :  111 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l4.csv']\n",
      "VALID MAPPED POSITIONS:  53 \n",
      "INVALID POSITIONS     :  162 \n",
      "TOTAL POSITIONS       :  215\n",
      "This should be a list of JSON Files.\n",
      "WARNING: Multiple matching JSON Files detected: ['../repo_data/gain_json/d2_gain.json', '../repo_data/gain_json/uncurated.d2_gain.json']\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/d2.csv']\n",
      "VALID MAPPED POSITIONS:  4 \n",
      "INVALID POSITIONS     :  0 \n",
      "TOTAL POSITIONS       :  4\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f2.csv']\n",
      "VALID MAPPED POSITIONS:  17 \n",
      "INVALID POSITIONS     :  86 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f4.csv']\n",
      "VALID MAPPED POSITIONS:  26 \n",
      "INVALID POSITIONS     :  114 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g2.csv']\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  159 \n",
      "TOTAL POSITIONS       :  177\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/a3.csv']\n",
      "VALID MAPPED POSITIONS:  21 \n",
      "INVALID POSITIONS     :  170 \n",
      "TOTAL POSITIONS       :  191\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e5.csv']\n",
      "VALID MAPPED POSITIONS:  22 \n",
      "INVALID POSITIONS     :  81 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g1.csv']\n",
      "VALID MAPPED POSITIONS:  13 \n",
      "INVALID POSITIONS     :  61 \n",
      "TOTAL POSITIONS       :  74\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/d1.csv']\n",
      "VALID MAPPED POSITIONS:  30 \n",
      "INVALID POSITIONS     :  151 \n",
      "TOTAL POSITIONS       :  181\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e3.csv']\n",
      "VALID MAPPED POSITIONS:  15 \n",
      "INVALID POSITIONS     :  95 \n",
      "TOTAL POSITIONS       :  110\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/a2.csv']\n",
      "VALID MAPPED POSITIONS:  25 \n",
      "INVALID POSITIONS     :  121 \n",
      "TOTAL POSITIONS       :  146\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e2.csv']\n",
      "VALID MAPPED POSITIONS:  19 \n",
      "INVALID POSITIONS     :  122 \n",
      "TOTAL POSITIONS       :  141\n",
      "This should be a list of JSON Files.\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e1.csv']\n",
      "VALID MAPPED POSITIONS:  33 \n",
      "INVALID POSITIONS     :  153 \n",
      "TOTAL POSITIONS       :  186\n",
      "[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\n",
      "TOTAL MUTATIONS WITHIN GAIN: 762 \n",
      "TOTAL MUTATIONS OUTSIDE GAIN: 6881 \n",
      "TOTAL MUTATIONS: 7643\n"
     ]
    }
   ],
   "source": [
    "generalized_counts = {}     # For every nomenclature label, the number of MUTATION entries is denoted\n",
    "generalized_mutations = {}  #   ^ The info behind the entries is here.\n",
    "gen_var_counts = {}         # For every nomenclature label, the number of VARIANT entries is denoted\n",
    "gen_vars = {}               #   ^ The info behind the entries is here.\n",
    "invalid = 0\n",
    "valid = 0\n",
    "\n",
    "mismatch_flag = False\n",
    "fasta_offsets = [2027, 2106, 2175, 479, 632, 588, 568, 528, 517, 579, 723, 2449, \n",
    "                 5619, 546, 0, 300, 22, 140, 112, 356, 154, 96, 365, 415, 265, \n",
    "                 170, 235, 120, 418, 269, 324]\n",
    "total_within = 0\n",
    "for gain_ndx, gain in enumerate(human_collection.collection):\n",
    "    gain_valid = 0\n",
    "    gain_invalid = 0\n",
    "    fasta_offset = fasta_offsets[gain_ndx]\n",
    "    # Retrieve mutations for respective receptor\n",
    "    positions, gain_mutation_dict = extract_variants(gain, jsons, 'x')\n",
    "    # Retrieve variants for respective receptor\n",
    "    var_positions, var_dict = extract_variants(gain, csvs, 'resid')\n",
    "    # Find all mutations whose RESID (fasta) matches the corrected INTERVAL(for fasta resids)\n",
    "    within = [pos for pos in positions if pos in range(fasta_offset, fasta_offset + 1 + gain.end - gain.start)]\n",
    "    #print(\"DEBUG: Mutations within GAIN space\\n\", sorted(within))\n",
    "    # Go through the gain_mutation_dict and add into the two dicts\n",
    "    total_within += len(within)\n",
    "    \"\"\"print(\"EXTENT OF MUTATIONS IN TARGET RESIDUE INTERVAL:\\nINTERVAL\\t\\t:\",fasta_offset, fasta_offset + 1 + gain.end - gain.start,\n",
    "        \"\\n FIRST OCC\\t\\t:\",within[0], \n",
    "        \"\\n LAST OCC\\t\\t:\",within[-1],\n",
    "        \"\\n TOTAL MUT\\t\\t:\",len(positions),\n",
    "        \"\\n WITHIN INTERVAL\\t:\", len(within),\n",
    "        \"\\n DEBUG: RESID WITHIN\", within\n",
    "        )\"\"\"\n",
    "    # Evaluate the mutations. If the residue in question (corrected index) is named (i.e. \"H6.50\"), get its resepective label.\n",
    "    for mutated_resid in gain_mutation_dict.keys():\n",
    "        # Since fasta_offset maps to the first gain residue (i.e. 2027 with gain.start being 459), an offset needs to be set by the difference\n",
    "        corrected_resid = mutated_resid - fasta_offset + gain.start -1\n",
    "        # map it to its corresponding label. If there is none, continue\n",
    "        try:\n",
    "            mutated_label = gain.rev_idx_dir[corrected_resid]\n",
    "            #print(mutated_label)\n",
    "            gain_valid +=1\n",
    "        except KeyError:\n",
    "            gain_invalid += 1\n",
    "            continue\n",
    "        # SANITY CHECK:\n",
    "        if gain_mutation_dict[mutated_resid][0]['aa_change'][0] != gain.sequence[corrected_resid-gain.start]:\n",
    "            print(f\"MISMATCH! {mutated_resid} -> {corrected_resid}\",\n",
    "                 f\"{gain_mutation_dict[mutated_resid][0]['aa_change'][0]} : {gain.sequence[corrected_resid-gain.start]} in GAIN\",\n",
    "                 f\"{gain.sequence[corrected_resid-gain.start-2:corrected_resid-gain.start+3]}\")\n",
    "            mismatch_flag = True\n",
    "        #else:\n",
    "            #print(f\"MATCH {mutated_resid} -> {corrected_resid} | {gain_mutation_dict[mutated_resid][0]['aa_change'][0]} : {gain.sequence[corrected_resid-gain.start]}\")\n",
    "        \n",
    "        # with the Label, map into generalized dictionary\n",
    "        if mutated_label not in generalized_mutations.keys():\n",
    "            generalized_mutations[mutated_label] = gain_mutation_dict[mutated_resid]\n",
    "        else:\n",
    "            [generalized_mutations[mutated_label].append(mutation) for mutation in gain_mutation_dict[mutated_resid]]\n",
    "    valid += gain_valid\n",
    "    invalid += gain_invalid\n",
    "    print(\"VALID MAPPED POSITIONS: \", gain_valid, \"\\nINVALID POSITIONS     : \", gain_invalid, \"\\nTOTAL POSITIONS       : \", len(gain_mutation_dict.keys()))\n",
    "if mismatch_flag: \n",
    "    print(\"[WARNING]: MISMATCHES HAVE BEEN FOUND. PLEASE CHECK THE OUTPUT.\")\n",
    "else: \n",
    "    print(\"[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\")\n",
    "print(f\"TOTAL MUTATIONS WITHIN GAIN:\", valid, \"\\nTOTAL MUTATIONS OUTSIDE GAIN:\", invalid, \"\\nTOTAL MUTATIONS:\", valid+invalid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In L3, there is a mismatch between canonical (UniProtKB) and the GDC form (equiv. Isoform 4) - see MSA mapping\n",
    "D2 (2/2)   | GDC ENST00000334810 = D2-201 (971 aa) - Uniprot Q7Z7M1 (963 aa) - Files OK - remap the four mutations in question : 496, 556, 460, 386, 507 -->  514, 529, 478, 404, X --> MSA mapping manual curation of these four mutations\n",
    "F4 (23/26) | GDC ENST00000283303 (695 aa) - Uniprot Q8IZF3 (695 aa) - Re-Download resolved all errors\n",
    "E2 (6/20)  | GDC ENST00000315576 (823 aa) - Uniprot Q9UHX3 (823 aa) - Re-Download resolved all errors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  24 \n",
      "INVALID POSITIONS     :  346 \n",
      "TOTAL POSITIONS       :  370\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  23 \n",
      "INVALID POSITIONS     :  358 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  15 \n",
      "INVALID POSITIONS     :  387 \n",
      "TOTAL POSITIONS       :  402\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  12 \n",
      "INVALID POSITIONS     :  81 \n",
      "TOTAL POSITIONS       :  93\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  28 \n",
      "INVALID POSITIONS     :  212 \n",
      "TOTAL POSITIONS       :  240\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  183 \n",
      "TOTAL POSITIONS       :  201\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  52 \n",
      "INVALID POSITIONS     :  465 \n",
      "TOTAL POSITIONS       :  517\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  24 \n",
      "INVALID POSITIONS     :  158 \n",
      "TOTAL POSITIONS       :  182\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  45 \n",
      "INVALID POSITIONS     :  328 \n",
      "TOTAL POSITIONS       :  373\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  157 \n",
      "TOTAL POSITIONS       :  175\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  27 \n",
      "INVALID POSITIONS     :  215 \n",
      "TOTAL POSITIONS       :  242\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  27 \n",
      "INVALID POSITIONS     :  714 \n",
      "TOTAL POSITIONS       :  741\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  17 \n",
      "INVALID POSITIONS     :  1102 \n",
      "TOTAL POSITIONS       :  1119\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  47 \n",
      "INVALID POSITIONS     :  334 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  22 \n",
      "INVALID POSITIONS     :  54 \n",
      "TOTAL POSITIONS       :  76\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  20 \n",
      "INVALID POSITIONS     :  134 \n",
      "TOTAL POSITIONS       :  154\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  67 \n",
      "TOTAL POSITIONS       :  85\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  29 \n",
      "INVALID POSITIONS     :  111 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  53 \n",
      "INVALID POSITIONS     :  162 \n",
      "TOTAL POSITIONS       :  215\n",
      "This should be a list of JSON Files.\n",
      "WARNING: Multiple matching JSON Files detected: ['../repo_data/gain_json/d2_gain.json', '../repo_data/gain_json/uncurated.d2_gain.json']\n",
      "VALID MAPPED POSITIONS:  4 \n",
      "INVALID POSITIONS     :  0 \n",
      "TOTAL POSITIONS       :  4\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  17 \n",
      "INVALID POSITIONS     :  86 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  26 \n",
      "INVALID POSITIONS     :  114 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  18 \n",
      "INVALID POSITIONS     :  159 \n",
      "TOTAL POSITIONS       :  177\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  21 \n",
      "INVALID POSITIONS     :  170 \n",
      "TOTAL POSITIONS       :  191\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  22 \n",
      "INVALID POSITIONS     :  81 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  13 \n",
      "INVALID POSITIONS     :  61 \n",
      "TOTAL POSITIONS       :  74\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  30 \n",
      "INVALID POSITIONS     :  151 \n",
      "TOTAL POSITIONS       :  181\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  15 \n",
      "INVALID POSITIONS     :  95 \n",
      "TOTAL POSITIONS       :  110\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  25 \n",
      "INVALID POSITIONS     :  121 \n",
      "TOTAL POSITIONS       :  146\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  19 \n",
      "INVALID POSITIONS     :  122 \n",
      "TOTAL POSITIONS       :  141\n",
      "This should be a list of JSON Files.\n",
      "VALID MAPPED POSITIONS:  33 \n",
      "INVALID POSITIONS     :  153 \n",
      "TOTAL POSITIONS       :  186\n",
      "[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\n",
      "TOTAL VARS WITHIN GAIN: 762 \n",
      "TOTAL VARS OUTSIDE GAIN: 6881 \n",
      "TOTAL VARS: 7643\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c2.csv']\n",
      "VALID MAPPED POSITIONS:  75 \n",
      "INVALID POSITIONS     :  1173 \n",
      "TOTAL POSITIONS       :  1248\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c1.csv']\n",
      "VALID MAPPED POSITIONS:  85 \n",
      "INVALID POSITIONS     :  1305 \n",
      "TOTAL POSITIONS       :  1390\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/c3.csv']\n",
      "VALID MAPPED POSITIONS:  62 \n",
      "INVALID POSITIONS     :  1199 \n",
      "TOTAL POSITIONS       :  1261\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f3.csv']\n",
      "VALID MAPPED POSITIONS:  75 \n",
      "INVALID POSITIONS     :  396 \n",
      "TOTAL POSITIONS       :  471\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b1.csv']\n",
      "VALID MAPPED POSITIONS:  33 \n",
      "INVALID POSITIONS     :  428 \n",
      "TOTAL POSITIONS       :  461\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b2.csv']\n",
      "VALID MAPPED POSITIONS:  44 \n",
      "INVALID POSITIONS     :  447 \n",
      "TOTAL POSITIONS       :  491\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/b3.csv']\n",
      "VALID MAPPED POSITIONS:  76 \n",
      "INVALID POSITIONS     :  499 \n",
      "TOTAL POSITIONS       :  575\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l1.csv']\n",
      "VALID MAPPED POSITIONS:  61 \n",
      "INVALID POSITIONS     :  475 \n",
      "TOTAL POSITIONS       :  536\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l2.csv']\n",
      "VALID MAPPED POSITIONS:  65 \n",
      "INVALID POSITIONS     :  527 \n",
      "TOTAL POSITIONS       :  592\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g6.csv']\n",
      "VALID MAPPED POSITIONS:  48 \n",
      "INVALID POSITIONS     :  415 \n",
      "TOTAL POSITIONS       :  463\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f5.csv']\n",
      "VALID MAPPED POSITIONS:  69 \n",
      "INVALID POSITIONS     :  514 \n",
      "TOTAL POSITIONS       :  583\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g4.csv']\n",
      "VALID MAPPED POSITIONS:  42 \n",
      "INVALID POSITIONS     :  990 \n",
      "TOTAL POSITIONS       :  1032\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/v1.csv']\n",
      "VALID MAPPED POSITIONS:  54 \n",
      "INVALID POSITIONS     :  2603 \n",
      "TOTAL POSITIONS       :  2657\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l3.csv']\n",
      "VALID MAPPED POSITIONS:  76 \n",
      "INVALID POSITIONS     :  435 \n",
      "TOTAL POSITIONS       :  511\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g5.csv']\n",
      "VALID MAPPED POSITIONS:  46 \n",
      "INVALID POSITIONS     :  180 \n",
      "TOTAL POSITIONS       :  226\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f1.csv']\n",
      "VALID MAPPED POSITIONS:  60 \n",
      "INVALID POSITIONS     :  354 \n",
      "TOTAL POSITIONS       :  414\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g3.csv']\n",
      "VALID MAPPED POSITIONS:  44 \n",
      "INVALID POSITIONS     :  217 \n",
      "TOTAL POSITIONS       :  261\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g7.csv']\n",
      "VALID MAPPED POSITIONS:  61 \n",
      "INVALID POSITIONS     :  273 \n",
      "TOTAL POSITIONS       :  334\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/l4.csv']\n",
      "VALID MAPPED POSITIONS:  92 \n",
      "INVALID POSITIONS     :  208 \n",
      "TOTAL POSITIONS       :  300\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/d2.csv']\n",
      "VALID MAPPED POSITIONS:  75 \n",
      "INVALID POSITIONS     :  338 \n",
      "TOTAL POSITIONS       :  413\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f2.csv']\n",
      "VALID MAPPED POSITIONS:  49 \n",
      "INVALID POSITIONS     :  275 \n",
      "TOTAL POSITIONS       :  324\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/f4.csv']\n",
      "VALID MAPPED POSITIONS:  55 \n",
      "INVALID POSITIONS     :  274 \n",
      "TOTAL POSITIONS       :  329\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g2.csv']\n",
      "VALID MAPPED POSITIONS:  29 \n",
      "INVALID POSITIONS     :  222 \n",
      "TOTAL POSITIONS       :  251\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/a3.csv']\n",
      "VALID MAPPED POSITIONS:  74 \n",
      "INVALID POSITIONS     :  494 \n",
      "TOTAL POSITIONS       :  568\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e5.csv']\n",
      "VALID MAPPED POSITIONS:  59 \n",
      "INVALID POSITIONS     :  311 \n",
      "TOTAL POSITIONS       :  370\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/g1.csv']\n",
      "VALID MAPPED POSITIONS:  46 \n",
      "INVALID POSITIONS     :  255 \n",
      "TOTAL POSITIONS       :  301\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/d1.csv']\n",
      "VALID MAPPED POSITIONS:  63 \n",
      "INVALID POSITIONS     :  346 \n",
      "TOTAL POSITIONS       :  409\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e3.csv']\n",
      "VALID MAPPED POSITIONS:  53 \n",
      "INVALID POSITIONS     :  213 \n",
      "TOTAL POSITIONS       :  266\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/a2.csv']\n",
      "VALID MAPPED POSITIONS:  60 \n",
      "INVALID POSITIONS     :  451 \n",
      "TOTAL POSITIONS       :  511\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e2.csv']\n",
      "VALID MAPPED POSITIONS:  51 \n",
      "INVALID POSITIONS     :  306 \n",
      "TOTAL POSITIONS       :  357\n",
      "List of CSV files detected.\n",
      "DEBUG : parse_snp_csv: csv_file = ['/home/hildilab/projects/agpcr_nom/snp_mane/e1.csv']\n",
      "VALID MAPPED POSITIONS:  65 \n",
      "INVALID POSITIONS     :  315 \n",
      "TOTAL POSITIONS       :  380\n",
      "[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\n",
      "TOTAL VARS WITHIN GAIN: 1847 \n",
      "TOTAL VARS OUTSIDE GAIN: 16438 \n",
      "TOTAL VARS: 18285\n"
     ]
    }
   ],
   "source": [
    "# Generate the dictionaries for the SNPs and the Cancer mutations.\n",
    "fasta_offsets = [2027, 2106, 2175, 479, 632, 588, 568, 528, 517, 579, 723, 2449, \n",
    "                 5619, 546, 0, 300, 22, 140, 112, 356, 154, 96, 365, 415, 265, \n",
    "                 170, 235, 120, 418, 269, 324]\n",
    "mutations, mut_counts = compose_vars(human_collection, jsons, 'x', 'aa_change', fasta_offsets)\n",
    "snps, snp_counts = compose_vars(human_collection, csvs, 'resid', 'resname', fasta_offsets)\n",
    "\n",
    "#print(valid, invalid, valid+invalid, total_within)\n",
    "# For each GAIN domain in the collection, count the occurence of a specific label\n",
    "occupancy_dict = {}\n",
    "for gain in human_collection.collection:\n",
    "    #rev_dir = {v: k for k, v in gain.named_dir.items()}\n",
    "    for key in gain.idx_dir.keys():\n",
    "        if key not in occupancy_dict.keys():\n",
    "            occupancy_dict[key] = 1\n",
    "        else:\n",
    "            occupancy_dict[key] += 1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting the Cancer-enrichement scores of Positions. For each SSE, count the mutations/SNPs at each position and generate a score according to \n",
    "Wright et al. 2019 https://doi.org/10.1038/s41467-019-08630-2 - cancer-enriched positions will have a positive score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compose_y(x_range, sse, count_dict):\n",
    "    y_vals = np.zeros(shape=len(x_range))\n",
    "    for i, val in enumerate(x_range):\n",
    "        k = f\"{sse}{val}\"\n",
    "        try:\n",
    "            y_vals[i] = count_dict[k]\n",
    "        except:\n",
    "            pass\n",
    "    return y_vals\n",
    "\n",
    "def score(m,v):\n",
    "    mmax = max(m)\n",
    "    vmax = max(v)\n",
    "    return [math.log10( ((m[i]/mmax)+1) / ((v[i]/vmax)+1) ) for i in range(len(m))]            \n",
    "\n",
    "order = [f\"H{str(num)}.\" for num in range(1,9)]\n",
    "[order.append(f\"S{str(num)}.\") for num in range(1,14)]\n",
    "order.append(\"GPS\")\n",
    "\n",
    "for sse in order:\n",
    "# mutations, mut_counts\n",
    "# snps, snp_counts\n",
    "    varkeys = [k for k in mut_counts.keys() if sse in k]+[k for k in snp_counts.keys() if sse in k]\n",
    "    #print(subkeys)\n",
    "    if sse == \"GPS\":\n",
    "        x_range = (1,2,3)\n",
    "        mut_y = [mut_counts[\"GPS-1\"], mut_counts[\"GPS+1\"], 0]\n",
    "        var_y = [mut_counts[\"GPS-1\"], mut_counts[\"GPS+1\"], 0]\n",
    "    if sse != 'GPS':\n",
    "        x_positions = [int(x.split('.')[-1]) for x in varkeys]\n",
    "        x_range = range(min(x_positions), max(x_positions)+1)\n",
    "        mut_y = compose_y(x_range, sse, mut_counts)\n",
    "        var_y = compose_y(x_range, sse, snp_counts)\n",
    "\n",
    "    fig, ax = plt.subplots(1,3, figsize=(8,2))\n",
    "    fig.tight_layout()\n",
    "    fig.set_facecolor('w')\n",
    "    score_y = score(mut_y,var_y)\n",
    "    y_max = max(var_y)\n",
    "    xlab = [f\"{sse}{p}\" for p in x_range][::3]\n",
    "    for a in ax:    #ax = plt.subplot(1,3,2)\n",
    "        a.spines['top'].set_visible(False)\n",
    "        a.spines['right'].set_visible(False)\n",
    "        a.spines['bottom'].set_visible(False)\n",
    "        a.spines['left'].set_visible(False)\n",
    "        a.xaxis.set_minor_locator(MultipleLocator(1)) #AutoMinorLocator())\n",
    "        a.xaxis.set_major_locator(FixedLocator([a for a in range(2,100,3)]))#MultipleLocator(3))\n",
    "        a.tick_params(which='both', width=2)\n",
    "        a.tick_params(which='major', length=8)\n",
    "        a.tick_params(which='minor', length=6)\n",
    "\n",
    "    ax[0].set_ylim(-0.2, y_max)\n",
    "    ax[0].set_title(f'Cancer Variation ({sse[:-1]})')\n",
    "    ax[0].bar(x_range, mut_y, width=1.1, color='silver')  \n",
    "\n",
    "    ax[1].set_ylim(-0.2, y_max)\n",
    "    ax[1].set_title(f'Natural Variation ({sse[:-1]})')\n",
    "    ax[1].bar(x_range, var_y, width=1.1, color='silver')\n",
    "\n",
    "    ax[2].set_title(f'Combined Variation ({sse[:-1]})')\n",
    "    ax[2].hlines(0, min(x_range), max(x_range), color='dimgray', linewidth=0.7)\n",
    "    for i in range(len(x_range)):\n",
    "        if score_y[i] < 0:\n",
    "            cx = 'xkcd:blurple'\n",
    "        else:\n",
    "            cx = 'xkcd:tomato red'\n",
    "            if score_y[i] > 0.05:\n",
    "                label = f'{sse}{x_range[i]}'\n",
    "                print(f'{label},{round(score_y[i],2)}')\n",
    "        ax[2].bar(x_range[i], score_y[i], color=cx)\n",
    "\n",
    "    for a in ax:\n",
    "        a.set_xticklabels([f'{sse}{str(int(v))}' for v in a.get_xticks()], rotation=90)\n",
    "        \n",
    "    #fig.savefig(f\"../fig/{sse}.variants.svg\", bbox_inches='tight')\n",
    "    #plt.show()\n",
    "    #plt.show(fig)\n",
    "    #plt.close()\n",
    "    plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H2.43 4\n",
      "H2.45 8\n",
      "H2.48 6\n",
      "H2.49 8\n",
      "H3.41 9\n",
      "H3.42 11\n",
      "H3.45 15\n",
      "H3.46 11\n",
      "H3.47 6\n",
      "H3.49 18\n",
      "H3.51 14\n",
      "H3.52 15\n",
      "H3.53 8\n",
      "H3.54 9\n",
      "H4.40 19\n",
      "H4.43 15\n",
      "H4.46 11\n",
      "H4.50 11\n",
      "H4.51 7\n",
      "H4.52 19\n",
      "H4.56 10\n",
      "H5.48 11\n",
      "H5.51 13\n",
      "H5.52 11\n",
      "H5.55 21\n",
      "H6.51 7\n",
      "H6.52 9\n",
      "H6.58 5\n",
      "H8.47 12\n",
      "H8.48 23\n",
      "H8.49 30\n",
      "H8.50 13\n",
      "H8.53 16\n",
      "H8.54 12\n",
      "H8.56 25\n",
      "H8.58 17\n",
      "H8.63 15\n",
      "H8.66 4\n",
      "S1.52 11\n",
      "S2.50 17\n",
      "S2.51 11\n",
      "S2.52 12\n",
      "S2.54 11\n",
      "S2.56 19\n",
      "S2.58 4\n",
      "S4.49 13\n",
      "S5.48 8\n",
      "S5.50 12\n",
      "S5.52 7\n",
      "S5.53 17\n",
      "S5.54 15\n",
      "S5.56 2\n",
      "S7.50 14\n",
      "S7.51 11\n",
      "S7.52 10\n",
      "S7.55 18\n",
      "S8.45 3\n",
      "S8.49 5\n",
      "S8.50 13\n",
      "S8.51 9\n",
      "S8.52 15\n",
      "S8.53 16\n",
      "S8.55 18\n",
      "S9.54 14\n",
      "S10.51 18\n",
      "S11.53 19\n",
      "S11.54 19\n",
      "S11.55 11\n",
      "S12.48 20\n",
      "S12.51 29\n",
      "S12.52 7\n",
      "GPS-1 12\n",
      "GPS+1 17\n",
      "S13.48 9\n",
      "S13.51 25\n",
      "S13.52 12\n",
      "S13.49 15\n",
      "S13.46 14\n",
      "S12.49 28\n",
      "S11.51 26\n",
      "S10.49 25\n",
      "S9.53 5\n",
      "S9.51 17\n",
      "S7.54 13\n",
      "S7.53 18\n",
      "S5.51 18\n",
      "S5.49 15\n",
      "S4.51 14\n",
      "S4.50 13\n",
      "S2.55 13\n",
      "S1.50 13\n",
      "H8.68 1\n",
      "H8.67 2\n",
      "H8.65 1\n",
      "H8.62 8\n",
      "H8.61 12\n",
      "H8.55 18\n",
      "H8.52 23\n",
      "H8.46 17\n",
      "H8.45 5\n",
      "H6.59 6\n",
      "H5.56 10\n",
      "H5.50 11\n",
      "H5.49 20\n",
      "H5.47 11\n",
      "H5.46 9\n",
      "H4.58 8\n",
      "H4.57 4\n",
      "H4.53 17\n",
      "H4.49 5\n",
      "H4.45 17\n",
      "H4.44 18\n",
      "H4.41 10\n",
      "H3.55 7\n",
      "H3.43 8\n",
      "H3.40 6\n",
      "H2.47 6\n",
      "H2.46 6\n",
      "H2.42 6\n",
      "H2.41 7\n",
      "H2.39 5\n",
      "H2.38 2\n",
      "GPS-2 19\n",
      "S12.53 2\n",
      "S8.54 10\n",
      "S8.48 3\n",
      "S8.46 3\n",
      "S5.55 9\n",
      "S4.53 9\n",
      "S1.49 5\n",
      "H8.60 11\n",
      "H6.56 4\n",
      "H6.55 8\n",
      "H6.50 7\n",
      "H6.49 6\n",
      "H5.53 10\n",
      "H4.55 7\n",
      "H4.48 21\n",
      "H3.44 6\n",
      "H2.51 5\n",
      "H2.40 5\n",
      "S13.50 6\n",
      "S11.52 18\n",
      "S11.50 10\n",
      "S9.50 9\n",
      "S9.47 10\n",
      "S8.44 2\n",
      "S8.43 3\n",
      "S7.57 2\n",
      "S7.56 5\n",
      "S4.48 8\n",
      "S3.49 11\n",
      "S3.48 7\n",
      "S3.47 2\n",
      "S1.48 6\n",
      "H8.51 14\n",
      "H7.53 2\n",
      "H7.50 5\n",
      "H5.45 10\n",
      "H5.43 16\n",
      "H4.47 13\n",
      "H3.50 10\n",
      "H3.48 15\n",
      "H3.39 4\n",
      "H2.52 8\n",
      "H1.45 1\n",
      "H1.53 4\n",
      "H1.55 6\n",
      "H6.48 8\n",
      "H6.60 3\n",
      "H8.59 7\n",
      "S6.50 3\n",
      "S10.50 10\n",
      "S9.52 8\n",
      "S9.46 8\n",
      "H4.59 5\n",
      "H3.56 5\n",
      "H3.38 2\n",
      "H3.37 2\n",
      "H3.36 2\n",
      "H1.54 1\n",
      "H1.51 3\n",
      "H1.49 5\n",
      "H1.48 3\n",
      "H1.46 1\n",
      "H1.56 3\n",
      "H1.58 5\n",
      "H1.59 2\n",
      "H4.42 13\n",
      "H4.54 9\n",
      "H5.44 13\n",
      "H5.54 12\n",
      "H8.57 7\n",
      "S2.53 10\n",
      "S3.51 7\n",
      "S6.49 6\n",
      "S9.48 16\n",
      "S9.49 10\n",
      "S12.47 14\n",
      "S13.47 7\n",
      "S8.47 5\n",
      "S6.52 6\n",
      "S6.51 6\n",
      "S1.56 1\n",
      "S1.55 3\n",
      "S1.54 5\n",
      "S1.51 8\n",
      "H8.44 5\n",
      "H8.41 3\n",
      "H6.53 5\n",
      "H5.41 2\n",
      "H5.29 2\n",
      "H4.63 2\n",
      "H4.60 4\n",
      "H4.62 3\n",
      "H5.30 2\n",
      "H5.31 1\n",
      "H5.34 1\n",
      "H5.38 1\n",
      "S5.47 10\n",
      "S9.55 2\n",
      "H8.42 3\n",
      "H8.43 5\n",
      "S3.50 9\n",
      "S3.52 5\n",
      "S3.53 2\n",
      "S7.47 2\n",
      "S4.52 7\n",
      "S4.46 8\n",
      "H3.61 1\n",
      "H3.57 6\n",
      "H2.44 4\n",
      "S3.54 2\n",
      "H2.54 1\n",
      "H2.59 1\n",
      "H2.62 1\n",
      "H5.59 1\n",
      "H5.60 2\n",
      "H7.38 1\n",
      "H7.40 1\n",
      "H7.44 1\n",
      "H7.45 4\n",
      "H7.48 3\n",
      "H7.49 5\n",
      "S3.56 1\n",
      "S9.45 3\n",
      "H2.50 5\n",
      "H5.40 2\n",
      "H5.42 2\n",
      "H8.40 7\n",
      "S1.53 2\n",
      "S9.56 3\n",
      "S10.47 2\n",
      "S12.50 7\n",
      "H4.29 1\n",
      "H4.35 2\n",
      "H4.37 2\n",
      "H4.38 2\n",
      "H4.39 1\n",
      "S7.44 4\n",
      "S7.46 2\n",
      "S7.48 3\n",
      "S3.44 2\n",
      "H6.57 5\n",
      "H3.60 1\n",
      "H3.59 3\n",
      "H5.57 4\n",
      "H5.61 1\n",
      "S5.45 1\n",
      "S7.45 2\n",
      "S7.49 3\n",
      "S13.53 5\n",
      "H4.36 1\n",
      "H7.47 4\n",
      "S4.54 2\n",
      "S9.43 1\n",
      "S9.44 2\n",
      "S13.54 2\n",
      "S10.48 3\n",
      "S7.58 1\n",
      "S2.57 5\n",
      "H5.39 1\n",
      "H2.58 1\n",
      "H2.56 1\n",
      "H2.55 3\n",
      "H1.52 2\n",
      "H2.53 1\n",
      "H3.58 3\n",
      "H6.54 2\n",
      "S6.48 3\n",
      "S6.53 4\n",
      "S2.48 2\n",
      "S2.49 2\n",
      "S4.45 5\n",
      "S11.56 3\n",
      "S4.44 2\n",
      "H4.61 1\n",
      "S13.55 3\n",
      "H3.62 1\n",
      "H3.66 2\n",
      "H1.47 1\n",
      "H1.50 1\n",
      "S3.46 4\n",
      "H8.64 1\n",
      "H5.58 1\n",
      "S1.47 1\n",
      "H4.34 2\n",
      "H6.37 1\n",
      "H6.39 1\n",
      "H6.40 1\n",
      "H6.41 1\n",
      "H6.43 1\n",
      "H6.45 1\n",
      "H6.46 1\n"
     ]
    }
   ],
   "source": [
    "#generalized_mutations.keys()\n",
    "\n",
    "names = []\n",
    "nums = []\n",
    "occus = []\n",
    "# Construct a sequence of secondary structural elements.\n",
    "order = [f\"H{str(num)}.\" for num in range(1,9)]\n",
    "[order.append(f\"S{str(num)}.\") for num in range(1,13)]\n",
    "order.append(\"GPS-2\")\n",
    "order.append(\"GPS-1\")\n",
    "order.append(\"GPS+1\")\n",
    "order.append(\"S13.\")\n",
    "\n",
    "for k, v in snps.items():\n",
    "    print(k, len(v))\n",
    "\n",
    "#print(order)\n",
    "# Enumerate SSE Sequence and find the corresponding keys, smash together into $names, $nums, $occus\n",
    "for sse in order:\n",
    "    #print(sse)\n",
    "    keys = sorted([ki for ki in generalized_mutations.keys() if sse in ki])\n",
    "    for ki in keys:\n",
    "        #print(ki,  len(generalized_mutations[ki]))#, \"\\tOCC\",occupancy_dict[ki])\n",
    "        pass\n",
    "        #mut_names.append(ki)                                # The position\n",
    "        #mut_nums.append(len(generalized_mutations[ki]))     # Number of mutations at this position\n",
    "        #mut_occus.append(occupancy_dict[ki])                # The mutations (data) in question\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f708db767c0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAD4CAYAAADCWc2+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAS6UlEQVR4nO3c0U9b5xnH8Z8ri4tIbW4aUlPTZd5h1GIhbWpWbpuEsLDKVRupYkoXpKxiyZiydlqq/AFjYVErNRJTI180IqsEF7uIEcSspN2kqVrEPCpNnXdhZaDZjhPIEtRV1UbmvruowspMwMuh8Nj5fu6Oz3vsh7dW/ZU5IeCccwIAADDogc0eAAAA4G4IFQAAYBahAgAAzCJUAACAWYQKAAAwK7jZA9yLhx9+WDt27NjsMQAAwDqYnZ3VjRs3VjxXlaGyY8cOpdPpzR4DAACsg1gsdtdz/OoHAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYtS6hMjExoebmZnmep4GBgbLzzjkdP35cnueptbVV09PTy86XSiU9+eSTevbZZ9djHAAAUCN8h0qpVFJfX59SqZQymYyGh4eVyWSWrUmlUspms8pms0okEjp27Niy82fOnFE0GvU7CgAAqDG+Q2Vqakqe5ykSiaiurk7d3d1KJpPL1iSTSR0+fFiBQEDt7e1aWFhQsViUJOXzeY2Pj+vll1/2OwoAAKgxvkOlUCiosbFx6TgcDqtQKFS85pVXXtHp06f1wAOrj5JIJBSLxRSLxTQ/P+93bAAAUAV8h4pzruyxQCBQ0ZqxsTHV19frqaeeWvN1ent7lU6nlU6ntW3btnsfGAAAVA3foRIOh5XL5ZaO8/m8GhoaKlrzwQcfaHR0VDt27FB3d7fef/99vfTSS35HAgAANcJ3qLS1tSmbzWpmZkaLi4saGRlRPB5ftiYej+v8+fNyzuny5cvaunWrQqGQTp06pXw+r9nZWY2MjGjPnj165513/I4EAABqRND3EwSDGhwcVGdnp0qlko4cOaKWlhadPXtWknT06FF1dXXp4sWL8jxPW7Zs0blz53wPDgAAal/ArXQDiXGxWEzpdHqzxwAAAOtgtc91/jItAAAwi1ABAABmESoAAMAsQgUAAJhFqAAAALMIFQAAYBahAgAAzCJUAACAWYQKAAAwi1ABAABmESoAAMAsQgUAAJhFqAAAALMIFQAAYBahAgAAzCJUAACAWYQKAAAwi1ABAABmESoAAMAsQgUAAJhFqAAAALMIFQAAYBahAgAAzCJUAACAWYQKAAAwi1ABAABmESoAAMAsQgUAAJhFqAAAALMIFQAAYNa6hMrExISam5vleZ4GBgbKzjvndPz4cXmep9bWVk1PT0uScrmcnnnmGUWjUbW0tOjMmTPrMQ4AAKgRvkOlVCqpr69PqVRKmUxGw8PDymQyy9akUills1lls1klEgkdO3ZMkhQMBvXGG2/oL3/5iy5fvqxf/OIXZdcCAID7l+9QmZqakud5ikQiqqurU3d3t5LJ5LI1yWRShw8fViAQUHt7uxYWFlQsFhUKhbR7925J0oMPPqhoNKpCoeB3JAAAUCN8h0qhUFBjY+PScTgcLouNStbMzs7qww8/1NNPP+13JAAAUCOCfp/AOVf2WCAQ+L/WfPLJJzp48KDefPNNPfTQQyu+TiKRUCKRkCTNz8/7GRkAAFQJ39+ohMNh5XK5peN8Pq+GhoaK19y+fVsHDx7UoUOH9MILL9z1dXp7e5VOp5VOp7Vt2za/YwMAgCrgO1Ta2tqUzWY1MzOjxcVFjYyMKB6PL1sTj8d1/vx5Oed0+fJlbd26VaFQSM45fe9731M0GtWPf/xjv6MAAIAa4/tXP8FgUIODg+rs7FSpVNKRI0fU0tKis2fPSpKOHj2qrq4uXbx4UZ7nacuWLTp37pwk6YMPPtAvf/lL7dy5U0888YQk6Wc/+5m6urr8jgUAAGpAwK10A4lxsVhM6XR6s8cAAADrYLXPdf4yLQAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMxal1CZmJhQc3OzPM/TwMBA2XnnnI4fPy7P89Ta2qrp6emKrwUAAPcv36FSKpXU19enVCqlTCaj4eFhZTKZZWtSqZSy2ayy2awSiYSOHTtW8bUAAOD+5TtUpqam5HmeIpGI6urq1N3drWQyuWxNMpnU4cOHFQgE1N7eroWFBRWLxYquBQAA9y/foVIoFNTY2Lh0HA6HVSgUKlpTybV3JBIJxWIxxWIxzc/P+x0bAABUAd+h4pwreywQCFS0ppJr7+jt7VU6nVY6nda2bdvucVoAAFBNgn6fIBwOK5fLLR3n83k1NDRUtGZxcXHNawEAwP3L9zcqbW1tymazmpmZ0eLiokZGRhSPx5eticfjOn/+vJxzunz5srZu3apQKFTRtQAA4P7l+xuVYDCowcFBdXZ2qlQq6ciRI2ppadHZs2clSUePHlVXV5cuXrwoz/O0ZcsWnTt3btVrAQAAJCngVrpRxLhYLKZ0Or3ZYwAAgHWw2uc6f5kWAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMMtXqNy8eVMdHR1qampSR0eHbt26teK6iYkJNTc3y/M8DQwMLD1+4sQJPf7442ptbdXzzz+vhYUFP+MAAIAa4ytUBgYGtHfvXmWzWe3du3dZhNxRKpXU19enVCqlTCaj4eFhZTIZSVJHR4c++ugj/elPf9LXv/51nTp1ys84AACgxvgKlWQyqZ6eHklST0+PLly4ULZmampKnucpEomorq5O3d3dSiaTkqT9+/crGAxKktrb25XP5/2MAwAAaoyvULl+/bpCoZAkKRQKaW5urmxNoVBQY2Pj0nE4HFahUChb9/bbb+vAgQN+xgEAADUmuNaCffv26dq1a2WP9/f3V/QCzrmyxwKBQNlzBYNBHTp06K7Pk0gklEgkJEnz8/MVvTYAAKhua4bKpUuX7npu+/btKhaLCoVCKhaLqq+vL1sTDoeVy+WWjvP5vBoaGpaOh4aGNDY2pvfee68sYL6ot7dXvb29kqRYLLbW2AAAoAb4+tVPPB7X0NCQpM+D47nnnitb09bWpmw2q5mZGS0uLmpkZETxeFzS5/8a6Oc//7lGR0e1ZcsWP6MAAIAa5CtUTp48qcnJSTU1NWlyclInT56UJF29elVdXV2SpGAwqMHBQXV2dioajerFF19US0uLJOmHP/yh/vGPf6ijo0NPPPGEjh496vPHAQAAtSTgVrqJxLhYLKZ0Or3ZYwAAgHWw2uc6f5kWAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZhEqAADALEIFAACYRagAAACzCBUAAGAWoQIAAMwiVAAAgFmECgAAMItQAQAAZvkKlZs3b6qjo0NNTU3q6OjQrVu3Vlw3MTGh5uZmeZ6ngYGBsvOvv/66AoGAbty44WccAABQY3yFysDAgPbu3atsNqu9e/euGCGlUkl9fX1KpVLKZDIaHh5WJpNZOp/L5TQ5OanHHnvMzygAAKAG+QqVZDKpnp4eSVJPT48uXLhQtmZqakqe5ykSiaiurk7d3d1KJpNL51999VWdPn1agUDAzygAAKAG+QqV69evKxQKSZJCoZDm5ubK1hQKBTU2Ni4dh8NhFQoFSdLo6KgeffRR7dq1a83XSiQSisViisVimp+f9zM2AACoEsG1Fuzbt0/Xrl0re7y/v7+iF3DOlT0WCAT06aefqr+/X++++25Fz9Pb26ve3l5JUiwWq+gaAABQ3dYMlUuXLt313Pbt21UsFhUKhVQsFlVfX1+2JhwOK5fLLR3n83k1NDToypUrmpmZWfo2JZ/Pa/fu3ZqamtIjjzxyLz8LAACoMb5+9ROPxzU0NCRJGhoa0nPPPVe2pq2tTdlsVjMzM1pcXNTIyIji8bh27typubk5zc7OanZ2VuFwWNPT00QKAABY4itUTp48qcnJSTU1NWlyclInT56UJF29elVdXV2SpGAwqMHBQXV2dioajerFF19US0uL/8kBAEDNC7iVbiIxLhaLKZ1Ob/YYAABgHaz2uc5fpgUAAGYRKgAAwCxCBQAAmEWoAAAAswgVAABgFqECAADMIlQAAIBZhAoAADCLUAEAAGYRKgAAwCxCBQAAmEWoAAAAswgVAABgFqECAADMIlQAAIBZhAoAADCLUAEAAGYRKgAAwCxCBQAAmEWoAAAAswgVAABgFqECAADMIlQAAIBZhAoAADAr4Jxzmz3E/+vhhx/Wjh07NnsMM+bn57Vt27bNHuO+wX5vHPZ647DXG4v9Xm52dlY3btxY8VxVhgqWi8ViSqfTmz3GfYP93jjs9cZhrzcW+105fvUDAADMIlQAAIBZhEoN6O3t3ewR7ivs98ZhrzcOe72x2O/KcY8KAAAwi29UAACAWYQKAAAwi1CpEjdv3lRHR4eamprU0dGhW7durbhuYmJCzc3N8jxPAwMDZedff/11BQKBu/57dfjf6xMnTujxxx9Xa2urnn/+eS0sLGzQ5NVjrfepc07Hjx+X53lqbW3V9PR0xdei3L3udy6X0zPPPKNoNKqWlhadOXNmo0evOn7e25JUKpX05JNP6tlnn92oke1zqAonTpxwp06dcs45d+rUKffaa6+Vrfn3v//tIpGIu3LlivvXv/7lWltb3Z///Oel83/729/c/v373WOPPebm5+c3bPZq43evf/3rX7vbt28755x77bXXVrz+frbW+9Q558bHx923vvUt99lnn7nf//737pvf/GbF12I5P/t99epV98c//tE559zHH3/smpqa2O9V+NnrO9544w33ne98x33729/eyNFN4xuVKpFMJtXT0yNJ6unp0YULF8rWTE1NyfM8RSIR1dXVqbu7W8lkcun8q6++qtOnTysQCGzU2FXJ717v379fwWBQktTe3q58Pr9hs1eDtd6n0uf/DQ4fPqxAIKD29nYtLCyoWCxWdC2W87PfoVBIu3fvliQ9+OCDikajKhQKm/FjVAU/ey1J+Xxe4+PjevnllzdjfLMIlSpx/fp1hUIhSVIoFNLc3FzZmkKhoMbGxqXjcDi89D+V0dFRPfroo9q1a9fGDFzF/O71F7399ts6cODAlzdsFapk7+62ptJ9x3/52e8vmp2d1Ycffqinn376yx24ivnd61deeUWnT5/WAw/w0fxFwc0eAP+1b98+Xbt2rezx/v7+iq53K/xL80AgoE8//VT9/f169913fc9YK76svf7f5woGgzp06NC9DVmjKtm7u62p5Fos52e/7/jkk0908OBBvfnmm3rooYfWf8ga4Wevx8bGVF9fr6eeekq//e1vv6wRqxKhYsilS5fuem779u1LX8UWi0XV19eXrQmHw8rlckvH+XxeDQ0NunLlimZmZpa+Tcnn89q9e7empqb0yCOPrP8PUgW+rL2+Y2hoSGNjY3rvvff4IP0fa+3damsWFxfXvBbL+dlvSbp9+7YOHjyoQ4cO6YUXXtiYoauUn73+1a9+pdHRUV28eFH//Oc/9fHHH+ull17SO++8s2Hzm7Vpd8fg//KTn/xk2Q2eJ06cKFtz+/Zt99WvftX99a9/XbqR66OPPipb95WvfIWbaVfhd69TqZSLRqNubm5uQ+euFpW8T8fGxpbdcNjW1lbxtVjOz35/9tln7rvf/a770Y9+tAmTVx8/e/1Fv/nNb7iZ9gsIlSpx48YNt2fPHud5ntuzZ4/7+9//7pxzrlAouAMHDiytGx8fd01NTS4Sibif/vSnKz4XobI6v3v9ta99zYXDYbdr1y63a9cu9/3vf3/DfwbrVtq7t956y7311lvOuc8/IH/wgx+4SCTivvGNb7g//OEPq16L1d3rfv/ud79zktzOnTuX3s/j4+Ob9nNUAz/v7TsIleX4E/oAAMAsbi0GAABmESoAAMAsQgUAAJhFqAAAALMIFQAAYBahAgAAzCJUAACAWf8B8RdVj5rLnjkAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 648x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "#print(names)\n",
    "fig = plt.figure(figsize=[9,4])\n",
    "fig.set_facecolor(\"w\")\n",
    "ymax = 15\n",
    "col_dict = { \"H\" : \"dodgerblue\",\n",
    "             \"S\" : \"darkorange\",\n",
    "             \"G\" : \"forestgreen\"}\n",
    "label_x = [0]\n",
    "label_txt = [\"GPS\"]\n",
    "stored = \"H\"\n",
    "stored_x = 0\n",
    "for lab_idx, label in enumerate(names):\n",
    "    #print(label.split(\".\")[0])\n",
    "    if label.split(\".\")[0] != stored:\n",
    "        if label not in [\"GPS-1\",\"GPS+1\"]:\n",
    "            box_col = col_dict[stored[0]]\n",
    "            col = col_dict[label[0]]\n",
    "            \n",
    "            plt.vlines(lab_idx, 0,ymax, color=col, alpha=0.4)\n",
    "            plt.fill_between([stored_x, lab_idx], 0, ymax, color=box_col, alpha=0.2)\n",
    "            \n",
    "            stored_x = lab_idx\n",
    "            stored = label.split(\".\")[0]\n",
    "            \n",
    "            label_x.append(lab_idx)\n",
    "            label_txt.append(label.split(\".\")[0])\n",
    "\n",
    "        plt.xticks(ticks = label_x, labels = label_txt, rotation=45)\n",
    "        #plt.yticks(ticks = [0,1,2], labels = [\"0\", \"100%\", \"200%\"])\n",
    "        #plt.ylabel(\"Mutation Number over Position Occurence\")\n",
    "        plt.xlabel(\"Element name\")\n",
    "#plt.fill_between([stored_x, lab_idx], 0, ymax, color=\"darkorange\", alpha=0.2)\n",
    "#for x in range(len(nums)):\n",
    "#    alfa = 0.2+(0.8*occus[x]/31)\n",
    "#    plt.scatter(x, nums[x]/occus[x], color='red', alpha=alfa, s=11)\n",
    "#plt.plot(range(len(nums)),nums, linewidth=1.5, color='black')\n",
    "#plt.bar(range(len(nums)),nums, color='black')\n",
    "plt.scatter(range(len(nums)),np.divide(nums, occus), color='red', s=6)\n",
    "#plt.savefig(\"corr.human_indexed_.png\", bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#b-factor : line[60:66]\n",
    "d1_gain = human_collection.collection[26]\n",
    "print(d1_gain.name, d1_gain.start, d1_gain.end)\n",
    "ref_pdb = \"d1_gain.pdb\"\n",
    "with open(ref_pdb) as inpdb:\n",
    "    pdb_data = inpdb.readlines()\n",
    "#names, nums, occus\n",
    "print(d1_gain.idx_dir)\n",
    "new_data = []\n",
    "for line in pdb_data:\n",
    "    if not line.startswith(\"ATOM\"):\n",
    "        new_data.append(line)\n",
    "        continue\n",
    "    resid = int(line[22:26])\n",
    "    try:\n",
    "        position_label = d1_gain.rev_idx_dir[resid-1] # i.e. \"H2.54\", account for zero vs. one-indexed!\n",
    "        try: \n",
    "            b = float(len(generalized_mutations[position_label]))\n",
    "            #print(position_label, len(generalized_mutations[position_label]))\n",
    "        except KeyError:\n",
    "            b = 0.0\n",
    "    except KeyError: \n",
    "        b = 0.0\n",
    "    new_data.append(line[:61]+\"%6.2f\"%(b)+line[68:])\n",
    "\n",
    "with open(\"d1_gain_mut.pdb\",\"w\") as outpdb:\n",
    "    for line in new_data:\n",
    "        outpdb.write(line)\n",
    "print(\"Done.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#b-factor : line[60:66]\n",
    "for i, gain in enumerate(human_collection.collection):\n",
    "    if \"AGRL1\" in gain.name: \n",
    "        print(i)\n",
    "        break\n",
    "l1_gain = human_collection.collection[i]\n",
    "print(l1_gain.name, l1_gain.start, l1_gain.end)\n",
    "ref_pdb = \"../repo_data/l1_gain.pdb\"\n",
    "with open(ref_pdb) as inpdb:\n",
    "    pdb_data = inpdb.readlines()\n",
    "#names, nums, occus\n",
    "print(l1_gain.idx_dir)\n",
    "new_data = []\n",
    "for line in pdb_data:\n",
    "    if not line.startswith(\"ATOM\"):\n",
    "        new_data.append(line)\n",
    "        continue\n",
    "    resid = int(line[22:26])\n",
    "    try:\n",
    "        position_label = l1_gain.rev_idx_dir[resid-1] # i.e. \"H2.54\" account for 0-indexed residues in the dictionary!\n",
    "        try: \n",
    "            b = float(len(generalized_mutations[position_label]))\n",
    "            #print(position_label, len(generalized_mutations[position_label]))\n",
    "        except KeyError:\n",
    "            b = 0.0\n",
    "    except KeyError: \n",
    "        b = 0.0\n",
    "    new_data.append(line[:61]+\"%6.2f\"%(b)+line[68:])\n",
    "\n",
    "with open(\"l1_gain_mut.pdb\",\"w\") as outpdb:\n",
    "    for line in new_data:\n",
    "        outpdb.write(line)\n",
    "print(\"Done.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H1.50 , L\n",
      "H1.54 , II\n",
      "H1.57 , A\n",
      "H1.61 , A\n",
      "H2.56 , KKK\n",
      "H2.57 , I\n",
      "H2.60 , D\n",
      "H2.61 , Q\n",
      "H3.36 , RR\n",
      "H3.43 , AAAPPA\n",
      "H3.44 , SSEAS\n",
      "H3.51 , RRRAVCRR\n",
      "H3.53 , ALLSQ\n",
      "H3.56 , RRRR\n",
      "H4.38 , ASS\n",
      "H4.41 , GGGEGMGGA\n",
      "H4.51 , ARREDMDAKKEQD\n",
      "H5.37 , K\n",
      "H5.38 , RRRR\n",
      "H5.42 , AS\n",
      "H5.44 , GEEKNN\n",
      "H5.48 , AAESE\n",
      "H5.50 , VMMSI\n",
      "H5.59 , EL\n",
      "H6.42 , E\n",
      "H6.54 , WWWWWWW\n",
      "H6.56 , EDVVF\n",
      "H7.40 , R\n",
      "H7.51 , A\n",
      "H8.46 , GAAASAATL\n",
      "H8.58 , AASSTGMD\n",
      "H8.60 , VIVIAAA\n",
      "S1.48 , AAVMAA\n",
      "S2.47 , S\n",
      "S2.51 , VAAQAAS\n",
      "S2.53 , EAEEE\n",
      "S2.58 , PPSD\n",
      "S3.53 , IIAA\n",
      "S3.55 , ST\n",
      "S4.56 , R\n",
      "S5.48 , VILVS\n",
      "S5.52 , GTCHG\n",
      "S5.55 , ILSYSF\n",
      "S6.50 , SSV\n",
      "S7.45 , VV\n",
      "S7.52 , SVASASS\n",
      "S7.57 , SA\n",
      "S9.53 , WWWWWWWWWW\n",
      "S10.50 , WWWWWWW\n",
      "S11.54 , SDEQVYYQR\n",
      "S11.55 , RESSEGG\n",
      "S12.47 , HRYGEEQ\n",
      "S12.50 , CCCC\n",
      "S12.52 , CCCC\n",
      "S13.48 , AAGAA\n",
      "S13.50 , LYLLI\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#print(generalized_mutations[label])\n",
    "def position_vars(gen_vars, label, aa_key=None, text_out=None, return_aa=False):\n",
    "    # Prints or writes to File the mutations at a specific label position.\n",
    "    def tee(file_obj=None, text_string=\"\"):\n",
    "        # Selector function similar to UNIX \"tee\" -> where it prints and - if existing - writes to File.\n",
    "        if file_obj is not None:\n",
    "            file_obj.write(text_string)\n",
    "        print(text_string, end='')\n",
    "\n",
    "    if text_out is not None: # Initialize TEXT output file if specified. Otherwise construct None\n",
    "        text = open(text_out, 'w')\n",
    "    else: text = None\n",
    "\n",
    "    #tee(text, f\"CANCER GENOME ATLAS MUTATIONS AT POSITION {label.upper()} \\nTOTAL: {len(gen_vars[label])} MUTATIONS\\n\\n\")\n",
    "\n",
    "    rstring=''\n",
    "    for var in gen_vars[label]:\n",
    "        for key in var.keys():\n",
    "            if return_aa and key == 'aa_change':\n",
    "                    rlet = var[key][0]\n",
    "                    rstring = rstring+rlet\n",
    "            #tee(text, f\"{key.ljust(30)} {var[key]}\\n\")\n",
    "        if aa_key is not None:\n",
    "            resid = int(var[aa_key]) # X is the residue ID\n",
    "        #tee(text, \"\\n\"+\"_\"*30+\"\\n\")\n",
    "    print(label, \",\", rstring)\n",
    "\n",
    "enriched_positions = ['H1.50','H1.54','H1.57','H1.61','H2.56','H2.57','H2.60','H2.61','H3.36','H3.43',\n",
    "'H3.44','H3.51','H3.53','H3.56','H4.38','H4.41','H4.51','H5.37','H5.38','H5.42','H5.44',\n",
    "'H5.48','H5.50','H5.59','H6.42','H6.54','H6.56','H7.40','H7.51','H8.46','H8.58','H8.60',\n",
    "'S1.48','S2.47','S2.51','S2.53','S2.58','S3.53','S3.55','S4.56','S5.48','S5.52','S5.55',\n",
    "'S6.50','S7.45','S7.52','S7.57','S9.53','S10.50','S11.54','S11.55','S12.47','S12.50','S12.52','S13.48','S13.50']\n",
    "\n",
    "for label in enriched_positions:\n",
    "    position_vars(generalized_mutations, label, return_aa=True)#text_out=f\"../mut/{label}.out\", return_aa=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_criteria(*pairs, sift=None, poly=None, mutation_dict):\n",
    "    # Defines a function where key - value pairs will be used to specifically filter criteria to be applied for subselection of mutations.\n",
    "    #       i.e. (\"consequence\", \"missense\"), (\"sift_impact\",\"deleterious\")\n",
    "    # Scores \"sift\" and \"polyphen\" define minimum and maximum values with an operator to determine lower than or higher than $val cutoffs\n",
    "    #       i.e. (\"sift_score\", \">\", 0.22) as tuples\n",
    "    print(\"DEBUG:\", pairs)\n",
    "    def eval_entry(pairs, entry):\n",
    "        if pairs is None:\n",
    "            return True\n",
    "\n",
    "        for tup in pairs:\n",
    "            if entry[tup[0]] != tup[1]:\n",
    "                return False\n",
    "        return True\n",
    "\n",
    "    # Set a custom_func flag indicating if there are custom dependent lambda functions for scores specified\n",
    "    custom_func = True\n",
    "    if sift is None and poly is None:\n",
    "        custom_func = False\n",
    "    if sift is None:\n",
    "        sift = lambda n: True\n",
    "    if poly is None:\n",
    "        poly = lambda n: True\n",
    "\n",
    "        \n",
    "    print(\"DEBUG\", custom_func)\n",
    "\n",
    "    valid_ct = 0\n",
    "    out_dict = {}\n",
    "    # We want to filter the generalized dictionary by key\n",
    "    for ki in mutation_dict.keys():\n",
    "        mutation_list = mutation_dict[ki]\n",
    "        filtered_mutation_list = []\n",
    "        for mutation in mutation_list:\n",
    "            # Check the scores first. If there are no scores, but custom functions specified, thats ok. Otherwise, do not consider this entry\n",
    "            try: \n",
    "                s = float(mutation[\"sift_score\"])\n",
    "                p = float(mutation[\"polyphen_score\"])\n",
    "            except:\n",
    "                #print(\"Scores not in Entry...\", end ='')\n",
    "                if custom_func == True:\n",
    "                    continue\n",
    "                print(\"But no score functions specified\")\n",
    "            if custom_func and not sift(float(mutation[\"sift_score\"])):\n",
    "                continue\n",
    "            if custom_func and not poly(float(mutation[\"polyphen_score\"])):\n",
    "                continue\n",
    "            # Here, check the *pairs argument if that matches the specifications.\n",
    "            if not eval_entry(pairs, mutation):\n",
    "                continue\n",
    "            \n",
    "            valid_ct += 1\n",
    "            filtered_mutation_list.append(mutation)\n",
    "        out_dict[ki] = filtered_mutation_list\n",
    "    print(f\"Found {valid_ct} entries matching criteria.\")\n",
    "    return out_dict\n",
    "            \n",
    "# Example Lambda Functions for score evaluation\n",
    "poly = lambda n : n > 0.446 # > 0.446 for possibly damaging, > 0.908 for Probably Damaging\n",
    "sift = lambda n : n < 0.05 # For deleterious impacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: ()\n",
      "DEBUG False\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "But no score functions specified\n",
      "Found 853 entries matching criteria.\n"
     ]
    }
   ],
   "source": [
    "#missense_dict = query_criteria((\"consequence\", \"missense\"), poly=lambda n : n > 0.446, mutation_dict=generalized_mutations)\n",
    "missense_dict2 = query_criteria(mutation_dict=generalized_mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gain in human_collection.collection:\n",
    "    if \"AGRL3\" in gain.name:\n",
    "        print(\"DEBUG:\", gain.name, gain.idx_dir, \"\".join(gain.sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "retrieve_csv_vars() got an unexpected keyword argument 'return_keys'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-f1504507d29d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcsvs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mglob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/home/hildilab/projects/agpcr_nom/snp_mane/*csv\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m parse_dict, data = retrieve_csv_vars(csvs[0], \n\u001b[0m\u001b[1;32m      4\u001b[0m                                     \u001b[0mreturn_keys\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Protein Consequence'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Transcript'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Transcript Consequence'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'VEP Annotation'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m                                     \u001b[0mfilter_str\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'missense_variant'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: retrieve_csv_vars() got an unexpected keyword argument 'return_keys'"
     ]
    }
   ],
   "source": [
    "csvs = glob.glob(\"/home/hildilab/projects/agpcr_nom/snp_mane/*csv\")\n",
    "\n",
    "parse_dict, data = retrieve_csv_vars(csvs[0], \n",
    "                                    return_keys=['Protein Consequence', 'Transcript', 'Transcript Consequence', 'VEP Annotation'], \n",
    "                                    filter_str = 'missense_variant'\n",
    "                                    )\n",
    "print(data)\n",
    "# Filter for VEP annotation to only get the missense mutation\n",
    "oneletter = {'Ala':'A','Arg':'R','Asn':'N','Asp':'D','Cys':'C','Gln':'Q','Glu':'E','Gly':'G','His':'H','Ile':'I','Leu':'L','Lys':'K','Met':'M','Phe':'F','Pro':'P','Pyl':'O','Ser':'S','Thr':'T','Trp':'W','Tyr':'Y','Val':'V'}\n",
    "idx = parse_dict['Protein Consequence']\n",
    "for mutation in data[idx,:]:\n",
    "    res = oneletter[re.match(r'[A-z]{3}')[0]] # A, D, ...\n",
    "    resnum = re.match(r'[\\d]+', mutation) # 1364\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lets map the vars/mutations to the loops as wäll\n",
    "\n",
    "loop_lengths = {}\n",
    "loop_seqs = {}\n",
    "loop_seq = {}\n",
    "\n",
    "loop_info = {}\n",
    "#[loop_info[loop] = {} for loop in loop_seqs.keys()] # into each of these keys, any entry is composed of \"name\":$name, \"sequence\":$seq\n",
    "\n",
    "for idx, gain in enumerate(human_collection.collection):\n",
    "    curr_name = gain.name\n",
    "    idx_dir, cent, named_dir, _ = sse_func.create_indexing(gain, anchors, anchor_occupation, anchor_dict, outdir=None, offset=fasta_offsets[idx], silent=True, split_mode='double')\n",
    "    #print(f'{named_dir = }')\n",
    "    i_loc, i_dir = get_loop_stats(idx_dir, gain.sequence)\n",
    "    for k, seq in i_dir.items():\n",
    "        if k not in loop_info.keys():\n",
    "            loop_info[k] = []\n",
    "        loop_info[k].append({'name':f'{gain.name}_{i_loc[k][0]+gain.start}-{i_loc[k][1]+gain.start}', 'sequence':''.join(seq)})\n",
    "\n",
    "#for loop in loop_info.keys():\n",
    "#    loop2fasta(f\"../loops_human/{loop}.fa\", loop_info[loop])\n",
    "\n",
    "# mafft\n",
    "mafft_aln = glob.glob('../loops_human/aln/*fa')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], [], [0, 18]]\n",
      "[(array([ 0, 13]), ['H2.38', 'H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51'], (array([ 0, 13]), 12, 'H2'))]\n",
      "[(array([ 90, 101]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([ 90, 101]), 92, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  21 \n",
      "INVALID POSITIONS     :  349 \n",
      "TOTAL POSITIONS       :  370\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [1], [], [], [0, 18]]\n",
      "[(array([ 0, 13]), ['H2.38', 'H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51'], (array([ 0, 13]), 12, 'H2'))]\n",
      "[(array([ 90, 101]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([ 90, 101]), 92, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  47 \n",
      "INVALID POSITIONS     :  704 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "[[1], [], [], [], [], [], []]\n",
      "DEBUG gain_classes.disambiguate_anchors : no break found between anchors, will just overwrite.\n",
      "[(array([ 5, 18]), ['H2.38', 'H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51'], (array([ 5, 18]), 17, 'H2'))]\n",
      "[(array([ 95, 106]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([ 95, 106]), 97, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], [], [7]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [333, 339], seg_new = [341, 349]\n",
      "[([333, 339], ['S12.47', 'S12.48', 'S12.49', 'S12.50', 'S12.51', 'S12.52', 'S12.53'], ([333, 339], 336, 'S12')), ([341, 349], ['S13.44', 'S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52'], ([341, 349], 347, 'S13'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [341, 349],\tname_2 = ['S13.44', 'S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52'],\tcast_2 =  ([341, 349], 347, 'S13')\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  65 \n",
      "INVALID POSITIONS     :  1088 \n",
      "TOTAL POSITIONS       :  402\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], [], []]\n",
      "[(array([ 1, 15]), ['H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51', 'H2.52', 'H2.53'], (array([ 1, 15]), 12, 'H2'))]\n",
      "[[], [], [], [], [], [], [2], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  74 \n",
      "INVALID POSITIONS     :  1172 \n",
      "TOTAL POSITIONS       :  93\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 89, 101]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59', 'H6.60'], (array([ 89, 101]), 91, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  96 \n",
      "INVALID POSITIONS     :  1390 \n",
      "TOTAL POSITIONS       :  240\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([19, 39]), ['H3.36', 'H3.37', 'H3.38', 'H3.39', 'H3.40', 'H3.41', 'H3.42', 'H3.43', 'H3.44', 'H3.45', 'H3.46', 'H3.47', 'H3.48', 'H3.49', 'H3.50', 'H3.51', 'H3.52', 'H3.53', 'H3.54', 'H3.55', 'H3.56'], (array([19, 39]), 33, 'H3'))]\n",
      "[(array([86, 97]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([86, 97]), 88, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  116 \n",
      "INVALID POSITIONS     :  1571 \n",
      "TOTAL POSITIONS       :  201\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([19, 39]), ['H3.36', 'H3.37', 'H3.38', 'H3.39', 'H3.40', 'H3.41', 'H3.42', 'H3.43', 'H3.44', 'H3.45', 'H3.46', 'H3.47', 'H3.48', 'H3.49', 'H3.50', 'H3.51', 'H3.52', 'H3.53', 'H3.54', 'H3.55', 'H3.56'], (array([19, 39]), 33, 'H3'))]\n",
      "[(array([86, 97]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([86, 97]), 88, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  166 \n",
      "INVALID POSITIONS     :  2038 \n",
      "TOTAL POSITIONS       :  517\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 6, 17]), ['H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51', 'H2.52'], (array([ 6, 17]), 15, 'H2'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [8]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [301, 308], seg_new = [310, 319]\n",
      "[([301, 308], ['S12.47', 'S12.48', 'S12.49', 'S12.50', 'S12.51', 'S12.52', 'S12.53', 'S12.54'], ([301, 308], 304, 'S12')), ([310, 319], ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'], ([310, 319], 315, 'S13'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [310, 319],\tname_2 = ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'],\tcast_2 =  ([310, 319], 315, 'S13')\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  185 \n",
      "INVALID POSITIONS     :  2201 \n",
      "TOTAL POSITIONS       :  182\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 6, 17]), ['H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51', 'H2.52'], (array([ 6, 17]), 15, 'H2'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [8]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [298, 305], seg_new = [307, 316]\n",
      "[([298, 305], ['S12.47', 'S12.48', 'S12.49', 'S12.50', 'S12.51', 'S12.52', 'S12.53', 'S12.54'], ([298, 305], 301, 'S12')), ([307, 316], ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'], ([307, 316], 312, 'S13'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [307, 316],\tname_2 = ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'],\tcast_2 =  ([307, 316], 312, 'S13')\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  213 \n",
      "INVALID POSITIONS     :  2546 \n",
      "TOTAL POSITIONS       :  373\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [2]]\n",
      "DEBUG gain_classes.disambiguate_anchors : no break found between anchors, will just overwrite.\n",
      "[(array([63, 82]), ['H8.40', 'H8.41', 'H8.42', 'H8.43', 'H8.44', 'H8.45', 'H8.46', 'H8.47', 'H8.48', 'H8.49', 'H8.50', 'H8.51', 'H8.52', 'H8.53', 'H8.54', 'H8.55', 'H8.56', 'H8.57', 'H8.58', 'H8.59'], (array([63, 82]), 73, 'H8'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  233 \n",
      "INVALID POSITIONS     :  2701 \n",
      "TOTAL POSITIONS       :  175\n",
      "This should be a list of JSON Files.\n",
      "[[], [17], [], [], [], []]\n",
      "[(array([ 0, 12]), ['H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51'], (array([ 0, 12]), 11, 'H2'))]\n",
      "[(array([81, 92]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59'], (array([81, 92]), 83, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  254 \n",
      "INVALID POSITIONS     :  2922 \n",
      "TOTAL POSITIONS       :  242\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  285 \n",
      "INVALID POSITIONS     :  3632 \n",
      "TOTAL POSITIONS       :  741\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  299 \n",
      "INVALID POSITIONS     :  4737 \n",
      "TOTAL POSITIONS       :  1119\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 6, 17]), ['H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51', 'H2.52'], (array([ 6, 17]), 15, 'H2'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [8]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [286, 293], seg_new = [295, 304]\n",
      "[([286, 293], ['S12.47', 'S12.48', 'S12.49', 'S12.50', 'S12.51', 'S12.52', 'S12.53', 'S12.54'], ([286, 293], 289, 'S12')), ([295, 304], ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'], ([295, 304], 300, 'S13'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [295, 304],\tname_2 = ['S13.45', 'S13.46', 'S13.47', 'S13.48', 'S13.49', 'S13.50', 'S13.51', 'S13.52', 'S13.53', 'S13.54'],\tcast_2 =  ([295, 304], 300, 'S13')\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  338 \n",
      "INVALID POSITIONS     :  5079 \n",
      "TOTAL POSITIONS       :  381\n",
      "This should be a list of JSON Files.\n",
      "[[], []]\n",
      "[(array([ 4, 35]), ['H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51', 'H2.52', 'H2.53', 'H2.54', 'H2.55', 'H2.56', 'H2.57', 'H2.58', 'H2.59', 'H2.60', 'H2.61', 'H2.62', 'H2.63', 'H2.64', 'H2.65', 'H2.66', 'H2.67', 'H2.68', 'H2.69', 'H2.70', 'H2.71', 'H2.72', 'H2.73', 'H2.74', 'H2.75'], (array([ 4, 35]), 10, 'H2'))]\n",
      "[(array([ 4, 35]), ['H3.35', 'H3.36', 'H3.37', 'H3.38', 'H3.39', 'H3.40', 'H3.41', 'H3.42', 'H3.43', 'H3.44', 'H3.45', 'H3.46', 'H3.47', 'H3.48', 'H3.49', 'H3.50', 'H3.51', 'H3.52', 'H3.53', 'H3.54', 'H3.55', 'H3.56', 'H3.57', 'H3.58', 'H3.59', 'H3.60', 'H3.61', 'H3.62', 'H3.63', 'H3.64', 'H3.65', 'H3.66'], (array([ 4, 35]), 19, 'H3'))]\n",
      "[(array([ 4, 35]), ['H4.24', 'H4.25', 'H4.26', 'H4.27', 'H4.28', 'H4.29', 'H4.30', 'H4.31', 'H4.32', 'H4.33', 'H4.34', 'H4.35', 'H4.36', 'H4.37', 'H4.38', 'H4.39', 'H4.40', 'H4.41', 'H4.42', 'H4.43', 'H4.44', 'H4.45', 'H4.46', 'H4.47', 'H4.48', 'H4.49', 'H4.50', 'H4.51', 'H4.52', 'H4.53', 'H4.54', 'H4.55'], (array([ 4, 35]), 30, 'H4'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  354 \n",
      "INVALID POSITIONS     :  5139 \n",
      "TOTAL POSITIONS       :  76\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], [], []]\n",
      "[(array([ 1, 12]), ['H2.39', 'H2.40', 'H2.41', 'H2.42', 'H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50'], (array([ 1, 12]), 12, 'H2'))]\n",
      "[(array([83, 95]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58', 'H6.59', 'H6.60'], (array([83, 95]), 85, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  383 \n",
      "INVALID POSITIONS     :  5264 \n",
      "TOTAL POSITIONS       :  154\n",
      "This should be a list of JSON Files.\n",
      "[[], [10], [8]]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  401 \n",
      "INVALID POSITIONS     :  5331 \n",
      "TOTAL POSITIONS       :  85\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [14], [], [], []]\n",
      "[(array([ 2, 10]), ['H2.43', 'H2.44', 'H2.45', 'H2.46', 'H2.47', 'H2.48', 'H2.49', 'H2.50', 'H2.51'], (array([ 2, 10]), 9, 'H2'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  425 \n",
      "INVALID POSITIONS     :  5447 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [12], [], [], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  479 \n",
      "INVALID POSITIONS     :  5608 \n",
      "TOTAL POSITIONS       :  215\n",
      "This should be a list of JSON Files.\n",
      "WARNING: Multiple matching JSON Files detected: ['../repo_data/gain_json/d2_gain.json', '../repo_data/gain_json/uncurated.d2_gain.json']\n",
      "[[7], [], [], [], [], [], [8], [9]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [115, 122], seg_new = [124, 126]\n",
      "[([115, 122], ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55'], ([115, 122], 117, 'H6')), ([124, 126], ['H7.50', 'H7.51', 'H7.52'], ([124, 126], 124, 'H7'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [124, 126],\tname_2 = ['H7.50', 'H7.51', 'H7.52'],\tcast_2 =  ([124, 126], 124, 'H7')\n",
      "[[], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  479 \n",
      "INVALID POSITIONS     :  5612 \n",
      "TOTAL POSITIONS       :  4\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 92, 100]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56'], (array([ 92, 100]), 94, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  503 \n",
      "INVALID POSITIONS     :  5691 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [], []]\n",
      "[(array([ 99, 107]), ['H6.48', 'H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56'], (array([ 99, 107]), 101, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  523 \n",
      "INVALID POSITIONS     :  5811 \n",
      "TOTAL POSITIONS       :  140\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  545 \n",
      "INVALID POSITIONS     :  5966 \n",
      "TOTAL POSITIONS       :  177\n",
      "This should be a list of JSON Files.\n",
      "[[], [12], [], [], [], []]\n",
      "[(array([82, 91]), ['H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58'], (array([82, 91]), 83, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  569 \n",
      "INVALID POSITIONS     :  6133 \n",
      "TOTAL POSITIONS       :  191\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], []]\n",
      "[(array([28, 44]), ['H4.46', 'H4.47', 'H4.48', 'H4.49', 'H4.50', 'H4.51', 'H4.52', 'H4.53', 'H4.54', 'H4.55', 'H4.56', 'H4.57', 'H4.58', 'H4.59', 'H4.60', 'H4.61', 'H4.62'], (array([28, 44]), 32, 'H4'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  587 \n",
      "INVALID POSITIONS     :  6218 \n",
      "TOTAL POSITIONS       :  103\n",
      "This should be a list of JSON Files.\n",
      "[[15], []]\n",
      "DEBUG gain_classes.disambiguate_anchors : no break found between anchors, will just overwrite.\n",
      "[(array([ 5, 24]), ['H4.43', 'H4.44', 'H4.45', 'H4.46', 'H4.47', 'H4.48', 'H4.49', 'H4.50', 'H4.51', 'H4.52', 'H4.53', 'H4.54', 'H4.55', 'H4.56', 'H4.57', 'H4.58', 'H4.59', 'H4.60', 'H4.61', 'H4.62'], (array([ 5, 24]), 12, 'H4'))]\n",
      "DEBUG gain_classes.disambiguate_anchors : no break found between anchors, will just overwrite.\n",
      "[(array([ 5, 24]), ['H3.50', 'H3.51', 'H3.52', 'H3.53', 'H3.54', 'H3.55', 'H3.56', 'H3.57', 'H3.58', 'H3.59', 'H3.60', 'H3.61', 'H3.62', 'H3.63', 'H3.64', 'H3.65', 'H3.66', 'H3.67', 'H3.68', 'H3.69'], (array([ 5, 24]), 5, 'H3'))]\n",
      "[(array([34, 46]), ['H8.46', 'H8.47', 'H8.48', 'H8.49', 'H8.50', 'H8.51', 'H8.52', 'H8.53', 'H8.54', 'H8.55', 'H8.56', 'H8.57', 'H8.58'], (array([34, 46]), 38, 'H8'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  600 \n",
      "INVALID POSITIONS     :  6279 \n",
      "TOTAL POSITIONS       :  74\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], [], [5]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [101, 105], seg_new = [107, 127]\n",
      "[([101, 105], ['H7.47', 'H7.48', 'H7.49', 'H7.50', 'H7.51'], ([101, 105], 104, 'H7')), ([107, 127], ['H8.43', 'H8.44', 'H8.45', 'H8.46', 'H8.47', 'H8.48', 'H8.49', 'H8.50', 'H8.51', 'H8.52', 'H8.53', 'H8.54', 'H8.55', 'H8.56', 'H8.57', 'H8.58', 'H8.59', 'H8.60', 'H8.61', 'H8.62', 'H8.63'], ([107, 127], 114, 'H8'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [107, 127],\tname_2 = ['H8.43', 'H8.44', 'H8.45', 'H8.46', 'H8.47', 'H8.48', 'H8.49', 'H8.50', 'H8.51', 'H8.52', 'H8.53', 'H8.54', 'H8.55', 'H8.56', 'H8.57', 'H8.58', 'H8.59', 'H8.60', 'H8.61', 'H8.62', 'H8.63'],\tcast_2 =  ([107, 127], 114, 'H8')\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  642 \n",
      "INVALID POSITIONS     :  6418 \n",
      "TOTAL POSITIONS       :  181\n",
      "This should be a list of JSON Files.\n",
      "[[], [2], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  665 \n",
      "INVALID POSITIONS     :  6505 \n",
      "TOTAL POSITIONS       :  110\n",
      "This should be a list of JSON Files.\n",
      "[[], [12], [15, 19], [], [9]]\n",
      "[NOTE] disambiguate_anchors: Split the segment into: seg_stored = [41, 55], seg_new = [61, 76]\n",
      "[([41, 55], ['H4.40', 'H4.41', 'H4.42', 'H4.43', 'H4.44', 'H4.45', 'H4.46', 'H4.47', 'H4.48', 'H4.49', 'H4.50', 'H4.51', 'H4.52', 'H4.53', 'H4.54'], ([41, 55], 51, 'H4')), ([61, 76], ['H5.43', 'H5.44', 'H5.45', 'H5.46', 'H5.47', 'H5.48', 'H5.49', 'H5.50', 'H5.51', 'H5.52', 'H5.53', 'H5.54', 'H5.55', 'H5.56', 'H5.57', 'H5.58'], ([61, 76], 68, 'H5'))]\n",
      "[DEBUG] GainDomain.create_indexing : Found a split list:\n",
      "sse_adj_2  = [61, 76],\tname_2 = ['H5.43', 'H5.44', 'H5.45', 'H5.46', 'H5.47', 'H5.48', 'H5.49', 'H5.50', 'H5.51', 'H5.52', 'H5.53', 'H5.54', 'H5.55', 'H5.56', 'H5.57', 'H5.58'],\tcast_2 =  ([61, 76], 68, 'H5')\n",
      "[(array([80, 89]), ['H6.49', 'H6.50', 'H6.51', 'H6.52', 'H6.53', 'H6.54', 'H6.55', 'H6.56', 'H6.57', 'H6.58'], (array([80, 89]), 81, 'H6'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  687 \n",
      "INVALID POSITIONS     :  6629 \n",
      "TOTAL POSITIONS       :  146\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [], []]\n",
      "[(array([24, 43]), ['H4.43', 'H4.44', 'H4.45', 'H4.46', 'H4.47', 'H4.48', 'H4.49', 'H4.50', 'H4.51', 'H4.52', 'H4.53', 'H4.54', 'H4.55', 'H4.56', 'H4.57', 'H4.58', 'H4.59', 'H4.60', 'H4.61', 'H4.62'], (array([24, 43]), 31, 'H4'))]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  710 \n",
      "INVALID POSITIONS     :  6747 \n",
      "TOTAL POSITIONS       :  141\n",
      "This should be a list of JSON Files.\n",
      "[[], [], [6, 12], []]\n",
      "[[], [], [], [], [], [], [], [], [], [], [], [], []]\n",
      "DEBUG\n",
      "VALID LOOP POSITIONS:  745 \n",
      "INVALID POSITIONS     :  6898 \n",
      "TOTAL POSITIONS       :  186\n",
      "[NOTE]: NO MISMATCHES HAVE BEEN FOUND.\n",
      "TOTAL VARS WITHIN GAIN: 11689 \n",
      "TOTAL VARS OUTSIDE GAIN: 133826 \n",
      "TOTAL VARS: 145515\n",
      "S4-S5 86\n",
      "S8-S9 55\n",
      "S2-S4 35\n",
      "S11-S12 17\n",
      "S1-S2 22\n",
      "H2-H3 17\n",
      "S5-S7 85\n",
      "S10-S11 16\n",
      "H4-H5 33\n",
      "S9-S10 44\n",
      "H3-H4 32\n",
      "S7-S8 45\n",
      "H8-S1 24\n",
      "S12-S13 19\n",
      "H6-H8 5\n",
      "S2-S3 48\n",
      "H7-H8 5\n",
      "S3-S4 39\n",
      "H1-H3 9\n",
      "H8-S2 34\n",
      "S6-S7 11\n",
      "S5-S6 33\n",
      "S7-S9 13\n",
      "H5-H6 6\n",
      "S4-S7 17\n",
      "H5-H8 11\n",
      "S4-S6 6\n",
      "H8-S3 13\n",
      "H4-H8 5\n",
      "S3-S7 14\n",
      "H5-H7 2\n",
      "S5-S8 14\n",
      "H1-H4 13\n",
      "H4-H6 4\n"
     ]
    }
   ],
   "source": [
    "# Construct a labeled dict with each connector containing the corresponding mutations for this - independent of the receptor\n",
    "# ie \"H8-S1\": [{mut1}, {mut2}]\n",
    "# we rather not use the Alignment to sort mutations, since the loops themselves have too much variability as shown in the ALN. Subfamily-level.\n",
    "\n",
    "loop_muts, loop_counts = compose_loop_vars(human_collection, jsons, resid_key='x', aa_key='aa_change', fasta_offsets=fasta_offsets)\n",
    "\n",
    "for loop in loop_counts.keys():\n",
    "    print(loop, loop_counts[loop])\n",
    "    #for k, seq in i_dir.items():\n",
    "    #    if k not in loop_info.keys():\n",
    "    #        loop_info[k] = []\n",
    "    #    loop_info[k].append({'name':f'{gain.name}_{i_loc[k][0]+gain.start}-{i_loc[k][1]+gain.start}', 'sequence':''.join(seq)})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.52 \n",
      "TOTAL: 6 MUTATIONS\n",
      "\n",
      "id                             739c9590-5175-56c3-b6a7-7b776c41f9b9\n",
      "y                              1\n",
      "genomic_dna_change             chr3:g.48646107C>G\n",
      "x                              2482\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.56\n",
      "polyphen_score                 0.261\n",
      "aa_change                      Q2482H\n",
      "receptor                       Q9NYQ7_O75092-CELR3_HUMAN-CadherinEGFLAGseven-passG-typereceptor3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             d2b87d5b-5cda-5580-8b6b-9eeae38b77f5\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.69076021G>C\n",
      "x                              821\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.11\n",
      "polyphen_score                 0.224\n",
      "aa_change                      L821F\n",
      "receptor                       O60242_B7Z1K0_O60297_Q2NKN6_Q5VY37_Q9BX54-AGRB3_HUMAN-AGRB3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             317203a4-9bef-5e79-af23-3b01143816af\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.81968058C>G\n",
      "x                              790\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.986\n",
      "aa_change                      F790L\n",
      "receptor                       O95490_A5XEI2_B1ALT8_B1ALT9_B1ALU0_B1ALU2_B1ALU4_B1ALU5_B1ALU6-AGRL2_HUMAN-AGRL2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             a3232752-a313-5739-a940-9a35d75c60fd\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.47681745G>A\n",
      "x                              396\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.5\n",
      "polyphen_score                 0.961\n",
      "aa_change                      G396S\n",
      "receptor                       Q8IZF7_Q2PNZ1_Q86SL6_Q8NGU5_Q8TDT5-AGRF2_HUMAN-AGRF2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             7cae44e1-80c0-5952-a779-92928f66caa1\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.47681746G>A\n",
      "x                              396\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.1\n",
      "polyphen_score                 0.997\n",
      "aa_change                      G396D\n",
      "receptor                       Q8IZF7_Q2PNZ1_Q86SL6_Q8NGU5_Q8TDT5-AGRF2_HUMAN-AGRF2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             b145c156-c51c-5e38-854c-a3ba4d9f36ab\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.47714296G>A\n",
      "x                              351\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0.08\n",
      "polyphen_score                 0.819\n",
      "aa_change                      G351S\n",
      "receptor                       Q8IZF3_B3KTD0_Q2PNZ0_Q5T5B5_Q5T5B6_Q86SN9_Q8IXE6-AGRF4_HUMAN-AGRF4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.50 \n",
      "TOTAL: 2 MUTATIONS\n",
      "\n",
      "id                             9f3f1f4f-d802-5456-8fc2-059af2512e29\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.31738611C>T\n",
      "x                              874\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.997\n",
      "aa_change                      C874Y\n",
      "receptor                       O60241_B9EGK9_Q5T6K0_Q8NGW8_Q96GZ9-AGRB2_HUMAN-AGRB2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             c61f0a54-f489-5561-975b-12aa558583d9\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948128G>A\n",
      "x                              805\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.998\n",
      "aa_change                      C805Y\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.51 \n",
      "TOTAL: 6 MUTATIONS\n",
      "\n",
      "id                             9add4ef5-3ca8-5f7a-a8e9-f25efa790b51\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.31738609C>T\n",
      "x                              875\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0.05\n",
      "polyphen_score                 0.788\n",
      "aa_change                      A875T\n",
      "receptor                       O60241_B9EGK9_Q5T6K0_Q8NGW8_Q96GZ9-AGRB2_HUMAN-AGRB2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             36b16b3f-de49-5a2b-aa2f-6fe64b1f34b5\n",
      "y                              1\n",
      "genomic_dna_change             chrX:g.136393556G>A\n",
      "x                              2686\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.01\n",
      "polyphen_score                 0.984\n",
      "aa_change                      A2686T\n",
      "receptor                       Q8IZF6_A2A2J1_A2A2J2_Q5EGP2_Q86SM6-AGRG4_HUMAN-AGRG4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             501c18d1-a334-58ae-b455-51c007b631b8\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948130T>A\n",
      "x                              806\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.04\n",
      "polyphen_score                 0.987\n",
      "aa_change                      S806T\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             95d51eb2-f1f6-5648-83b4-1cf42288e8e6\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948131C>T\n",
      "x                              806\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.994\n",
      "aa_change                      S806L\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             5c4ed059-97cf-579e-b1c1-5064abc786c3\n",
      "y                              1\n",
      "genomic_dna_change             chrX:g.19006243A>G\n",
      "x                              571\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.48\n",
      "polyphen_score                 0.268\n",
      "aa_change                      V571A\n",
      "receptor                       Q8IZP9_B1AWB3_B1AWB4_B1AWB6_B1AWB7_O00406_Q14CE0_Q8IWT2_Q8IZE4-AGRG2_HUMAN-AGRG2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             5fc42ccd-5edf-5ff8-8186-3c0d080ea0d6\n",
      "y                              2\n",
      "genomic_dna_change             chr12:g.131076858G>A\n",
      "x                              511\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.72\n",
      "aa_change                      A511T\n",
      "receptor                       Q6QNK2_B2CKK9_B7ZLF7_Q2M1L3_Q6ZMQ1_Q7Z7M2_Q86SM4-AGRD1_HUMAN-AGRD1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.48 \n",
      "TOTAL: 6 MUTATIONS\n",
      "\n",
      "id                             e97a6ac5-439a-5d46-86ad-bb2893c872c7\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.69076008C>T\n",
      "x                              817\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0.01\n",
      "polyphen_score                 0.736\n",
      "aa_change                      P817L\n",
      "receptor                       O60242_B7Z1K0_O60297_Q2NKN6_Q5VY37_Q9BX54-AGRB3_HUMAN-AGRB3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             86d3f99b-5053-5881-bb8d-bf54a532af2e\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.14158033G>T\n",
      "x                              800\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.665\n",
      "aa_change                      A800D\n",
      "receptor                       O94910_Q96IE7_Q9BU07_Q9HAR3-AGRL1_HUMAN-AGRL1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             62938ea0-0185-5f58-83c7-57e6692c4a7d\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.46859048G>A\n",
      "x                              952\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.522\n",
      "aa_change                      T952M\n",
      "receptor                       Q8IZF2_O94858_Q5TF06_Q6RGN2_Q86SP0_Q9Y3Z2-AGRF5_HUMAN-AGRF5-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             8ae25784-ba46-5118-8693-4b94e5cdcf26\n",
      "y                              1\n",
      "genomic_dna_change             chr3:g.100646599T>C\n",
      "x                              381\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.46\n",
      "polyphen_score                 0.039\n",
      "aa_change                      Y381H\n",
      "receptor                       Q96K78_Q14D94_Q86SQ2-AGRG7_HUMAN-AGRG7-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             3c1a6c88-3778-5edd-bb2d-36a25d28a05c\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.47714284G>T\n",
      "x                              347\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.32\n",
      "polyphen_score                 0.078\n",
      "aa_change                      A347S\n",
      "receptor                       Q8IZF3_B3KTD0_Q2PNZ0_Q5T5B5_Q5T5B6_Q86SN9_Q8IXE6-AGRF4_HUMAN-AGRF4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             cfa8ca04-9c1b-5b6a-bf5f-c3cae7474fc1\n",
      "y                              2\n",
      "genomic_dna_change             chr19:g.6921734C>T\n",
      "x                              548\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.31\n",
      "polyphen_score                 0.108\n",
      "aa_change                      P548S\n",
      "receptor                       Q14246_A6NHV2_B7Z486_B7Z489_E7EPX9_E9PD45_H9KV79_Q2I7G5_Q6ZMN0-AGRE1_HUMAN-AGRE1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.53 \n",
      "TOTAL: 10 MUTATIONS\n",
      "\n",
      "id                             00cda7b1-1103-5423-ab0d-c36b21f60dcf\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.69076023G>A\n",
      "x                              822\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      W822*\n",
      "receptor                       O60242_B7Z1K0_O60297_Q2NKN6_Q5VY37_Q9BX54-AGRB3_HUMAN-AGRB3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             8f4f54c9-df0d-5a13-a16f-c283113d071c\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.81968061G>T\n",
      "x                              791\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.998\n",
      "aa_change                      W791C\n",
      "receptor                       O95490_A5XEI2_B1ALT8_B1ALT9_B1ALU0_B1ALU2_B1ALU4_B1ALU5_B1ALU6-AGRL2_HUMAN-AGRL2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             40e8d265-871e-5282-ba7b-238ffda62bec\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.46859033C>A\n",
      "x                              957\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 1\n",
      "aa_change                      W957L\n",
      "receptor                       Q8IZF2_O94858_Q5TF06_Q6RGN2_Q86SP0_Q9Y3Z2-AGRF5_HUMAN-AGRF5-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             b67ea30f-e44e-5d6e-a7b4-1eaf7fb8cfdf\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948136T>G\n",
      "x                              808\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.996\n",
      "aa_change                      W808G\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             658e74b9-0aae-58e4-8ff0-6e9655503b68\n",
      "y                              2\n",
      "genomic_dna_change             chr1:g.78921751C>T\n",
      "x                              373\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      W373*\n",
      "receptor                       Q9HBW9_B1AR71_Q5KU34-AGRL4_HUMAN-AGRL4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             5008b73b-b202-5869-abd6-bf2f82f60915\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.78921751C>G\n",
      "x                              373\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 1\n",
      "aa_change                      W373C\n",
      "receptor                       Q9HBW9_B1AR71_Q5KU34-AGRL4_HUMAN-AGRL4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             32fbf4ac-329c-5c98-b0f0-85c965b4ef4b\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.78921752C>T\n",
      "x                              373\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      W373*\n",
      "receptor                       Q9HBW9_B1AR71_Q5KU34-AGRL4_HUMAN-AGRL4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             0ca9ee12-6d53-5f8e-bcec-6b73d55df634\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.22413302C>T\n",
      "x                              704\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      W704*\n",
      "receptor                       Q8IWK6_Q6UXK9_Q86SQ5_Q8TC55-AGRA3_HUMAN-AGRA3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             89fc6854-4844-570f-8f79-569d9e9add3d\n",
      "y                              1\n",
      "genomic_dna_change             chr16:g.57656255G>T\n",
      "x                              349\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.99\n",
      "aa_change                      W349C\n",
      "receptor                       Q9Y653_A6NIT7_A6NJV9_B0M0K4_B4DR54_O95966_Q6ZMP1_Q8NGB3_Q96HB4-AGRG1_HUMAN-AGRG1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             d011d4ae-3efd-5887-a4c2-4b0db4ad7e2f\n",
      "y                              2\n",
      "genomic_dna_change             chr19:g.14644238C>T\n",
      "x                              307\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      W307*\n",
      "receptor                       Q9BY15-AGRE3_HUMAN-AGRE3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.46 \n",
      "TOTAL: 6 MUTATIONS\n",
      "\n",
      "id                             9e6af50d-a02f-522a-bc4c-dacd2989b77f\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.81968040C>G\n",
      "x                              784\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.02\n",
      "polyphen_score                 0.986\n",
      "aa_change                      F784L\n",
      "receptor                       O95490_A5XEI2_B1ALT8_B1ALT9_B1ALU0_B1ALU2_B1ALU4_B1ALU5_B1ALU6-AGRL2_HUMAN-AGRL2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             c9ed0bd5-50ec-55b6-a291-c08273929744\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948117C>A\n",
      "x                              801\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.03\n",
      "polyphen_score                 0.987\n",
      "aa_change                      F801L\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             e594017c-e6df-545e-b2ad-9d687851ceb2\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948116T>G\n",
      "x                              801\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.998\n",
      "aa_change                      F801C\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             f4b19a23-1b7d-502e-95d3-26e98bf65d95\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948116T>C\n",
      "x                              801\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0.07\n",
      "polyphen_score                 0.996\n",
      "aa_change                      F801S\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             197a809b-1548-55a6-8553-0e5c0b0c4a51\n",
      "y                              1\n",
      "genomic_dna_change             chr3:g.100646594A>C\n",
      "x                              379\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.05\n",
      "polyphen_score                 0.014\n",
      "aa_change                      Y379S\n",
      "receptor                       Q96K78_Q14D94_Q86SQ2-AGRG7_HUMAN-AGRG7-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             20b6db8e-efe4-5d02-a206-dbf220bb7d5b\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.22413325C>A\n",
      "x                              697\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.53\n",
      "polyphen_score                 0.003\n",
      "aa_change                      A697S\n",
      "receptor                       Q8IWK6_Q6UXK9_Q86SQ5_Q8TC55-AGRA3_HUMAN-AGRA3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.47 \n",
      "TOTAL: 5 MUTATIONS\n",
      "\n",
      "id                             48385186-2da9-58ae-a443-210b4d85d107\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948118A>T\n",
      "x                              802\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.999\n",
      "aa_change                      N802Y\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             97882581-372e-59ea-9142-8cb98ee63422\n",
      "y                              1\n",
      "genomic_dna_change             chr16:g.57566608A>T\n",
      "x                              186\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.25\n",
      "polyphen_score                 0.027\n",
      "aa_change                      T186S\n",
      "receptor                       Q8IZF4_A0A024R6S3_A8K9R8_B3KXZ5_Q6ZMH7_Q6ZML4_Q86SL8_Q8IZ14-AGRG5_HUMAN-AGRG5-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             f6bf80a4-8ce9-5ba8-b5a5-350534babd5e\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.14404407G>A\n",
      "x                              492\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.66\n",
      "polyphen_score                 0.122\n",
      "aa_change                      E492K\n",
      "receptor                       P48960_A8K7Z4_B2RBJ9_O00718_O76101_Q8NG72_Q8TBQ7-AGRE5_HUMAN-AGRE5-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             ba58a26a-1828-5e97-9363-d252ca1f27a5\n",
      "y                              1\n",
      "genomic_dna_change             chr8:g.37837799G>A\n",
      "x                              707\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.01\n",
      "polyphen_score                 0.102\n",
      "aa_change                      E707K\n",
      "receptor                       Q96PE1_A6H8W3_D3DSW4_Q8N3R1_Q8TEM3_Q96KB2_Q9P1Z7_Q9UFY4-AGRA2_HUMAN-AGRA2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             955f23ea-148b-583e-919e-b624da90adbc\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.6921733G>T\n",
      "x                              547\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.55\n",
      "polyphen_score                 0.018\n",
      "aa_change                      R547S\n",
      "receptor                       Q14246_A6NHV2_B7Z486_B7Z489_E7EPX9_E9PD45_H9KV79_Q2I7G5_Q6ZMN0-AGRE1_HUMAN-AGRE1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.56 \n",
      "TOTAL: 1 MUTATIONS\n",
      "\n",
      "id                             d0621cdf-b171-588f-ba66-89fe59d237d4\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.61948145T>C\n",
      "x                              811\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                benign\n",
      "sift_score                     0\n",
      "polyphen_score                 0.376\n",
      "aa_change                      S811P\n",
      "receptor                       Q9HAR2_E9PE04_O94867_Q9NWK5-AGRL3_HUMAN-AGRL3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.54 \n",
      "TOTAL: 4 MUTATIONS\n",
      "\n",
      "id                             f5bf9812-f8f6-5356-bd61-86936c71902b\n",
      "y                              1\n",
      "genomic_dna_change             chr6:g.47009823C>A\n",
      "x                              538\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.961\n",
      "aa_change                      D538Y\n",
      "receptor                       Q5T601_Q5KU15_Q5T5Z9_Q5T600_Q86SM1_Q8IXE3_Q8IZF8_Q96DQ1_Q9H615-AGRF1_HUMAN-AGRF1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             ea2fb5f5-e268-5e0a-bc85-621e3a35fd0c\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.14755088C>T\n",
      "x                              486\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.9\n",
      "polyphen_score                 0.074\n",
      "aa_change                      E486K\n",
      "receptor                       Q9UHX3_B4DQ96_E7ESD7_E9PBR1_E9PEL6_E9PFQ5_E9PG91_Q8NG96_Q9Y4B1-AGRE2_HUMAN-AGRE2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             ce46070a-9384-5b1f-aae4-ba0cfd6cf5f6\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.6921753G>A\n",
      "x                              554\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.37\n",
      "polyphen_score                 0.007\n",
      "aa_change                      S554N\n",
      "receptor                       Q14246_A6NHV2_B7Z486_B7Z489_E7EPX9_E9PD45_H9KV79_Q2I7G5_Q6ZMN0-AGRE1_HUMAN-AGRE1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             98bf737d-eba4-5c5b-8247-60b6b9cc5874\n",
      "y                              1\n",
      "genomic_dna_change             chr19:g.6921752A>G\n",
      "x                              554\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.18\n",
      "polyphen_score                 0.136\n",
      "aa_change                      S554G\n",
      "receptor                       Q14246_A6NHV2_B7Z486_B7Z489_E7EPX9_E9PD45_H9KV79_Q2I7G5_Q6ZMN0-AGRE1_HUMAN-AGRE1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.55 \n",
      "TOTAL: 1 MUTATIONS\n",
      "\n",
      "id                             75234f16-5fa2-51a8-a88d-df7565f3c453\n",
      "y                              1\n",
      "genomic_dna_change             chr1:g.78921746T>C\n",
      "x                              375\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                probably_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.989\n",
      "aa_change                      Y375C\n",
      "receptor                       Q9HBW9_B1AR71_Q5KU34-AGRL4_HUMAN-AGRL4-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "CANCER GENOME ATLAS MUTATIONS AT POSITION S9.49 \n",
      "TOTAL: 4 MUTATIONS\n",
      "\n",
      "id                             6a28c77f-8e84-5e75-ad85-0d8ea35feda0\n",
      "y                              1\n",
      "genomic_dna_change             chrX:g.19006249C>A\n",
      "x                              569\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.11\n",
      "polyphen_score                 0.044\n",
      "aa_change                      R569I\n",
      "receptor                       Q8IZP9_B1AWB3_B1AWB4_B1AWB6_B1AWB7_O00406_Q14CE0_Q8IWT2_Q8IZE4-AGRG2_HUMAN-AGRG2-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             60012958-2def-5acd-bb1a-6b3fc8a011c0\n",
      "y                              1\n",
      "genomic_dna_change             chr4:g.22413316C>T\n",
      "x                              700\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    tolerated\n",
      "polyphen_impact                benign\n",
      "sift_score                     0.1\n",
      "polyphen_score                 0.028\n",
      "aa_change                      V700I\n",
      "receptor                       Q8IWK6_Q6UXK9_Q86SQ5_Q8TC55-AGRA3_HUMAN-AGRA3-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             f564ff9e-c99b-51c9-a766-a421cb339c8b\n",
      "y                              1\n",
      "genomic_dna_change             chr16:g.57656241C>T\n",
      "x                              345\n",
      "consequence                    stop_gained\n",
      "impact                         HIGH\n",
      "sift_impact                    \n",
      "polyphen_impact                \n",
      "aa_change                      Q345*\n",
      "receptor                       Q9Y653_A6NIT7_A6NJV9_B0M0K4_B4DR54_O95966_Q6ZMP1_Q8NGB3_Q96HB4-AGRG1_HUMAN-AGRG1-Homo_sapiens\n",
      "\n",
      "______________________________\n",
      "id                             6d4bb7ad-f9cc-5e25-a8eb-3c45015f7c09\n",
      "y                              1\n",
      "genomic_dna_change             chr12:g.131076852T>G\n",
      "x                              509\n",
      "consequence                    missense\n",
      "impact                         MODERATE\n",
      "sift_impact                    deleterious\n",
      "polyphen_impact                possibly_damaging\n",
      "sift_score                     0\n",
      "polyphen_score                 0.89\n",
      "aa_change                      Y509D\n",
      "receptor                       Q6QNK2_B2CKK9_B7ZLF7_Q2M1L3_Q6ZMQ1_Q7Z7M2_Q86SM4-AGRD1_HUMAN-AGRD1-Homo_sapiens\n",
      "\n",
      "______________________________\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None, None, None, None, None, None, None, None]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#position_vars(loop_muts,'S9-S10', aa_key='x')\n",
    "[position_vars(generalized_mutations, x) for x in generalized_mutations.keys() if 'S9' in x]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "vscode": {
   "interpreter": {
    "hash": "ad450da4e6f5b8cdd3942a14692e4e36ff09ff1e9c8df5cfaea1622af0db4001"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
