{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build a database from a dataset for nomenclating\n",
    "# INPUT: a collection of GAIN domain PDBs, their sequences as one large \".fa\" file\n",
    "from gain_classes import GainDomain, GainCollection, Anchors, GPS\n",
    "import sse_func\n",
    "import execute\n",
    "import numpy as np\n",
    "import glob\n",
    "import multiprocessing as mp\n",
    "from subprocess import Popen, PIPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 15839 best ranked models in target directories.\n"
     ]
    }
   ],
   "source": [
    "# The Goal is to create a GainCollection from all GAIN domains here\n",
    "# The set of proteins needs to be filtered and analyzed statistically before\n",
    "\n",
    "# Example PDB name:\n",
    "#  R0LVI9-R0LVI9_ANAPL-PutativeGR125-Anas_platyrhynchos_unrelaxed_rank_1_model_3.pdb\n",
    "# corresponding STRIDE name:\n",
    "#  R0LVI9-R0LVI9_ANAPL-PutativeGR125-Anas_platyrhynchos.stride\n",
    "pdbs = glob.glob(\"/home/flo/projects/agpcr_nom/*/batch*/*_rank_1_*.pdb\")\n",
    "print(f\"Found {len(pdbs)} best ranked models in target directories.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "stride_folder = f\"./all_gps_stride\"\n",
    "stride_bin = \"/home/hildilab/lib/stride/stride\"\n",
    "           \n",
    "def compile_stride_mp_list(pdbs, stride_folder,stride_bin):\n",
    "    stride_mp_list = []\n",
    "    \n",
    "    for pdb in pdbs:\n",
    "        pdb_name = pdb.split(\"/\")[-1]\n",
    "        name = pdb_name.split(\"_unrelaxed_\")[0]\n",
    "        out_file = f\"{stride_folder}/{name}.stride\"\n",
    "        arg = [pdb, out_file, stride_bin]\n",
    "        \n",
    "        stride_mp_list.append(arg)\n",
    "        \n",
    "    return stride_mp_list\n",
    "\n",
    "def run_stride(arg):\n",
    "    pdb_file, out_file, stride_bin = arg\n",
    "    stride_command = f\"{stride_bin} {pdb_file} -f{out_file}\"\n",
    "    execute.run_command(stride_command)\n",
    "\n",
    "def execute_stride_mp(stride_mp_list, n_threads=10):\n",
    "        stride_pool = mp.Pool(n_threads)\n",
    "        stride_pool.map(run_stride, stride_mp_list)\n",
    "        print(\"Completed mutithreaded creation of STRIDE files!\")\n",
    "        \n",
    "        #execute.run_stride(pdb, out_file, stride_bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'stride_mp_list = compile_stride_mp_list(pdbs, stride_folder, stride_bin)\\nstride_mp_list = compile_stride_mp_list(singlet_pdbs, stride_folder, stride_bin)\\nprint(len(stride_mp_list))\\n[print(arg) for arg in stride_mp_list[:10]]'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"stride_mp_list = compile_stride_mp_list(pdbs, stride_folder, stride_bin)\n",
    "stride_mp_list = compile_stride_mp_list(singlet_pdbs, stride_folder, stride_bin)\n",
    "print(len(stride_mp_list))\n",
    "[print(arg) for arg in stride_mp_list[:10]]\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MP execution of STRIDE\n",
    "#execute_stride_mp(stride_mp_list, n_threads=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reduced the initial set of 15839 PDB files down to 14773 files.\n",
      "Found 14773/14773 accessions in the accession list.\n"
     ]
    }
   ],
   "source": [
    "# Eliminate double entries (both in the original run and the added small runs)\n",
    "# Form the \"pdbs\" list\n",
    "\n",
    "stride_files = glob.glob(\"/home/flo/projects/agpcr_nom/all_gps_stride/*\")\n",
    "accessions = [f.split(\".strid\")[0].split(\"/\")[-1] for f in stride_files]\n",
    "pdb_accessions = np.array([p.split(\"_unrelaxed_\")[0].split(\"/\")[-1] for p in pdbs])\n",
    "\n",
    "# Find duplicate in the original pdbs list and indicate them via > is_duplicate = True <\n",
    "is_duplicate=np.zeros([len(pdbs)], dtype=bool)\n",
    "sort_pdb_ac = np.sort(pdb_accessions)\n",
    "duplicate_list = []\n",
    "for i, pdb in enumerate(sort_pdb_ac):\n",
    "    if i+1 == len(sort_pdb_ac):\n",
    "        break\n",
    "    if pdb == sort_pdb_ac[i+1]:\n",
    "        duplicate_list.append(pdb)\n",
    "        multi_indices = np.where(pdb == pdb_accessions)[0]\n",
    "        is_duplicate[multi_indices[0]] = True\n",
    "\n",
    "np_pdbs = np.array(pdbs)\n",
    "singlet_pdbs = np_pdbs[is_duplicate == False] # This is the reduced list with ONLY UNIQUE PDBs\n",
    "print(f\"Reduced the initial set of {len(pdbs)} PDB files down to {len(singlet_pdbs)} files.\")\n",
    "\n",
    "\n",
    "# This is a check routine if there are PDBs in the reduced list which have NOT a STRIDE file\n",
    "singlet_pdb_accessions = np.array([p.split(\"_unrelaxed_\")[0].split(\"/\")[-1] for p in singlet_pdbs])\n",
    "\n",
    "counter = 0\n",
    "for ac in singlet_pdb_accessions:\n",
    "    if ac not in accessions:\n",
    "        print(ac)\n",
    "    else:\n",
    "        counter += 1\n",
    "print(f\"Found {counter}/{len(singlet_pdb_accessions)} accessions in the accession list.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'stride_folder = f\"{pdb_folder}/stride_files\"\\nfasta_file = \"/home/hildilab/projects/agpcr_nom/batch_files/batch_60.fa\"\\n# CHECK FILES!\\nimport os\\nif not os.path.isfile(fasta_file) or os.path.islink(fasta_file):\\n    print(\"ERROR: Specify FASTA FILE CONTAINING ALL SEQUENCES OF THE PROTEINS!\")\\n    \\n    \\nalignment_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/mafft.fa\"\\n#alignment_file = None\\n\\n# If there is not a specified base alignment, create one (that might take a while tho.)\\nif not alignment_file:\\n    mafft_bin = \"mafft\"\\n    mafft_command = f\"{mafft_bin} --auto --thread 4 {fasta_file}\"\\n    out_dir = \"/\".join(fasta_file.split(\"/\")[:-2])\\n    out_file = f\"{out_dir}/mafft.fa\"\\n    execute.run_command(mafft_command, out_file = out_file)\\n    alignment_file = out_file\\ngps_minus_one_column = 1209'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"stride_folder = f\"{pdb_folder}/stride_files\"\n",
    "fasta_file = \"/home/hildilab/projects/agpcr_nom/batch_files/batch_60.fa\"\n",
    "# CHECK FILES!\n",
    "import os\n",
    "if not os.path.isfile(fasta_file) or os.path.islink(fasta_file):\n",
    "    print(\"ERROR: Specify FASTA FILE CONTAINING ALL SEQUENCES OF THE PROTEINS!\")\n",
    "    \n",
    "    \n",
    "alignment_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/mafft.fa\"\n",
    "#alignment_file = None\n",
    "\n",
    "# If there is not a specified base alignment, create one (that might take a while tho.)\n",
    "if not alignment_file:\n",
    "    mafft_bin = \"mafft\"\n",
    "    mafft_command = f\"{mafft_bin} --auto --thread 4 {fasta_file}\"\n",
    "    out_dir = \"/\".join(fasta_file.split(\"/\")[:-2])\n",
    "    out_file = f\"{out_dir}/mafft.fa\"\n",
    "    execute.run_command(mafft_command, out_file = out_file)\n",
    "    alignment_file = out_file\n",
    "gps_minus_one_column = 1209\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The GainCollection class needs to be modified to also parse a list of sequences \n",
    "# instead of a folder containing one fasta per seq\n",
    "class FilterCollection:\n",
    "    ''' \n",
    "    A collection of GainDomain objects used for filtering a set of AF2 models\n",
    "    This is used to condense the dataset towards one containing only GAIN domains\n",
    "\n",
    "    Attributes\n",
    "    ----------\n",
    "\n",
    "    collection : list\n",
    "        List of GainDomain instances\n",
    "   \n",
    "    valid_gps : np.array(bool)\n",
    "        For each protein, specify if the GPS detection is valid or not\n",
    "    \n",
    "    valid_subdomain : np.array(bool)\n",
    "        For each protein, specify if it has detected subdomains or not\n",
    "    \n",
    "    Methods\n",
    "    ----------\n",
    "    print_gps(self):\n",
    "        Prints info about all GainDomain GPS subinstances\n",
    "\n",
    "    write_all_seq(self, savename):\n",
    "        Writes all sequences in the Collection to one file\n",
    "\n",
    "    transform_alignment(self, input_alignment, output_alignment, aln_cutoff):\n",
    "        Transforms a given alignment with SSE data to a SSE-assigned version of the alignment\n",
    "        \n",
    "    write_filtered(self, savename, bool_mask):\n",
    "        Writes all sequences to File where a boolean mask (i.e. subdomain_criterion, gps_criterion)\n",
    "        is True at respective position\n",
    "    '''\n",
    "    def __init__(self,\n",
    "                alignment_file, \n",
    "                aln_cutoff,\n",
    "                quality,\n",
    "                gps_index,\n",
    "                stride_files,\n",
    "                sequence_files=None, # modified to object containing all seqs\n",
    "                sequences=None, # replaces sequence_files\n",
    "                subdomain_bracket_size=20,\n",
    "                domain_threshold=20,\n",
    "                coil_weight=0,\n",
    "                alignment_dict=None): \n",
    "        '''\n",
    "        Constructs the GainCollection objects by initializing one GainDomain instance per given sequence\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        alignment_file:     str, required\n",
    "            The base dataset alignment file\n",
    "\n",
    "        aln_cutoff:         int, required\n",
    "            Integer value of last alignment column\n",
    "\n",
    "        quality:            list, required\n",
    "            List of quality valus for each alignment column. Has to have $aln_cutoff items\n",
    "            By default, would take in the annotated Blosum62 values from the alignment exported from Jalview\n",
    "\n",
    "        gps_index:          int, required\n",
    "            Alignment column index of the GPS-1 residue (consensus: Leu)\n",
    "        \n",
    "        stride_files:       list, required\n",
    "            A list of stride files corresponding to the sequences contained within.\n",
    "        \n",
    "        sequence_files:     list, optional\n",
    "            A list of sequence files to be read as the collection - specify either this\n",
    "            or sequences as an object instead of files for sequences\n",
    "        \n",
    "        sequences:          object, optional\n",
    "            A list of (sequence_name, sequence) tuples containing all sequences. Can be specified\n",
    "            instead of sequence_files\n",
    "\n",
    "        subdomain_bracket_size: int, optional\n",
    "            Smoothing window size for the signal convolve function. Default = 20.\n",
    "\n",
    "        domain_threshold:   int, optional\n",
    "            Minimum size of a helical segment to be considered candidate for Subdomain A of GAIN domain. Default = 20.\n",
    "\n",
    "        coil_weight:        float, optional\n",
    "            Weight assigned to unordered residues during Subdomain detection. Enables decay of helical signal\n",
    "            default = 0. Recommended values < +0.2 for decay\n",
    "\n",
    "        Returns\n",
    "        ----------\n",
    "        None\n",
    "        '''\n",
    "        # Initialize collection (containing all GainDomain instances) and the anchor-histogram\n",
    "        if sequence_files:\n",
    "            # Compile all sequence files to a sequences object\n",
    "            sequences = np.empty([len(sequence_files)])\n",
    "            for i, seq_file in enumerate(sequence_files):\n",
    "                name, seq = read_seq(seq_file, return_name = True)\n",
    "                sequences[i] = (name, seq)\n",
    "        elif (sequences is not None):\n",
    "            print(f\"Found sequences object.\")\n",
    "        else: \n",
    "            print(f\"ERROR: no sequence_files or sequences parameter found. Aborting compilation.\")\n",
    "            return None\n",
    "        self.collection = np.empty([len(sequences)], dtype=object)\n",
    "        \n",
    "        valid_gps = np.zeros([len(sequences)], dtype=bool)\n",
    "        valid_subdomain = np.zeros([len(sequences)], dtype=bool)\n",
    "        #anchor_hist = np.zeros([aln_cutoff])#\n",
    "        # Create a GainDomain instance for each sequence file contained in the list\n",
    "        for i,seq_tup in enumerate(sequences):\n",
    "            # updated GainDomain class\n",
    "            name, sequence = seq_tup\n",
    "            explicit_stride = [stride for stride in stride_files if name.split(\"-\")[0] in stride]\n",
    "            if len(explicit_stride) == 0:\n",
    "                print(f\"Stride file not found for {name}\")\n",
    "            newGain = GainDomain(alignment_file, \n",
    "                                  aln_cutoff,\n",
    "                                  quality,\n",
    "                                  name = name,\n",
    "                                  sequence = sequence,\n",
    "                                  gps_index = gps_index, \n",
    "                                  subdomain_bracket_size = subdomain_bracket_size,\n",
    "                                  domain_threshold = domain_threshold,\n",
    "                                  coil_weight = coil_weight,\n",
    "                                  explicit_stride_file = explicit_stride[0],\n",
    "                                  without_anchors = True,\n",
    "                                  skip_naming = True,\n",
    "                                  alignment_dict = alignment_dict)\n",
    "\n",
    "            # Check if the object staisfies minimum criteria\n",
    "            if newGain.isValid: \n",
    "                  \n",
    "                self.collection[i] = newGain\n",
    "                  \n",
    "                if newGain.hasSubdomain:\n",
    "                    valid_subdomain[i] = True\n",
    "                if newGain.GPS.isConsensus:\n",
    "                    valid_gps[i] = True\n",
    "        \n",
    "        self.valid_subdomain = valid_subdomain\n",
    "        self.valid_gps = valid_gps\n",
    "        print(f\"Completed Collection initialitazion of {len(sequences)} sequences.\\n\"\n",
    "             f\"{np.count_nonzero(self.collection)} valid proteins were found.\\n\"\n",
    "             f\"{np.count_nonzero(self.valid_subdomain)} of which have detected Subdomains.\\n\"\n",
    "             f\"{np.count_nonzero(self.valid_gps)} of which have detected consensus GPS motifs.\\n\")\n",
    "\n",
    "    def print_gps(self):\n",
    "        '''\n",
    "        Prints information about the GPS of each GAIN domain.\n",
    "\n",
    "        Parameters:\n",
    "            None\n",
    "        Returns:\n",
    "            None\n",
    "        '''\n",
    "        for i, gain in enumerate(self.collection):\n",
    "            try:\n",
    "                gain.GPS.info()\n",
    "            except:\n",
    "                print(f\"No GPS data for {gain.name}. Likely not a GAIN Domain!\")\n",
    "\n",
    "    def write_all_seq(self, savename):\n",
    "        '''\n",
    "        Write all GAIN sequences of the GainCollection into one fasta file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        savename: str, required\n",
    "            Output name of the fasta file.\n",
    "\n",
    "        Returns\n",
    "        ----------\n",
    "        None\n",
    "        '''\n",
    "        with open(savename, 'w') as f:\n",
    "            for gain in self.collection:\n",
    "                f.write(f\">{gain.name[:-3]}\\n{''.join(gain.sequence)}\\n\")\n",
    "    \n",
    "    def write_filtered(self, savename, bool_mask=None, write_mode='w'):\n",
    "                  \n",
    "        '''\n",
    "        Internal function for writing filtered sequences to file.\n",
    "        Takes the Gain.sequence np.array type to write the truncated versions.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        savename: str, required\n",
    "            Output name of the fasta file.\n",
    "        bool_mask: list/array, required\n",
    "            A mask of len(self.collection) where a boolean defines whether to write the\n",
    "            sequence to file or not\n",
    "        Returns\n",
    "        ----------\n",
    "        None\n",
    "        '''\n",
    "        with open(savename, write_mode) as f:\n",
    "            print(f\"writing filtered to {savename}\")\n",
    "            for i, gain in enumerate(self.collection):\n",
    "                if gain is not None and bool_mask[i] == True:\n",
    "                    f.write(f\">{gain.name.replace('.fa','')}\\n{''.join(gain.sequence)}\\n\")\n",
    "\n",
    "    def transform_alignment(self, input_alignment, output_alignment, aln_cutoff):\n",
    "        ''' \n",
    "        Transform any input alignment containing all sequences in the GainCollection \n",
    "        into one where each residue is replaced with the respective \n",
    "        Secondary Structure from the STRIDE files\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_alignment: str, required\n",
    "            Input alignment file\n",
    "        output_alignment: str, required\n",
    "            Output alignment file\n",
    "        aln_cutoff: int, required\n",
    "            Last alignment column to be read from the Input Alignment\n",
    "\n",
    "        Returns\n",
    "        ---------\n",
    "        None\n",
    "        '''\n",
    "        initial_dict = sse_func.read_alignment(input_alignment, aln_cutoff)\n",
    "        out_dict = {}\n",
    "        for gain in self.collection:\n",
    "            sse_alignment_row = np.full([aln_cutoff], fill_value='-', dtype='<U1')\n",
    "            mapper = sse_func.get_indices(gain.name, gain.sequence, input_alignment, aln_cutoff)\n",
    "            for index, resid in enumerate(gain.sse_sequence):\n",
    "                sse_alignment_row[mapper[index]] = resid\n",
    "            out_dict[gain.name[:-3]] = sse_alignment_row\n",
    "\n",
    "        # Write to file\n",
    "        with open(output_alignment, \"w\") as f:\n",
    "            for key in out_dict.keys():\n",
    "                f.write(f\">{key}\\n{''.join(out_dict[key])}\\n\")\n",
    "        print(f\"Done transforming alignment {input_alignment} to {output_alignment} with SSE data.\")\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'alignment_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/big_mafft.fa\"\\nquality_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/big_mafft.jal\"\\nfasta_file = \"/home/hildilab/projects/agpcr_nom/batch_files/batch_60.fa\"\\nstride_folder = f\"{pdb_folder}/stride_files\"\\nquality = sse_func.read_quality(quality_file)\\ngps_minus_one = 19258\\naln_cutoff = 19822\\nsequences = sse_func.read_multi_seq(fasta_file)\\nprint(len(sequences))\\nstride_files = glob.glob(\"/home/hildilab/projects/agpcr_nom/batch_out_test/stride_files/*\")\\nprint(len(stride_files))'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"alignment_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/big_mafft.fa\"\n",
    "quality_file = \"/home/hildilab/projects/agpcr_nom/batch_out_test/big_mafft.jal\"\n",
    "fasta_file = \"/home/hildilab/projects/agpcr_nom/batch_files/batch_60.fa\"\n",
    "stride_folder = f\"{pdb_folder}/stride_files\"\n",
    "quality = sse_func.read_quality(quality_file)\n",
    "gps_minus_one = 19258\n",
    "aln_cutoff = 19822\n",
    "sequences = sse_func.read_multi_seq(fasta_file)\n",
    "print(len(sequences))\n",
    "stride_files = glob.glob(\"/home/hildilab/projects/agpcr_nom/batch_out_test/stride_files/*\")\n",
    "print(len(stride_files))\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'filterCollection = FilterCollection(alignment_file,\\n                                   aln_cutoff = 19822,\\n                                   quality = quality,\\n                                   gps_index = gps_minus_one,\\n                                   stride_files = stride_files,\\n                                   sequences = sequences)'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"filterCollection = FilterCollection(alignment_file,\n",
    "                                   aln_cutoff = 19822,\n",
    "                                   quality = quality,\n",
    "                                   gps_index = gps_minus_one,\n",
    "                                   stride_files = stride_files,\n",
    "                                   sequences = sequences)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filterCollection.write_filtered(savename=\"test.fa\", bool_mask = filterCollection.valid_gps, write_mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def batch_filter_seqs(arg_item):\n",
    "                   # [sequences,      # A number of sequences as tuple instances\n",
    "                   #  stride_folder,  # A folder containing ALL stride files\n",
    "                   #   output_prefix,  # A prefix for individual file identification\n",
    "                   #   alignment_file, # The big (initial) alignment file\n",
    "                   #   quality,        # The corr. parsed quality for BLOSUM62 score\n",
    "                   #   aln_cutoff,     # the left-most column (19822 for big_mafft.fa)\n",
    "                   #   gps_minus_one,  # The column index of GPS-1 (zero-indexed! 19258 big_mafft)\n",
    "                   #   ]\n",
    "    sequences, stride_folder, output_prefix, alignment_file, quality, aln_cutoff, gps_minus_one, alignment_dict = arg_item\n",
    "    # Parallelizable version of filtering sequences and models via FilterCollection\n",
    "    # This should create separate files for each valid, fragment and no-SD group\n",
    "    # These files should then be grouped together\n",
    "    # The batch size is arbitrary and is considered the number of sequences passed\n",
    "    \n",
    "    # Output: Profiles; 4 Text files (valid, fragment, invalidGPS, invalid)\n",
    "    filteredBatch = FilterCollection(alignment_file,\n",
    "                                   aln_cutoff = aln_cutoff,\n",
    "                                   quality = quality,\n",
    "                                   gps_index = gps_minus_one,\n",
    "                                   stride_files = stride_files,\n",
    "                                   sequences = sequences,\n",
    "                                   alignment_dict = alignment_dict)\n",
    "    outpath = \"/home/flo/projects/agpcr_nom/all_gps_profiles\"\n",
    "    \n",
    "    for Gain in filteredBatch.collection:\n",
    "        if Gain:\n",
    "            Gain.plot_profile(outdir=outpath)\n",
    "        \n",
    "    suffixes = [\"gain\", \n",
    "                \"fragments\", \n",
    "                \"noncons_gps\", \n",
    "                \"invalid\"]\n",
    "    masks = [np.logical_and(filteredBatch.valid_gps, filteredBatch.valid_subdomain),\n",
    "              np.logical_not(filteredBatch.valid_subdomain),\n",
    "              np.logical_not(filteredBatch.valid_gps),\n",
    "              np.logical_not(np.logical_and(filteredBatch.valid_gps, filteredBatch.valid_subdomain))]\n",
    "    # write four separate files, matching each criterion\n",
    "    for k in range(4):\n",
    "        filteredBatch.write_filtered(savename=f\"{outpath}/{output_prefix}_{suffixes[k]}.fa\", \n",
    "                                     bool_mask = masks[k],\n",
    "                                     write_mode = 'w')\n",
    "    return None\n",
    "\n",
    "def compile_fastas(prefixes, out_prefix):\n",
    "    # Compiles the fasta files together to construct one large file containing the sequences\n",
    "    # satisfying each criterion in the 2x2 matrix\n",
    "    # we want to have the GAIN sequence only that is output by the write_filtered() func.\n",
    "    \n",
    "    # Gather all files:\n",
    "    suffixes = [\"gain\", \n",
    "                \"fragments\", \n",
    "                \"noncons_gps\", \n",
    "                \"invalid\"]\n",
    "    all_files = np.asarray([glob.glob(f\"{prefix}*fa\") for prefix in prefixes]).flatten()\n",
    "    \n",
    "    for suffix in suffixes:\n",
    "        sub_list = [f for f in all_files if suffix in f.split(\"_\")[-1]]\n",
    "        print(f\"Sublist constructed for {suffix = } containing {len(sub_list)} files.\")\n",
    "        with open(f\"{out_prefix}_{suffixes}.fa\", \"w\") as all_file:\n",
    "            for file in sublist:\n",
    "                with open(file, \"w\") as f:\n",
    "                    data = f.read()\n",
    "                all_file.write(data)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_mp_collection(arg_list, n_threads=10):\n",
    "    pool = mp.Pool(n_threads)\n",
    "    pool.map(batch_filter_seqs, arg_list)\n",
    "    print(\"Completed mutithreaded filtering.\")\n",
    "\n",
    "def construct_arg_list(batch_sequence_files, \n",
    "                       output_folder,\n",
    "                       stride_folder, \n",
    "                       quality, \n",
    "                       alignment_file, \n",
    "                       aln_cutoff, \n",
    "                       gps_minus_one,\n",
    "                       alignment_dict = None):\n",
    "    \"\"\" each item looks like this:\n",
    "        sequences, \\ \n",
    "        stride_folder, \\\n",
    "        output_prefix, \\\n",
    "        alignment_file, \\\n",
    "        quality, \\\n",
    "        aln_cutoff, \\\n",
    "        gps_minus_one = arg_item\"\"\"\n",
    "    # static : stride_folder, quality, alignment_file, aln_cutoff, gps_minus_one\n",
    "    # flexible : sequences, output_prefix\n",
    "    arg_list = []\n",
    "    #\n",
    "    for idx, sequence_file in enumerate(batch_sequence_files):\n",
    "        \n",
    "        index_string = str(idx)\n",
    "        sequences = sse_func.read_multi_seq(sequence_file)\n",
    "        output_prefix = f\"{output_folder}_{index_string.zfill(3)}\"\n",
    "        item = [sequences, \n",
    "                stride_folder, \n",
    "                output_prefix, \n",
    "                alignment_file, \n",
    "                quality, \n",
    "                aln_cutoff, \n",
    "                gps_minus_one,\n",
    "                alignment_dict]\n",
    "        \n",
    "        arg_list.append(item)\n",
    "    \n",
    "    print(f\"[NOTE] : Compiled list of arguments for multithreaded filtering\"\n",
    "          f\" containing {len(arg_list)} items.\")\n",
    "    return arg_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14773\n",
      "465\n",
      "[NOTE] : Compiled list of arguments for multithreaded filtering containing 465 items.\n",
      "10\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n",
      "Next ITEM\n",
      "75\n",
      "43\n",
      "24\n",
      "41\n",
      "19822\n",
      "19822\n",
      "19258\n",
      "14792\n"
     ]
    }
   ],
   "source": [
    "batch_sequence_files = glob.glob(\"/home/flo/projects/agpcr_nom/all_gp*/batch_*.fa\")\n",
    "output_folder = \"all_gps_gain_domains\"\n",
    "\n",
    "alignment_file = \"/home/flo/projects/agpcr_nom/big_mafft.fa\"\n",
    "quality_file = \"/home/flo/projects/agpcr_nom/big_mafft.jal\"\n",
    "stride_folder = \"/home/flo/projects/agpcr_nom/all_gps_stride\"\n",
    "quality = sse_func.read_quality(quality_file)\n",
    "gps_minus_one = 19258\n",
    "aln_cutoff = 19822\n",
    "stride_files = glob.glob(\"/home/flo/projects/agpcr_nom/all_gps_stride/*\")\n",
    "alignment_dict = sse_func.read_alignment(alignment_file, aln_cutoff)\n",
    "print(len(stride_files))\n",
    "print(len(batch_sequence_files))\n",
    "#print(quality)\n",
    "#sequences = sse_func.read_multi_seq(fasta_file)\n",
    "\n",
    "arg_list = construct_arg_list(batch_sequence_files, \n",
    "                       output_folder,\n",
    "                       stride_folder, \n",
    "                       quality, \n",
    "                       alignment_file, \n",
    "                       aln_cutoff, \n",
    "                       gps_minus_one,\n",
    "                       alignment_dict)\n",
    "print(len(test_arg))\n",
    "for arg in test_arg:\n",
    "    print(\"Next ITEM\")\n",
    "    for item in arg:\n",
    "        try:print(len(item))\n",
    "        except: print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found sequences object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0SQF4-A0A2D0SQF4_ICTPU-mucin-2-likeisoformX2-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0SQF4-A0A2D0SQF4_ICTPU-mucin-2-likeisoformX2-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0SQF8-A0A2D0SQF8_ICTPU-mucin-2-likeisoformX1-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0SQF8-A0A2D0SQF8_ICTPU-mucin-2-likeisoformX1-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0SQM0-A0A2D0SQM0_ICTPU-AGRF5-likeisoformX5-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0SQM0-A0A2D0SQM0_ICTPU-AGRF5-likeisoformX5-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0SQM5-A0A2D0SQM5_ICTPU-AGRF5-likeisoformX10-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0SQM5-A0A2D0SQM5_ICTPU-AGRF5-likeisoformX10-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0LUM3-A0A2I0LUM3_COLLI-AGRL3-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0LUM3-A0A2I0LUM3_COLLI-AGRL3-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0LZF5-A0A2I0LZF5_COLLI-AGRV1-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 71(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0LZF5-A0A2I0LZF5_COLLI-AGRV1-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 269, start = 518\n",
      "DEBUG: self.indices = [268, 269, 270]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0ME39-A0A2I0ME39_COLLI-AGRL2,transcriptvariantX1-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0ME39-A0A2I0ME39_COLLI-AGRL2,transcriptvariantX1-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0ME41-A0A2I0ME41_COLLI-AGRL2,transcriptvariantX3-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0ME41-A0A2I0ME41_COLLI-AGRL2,transcriptvariantX3-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0ME47-A0A2I0ME47_COLLI-AGRL2,transcriptvariantX4-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0ME47-A0A2I0ME47_COLLI-AGRL2,transcriptvariantX4-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 292, start = 495\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0ME50-A0A2I0ME50_COLLI-AGRL2,transcriptvariantX5-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0ME50-A0A2I0ME50_COLLI-AGRL2,transcriptvariantX5-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0MHB3-A0A2I0MHB3_COLLI-Cadherin,EGFLAGseven-passG-typereceptor1-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0MHB3-A0A2I0MHB3_COLLI-Cadherin,EGFLAGseven-passG-typereceptor1-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0MLD0-A0A2I0MLD0_COLLI-AGRG2-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 44(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0MLD0-A0A2I0MLD0_COLLI-AGRG2-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 239, start = 548\n",
      "DEBUG: self.indices = [238, 239, 240]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0MTF5-A0A2I0MTF5_COLLI-AGRG6-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0MTF5-A0A2I0MTF5_COLLI-AGRG6-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0MU76-A0A2I0MU76_COLLI-AGRF5,transcriptvariantX1-Columba_livia\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A2I0MU76-A0A2I0MU76_COLLI-AGRF5,transcriptvariantX1-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0TKP4-A0A2I0TKP4_LIMLA-Uncharacterizedprotein-Limosa_lapponica_baueri.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 75(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A2I0TKP4-A0A2I0TKP4_LIMLA-Uncharacterizedprotein-Limosa_lapponica_baueri., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 275, start = 512\n",
      "DEBUG: self.indices = [274, 275, 276]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0U3M0-A0A2I0U3M0_LIMLA-Uncharacterizedprotein-Limosa_lapponica_baueri.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0U3M0-A0A2I0U3M0_LIMLA-Uncharacterizedprotein-Limosa_lapponica_baueri., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I0UQN4-A0A2I0UQN4_LIMLA-Agprotein-coupledreceptorl2isoformx1-Limosa_lapponica_baueri.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 99(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I0UQN4-A0A2I0UQN4_LIMLA-Agprotein-coupledreceptorl2isoformx1-Limosa_lapponica_baueri., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 292, start = 495\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2U088-A0A2I2U088_FELCA-AGRG6-Felis_catus\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2U088-A0A2I2U088_FELCA-AGRG6-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2U0D3-A0A2I2U0D3_FELCA-AGRL2-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2U0D3-A0A2I2U0D3_FELCA-AGRL2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found sequences object.\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663EIR9-A0A663EIR9_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A2I2U0N2-A0A2I2U0N2_FELCA-AGRL2-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A663EIR9-A0A663EIR9_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A2I2U0N2-A0A2I2U0N2_FELCA-AGRL2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone  A0A2I2U5R8-A0A2I2U5R8_FELCA-AGRL3-Felis_catusA0A663EIS9-A0A663EIS9_AQUCH-AGRB2-Aquila_chrysaetos_chrysaetos.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2U5R8-A0A2I2U5R8_FELCA-AGRL3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A663EIS9-A0A663EIS9_AQUCH-AGRB2-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2U7T9-A0A2I2U7T9_FELCA-AGRB2-Felis_catus A0A663EIV8-A0A663EIV8_AQUCH-AGRB2-Aquila_chrysaetos_chrysaetos.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A663EIV8-A0A663EIV8_AQUCH-AGRB2-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A2I2U7T9-A0A2I2U7T9_FELCA-AGRB2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "DEBUG: minus_one_residue = 323, start = 464\n",
      "DEBUG: self.indices = [322, 323, 324]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663EIW8-A0A663EIW8_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant) \n",
      "DEBUG: Getting alignment indices with: A0A663EIW8-A0A663EIW8_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A2I2UB34-A0A2I2UB34_FELCA-AGRL1-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UB34-A0A2I2UB34_FELCA-AGRL1-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 308, start = 479None \n",
      "A0A663EIX3-A0A663EIX3_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.DEBUG: self.indices = [307, 308, 309]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663EIX3-A0A663EIX3_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2UL54-A0A2I2UL54_FELCA-AGRL2-Felis_catus\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UL54-A0A2I2UL54_FELCA-AGRL2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663EIX8-A0A663EIX8_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663EIX8-A0A663EIX8_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [306, 307, 308] \n",
      "None A0A2I2UPT2-A0A2I2UPT2_FELCA-AGRB2-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "NoneDEBUG: Getting alignment indices with: A0A2I2UPT2-A0A2I2UPT2_FELCA-AGRB2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " A0A663EJ46-A0A663EJ46_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663EJ46-A0A663EJ46_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 323, start = 464\n",
      "DEBUG: self.indices = [322, 323, 324]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneDEBUG: self.indices = [306, 307, 308] A0A2I2US02-A0A2I2US02_FELCA-AGRL3-Felis_catus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2US02-A0A2I2US02_FELCA-AGRL3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663EJ79-A0A663EJ79_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: Getting alignment indices with: A0A663EJ79-A0A663EJ79_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2UTI9-A0A2I2UTI9_FELCA-AGRF5-Felis_catusDEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UTI9-A0A2I2UTI9_FELCA-AGRF5-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663EJQ3-A0A663EJQ3_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos.\n",
      "DEBUG: minus_one_residue = 265, start = 522\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663EJQ3-A0A663EJQ3_AQUCH-AGRL3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2UVD9-A0A2I2UVD9_FELCA-AGRL2-Felis_catus\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UVD9-A0A2I2UVD9_FELCA-AGRL2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A663ENR9-A0A663ENR9_AQUCH-AGRG6-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663ENR9-A0A663ENR9_AQUCH-AGRG6-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A2I2UWE8-A0A2I2UWE8_FELCA-AGRL3-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UWE8-A0A2I2UWE8_FELCA-AGRL3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 256, start = 531\n",
      "\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 307, start = 480None\n",
      " A0A663ETC3-A0A663ETC3_AQUCH-CadherinEGFLAGseven-passG-typereceptor1-Aquila_chrysaetos_chrysaetos.DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A663ETC3-A0A663ETC3_AQUCH-CadherinEGFLAGseven-passG-typereceptor1-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A2I2UZI8-A0A2I2UZI8_FELCA-AGRB2-Felis_catusDEBUG: minus_one_residue = 329, start = 458\n",
      "\n",
      "DEBUG: self.indices = [328, 329, 330]\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2UZI8-A0A2I2UZI8_FELCA-AGRB2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "None  A0A663FF54-A0A663FF54_AQUCH-AGRL1-Aquila_chrysaetos_chrysaetos.DEBUG: minus_one_residue = 323, start = 464\n",
      "\n",
      "DEBUG: self.indices = [322, 323, 324]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 104(new variant) \n",
      "A0A2I2V4L1-A0A2I2V4L1_FELCA-AGRB3-Felis_catusDEBUG: Getting alignment indices with: A0A663FF54-A0A663FF54_AQUCH-AGRL1-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2V4L1-A0A2I2V4L1_FELCA-AGRB3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 289, start = 498\n",
      "DEBUG: self.indices = [288, 289, 290]\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663FH35-A0A663FH35_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 73(new variant)None\n",
      "DEBUG: Getting alignment indices with: A0A663FH35-A0A663FH35_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "A0A2I2V523-A0A2I2V523_FELCA-AGRL2-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2V523-A0A2I2V523_FELCA-AGRL2-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 312, start = 475\n",
      "DEBUG: self.indices = [311, 312, 313]\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307] \n",
      "None A0A663FHW4-A0A663FHW4_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A663FHW4-A0A663FHW4_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None A0A2I2Y2R6-A0A2I2Y2R6_GORGO-AGRL1-Gorilla_gorilla_gorilla\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2Y2R6-A0A2I2Y2R6_GORGO-AGRL1-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 338, start = 449\n",
      "\n",
      "DEBUG: self.indices = [337, 338, 339]\n",
      "DEBUG: minus_one_residue = 309, start = 478\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [308, 309, 310]None\n",
      " A0A663FIF8-A0A663FIF8_AQUCH-AGRL1-Aquila_chrysaetos_chrysaetos.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 105(new variant) \n",
      "A0A2I2Y7F0-A0A2I2Y7F0_GORGO-AGRL2-Gorilla_gorilla_gorilla\n",
      "DEBUG: Getting alignment indices with: A0A663FIF8-A0A663FIF8_AQUCH-AGRL1-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2Y7F0-A0A2I2Y7F0_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 499\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [292, 293, 294]None\n",
      " A0A663FJ65-A0A663FJ65_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A663FJ65-A0A663FJ65_AQUCH-CadherinEGFLAGseven-passG-typereceptor3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A2I2Y9M5-A0A2I2Y9M5_GORGO-AGRL3-Gorilla_gorilla_gorilla\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: minus_one_residue = 326, start = 461DEBUG: Getting alignment indices with: A0A2I2Y9M5-A0A2I2Y9M5_GORGO-AGRL3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [325, 326, 327]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: self.indices = [306, 307, 308]None A0A663M824-A0A663M824_ATHCN-CadherinEGFLAGseven-passG-typereceptor3-Athene_cunicularia\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663M824-A0A663M824_ATHCN-CadherinEGFLAGseven-passG-typereceptor3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2YCA1-A0A2I2YCA1_GORGO-AGRL3-Gorilla_gorilla_gorilla\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)DEBUG: minus_one_residue = 333, start = 454\n",
      "DEBUG: Getting alignment indices with: A0A2I2YCA1-A0A2I2YCA1_GORGO-AGRL3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [332, 333, 334]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 307, start = 480 \n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "None \n",
      "Initializing GainDomain Object.\n",
      "A0A663M924-A0A663M924_ATHCN-AGRG6-Athene_cunicularia None\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)A0A2I2YS18-A0A2I2YS18_GORGO-AGRL2-Gorilla_gorilla_gorilla\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A663M924-A0A663M924_ATHCN-AGRG6-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2YS18-A0A2I2YS18_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone  A0A663M951-A0A663M951_ATHCN-AGRG6-Athene_cuniculariaA0A2I2YTV0-A0A2I2YTV0_GORGO-AGRL2-Gorilla_gorilla_gorilla\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)DEBUG: Getting alignment indices with: A0A663M951-A0A663M951_ATHCN-AGRG6-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]DEBUG: Getting alignment indices with: A0A2I2YTV0-A0A2I2YTV0_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 307, start = 480None\n",
      " DEBUG: self.indices = [306, 307, 308]\n",
      "A0A663M958-A0A663M958_ATHCN-AGRG6-Athene_cunicularia\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 59(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A663M958-A0A663M958_ATHCN-AGRG6-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None A0A2I2YWH2-A0A2I2YWH2_GORGO-AGRL3-Gorilla_gorilla_gorilla\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2YWH2-A0A2I2YWH2_GORGO-AGRL3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663MDG4-A0A663MDG4_ATHCN-AGRL2-Athene_cunicularia\n",
      "DEBUG: minus_one_residue = 307, start = 480Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 107(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: A0A663MDG4-A0A663MDG4_ATHCN-AGRL2-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2YYL5-A0A2I2YYL5_GORGO-AGRB2-Gorilla_gorilla_gorillaDEBUG: minus_one_residue = 305, start = 482\n",
      "\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2YYL5-A0A2I2YYL5_GORGO-AGRB2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663MQB6-A0A663MQB6_ATHCN-AGRL3-Athene_cunicularia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663MQB6-A0A663MQB6_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 322, start = 465\n",
      "\n",
      "DEBUG: self.indices = [321, 322, 323]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2YZB3-A0A2I2YZB3_GORGO-AGRL3-Gorilla_gorilla_gorilla\n",
      "Found sequences object.\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A2I2YZB3-A0A2I2YZB3_GORGO-AGRL3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663MSP5-A0A663MSP5_ATHCN-AGRL3-Athene_cunicularia\n",
      " None A0A6I9MIZ0-A0A6I9MIZ0_PERMB-AGRF1isoformX1-Peromyscus_maniculatus_bairdii\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663MSP5-A0A663MSP5_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6I9MIZ0-A0A6I9MIZ0_PERMB-AGRF1isoformX1-Peromyscus_maniculatus_bairdii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2Z070-A0A2I2Z070_GORGO-AGRL2-Gorilla_gorilla_gorillaDEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 261, start = 299\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant) None\n",
      " DEBUG: Getting alignment indices with: A0A2I2Z070-A0A2I2Z070_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "None A0A663MSQ2-A0A663MSQ2_ATHCN-AGRL3-Athene_cunicularia\n",
      "A0A6J0CN99-A0A6J0CN99_PERMB-AGRF1isoformX3-Peromyscus_maniculatus_bairdiiDEBUG: minus_one_residue = 307, start = 480Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: A0A663MSQ2-A0A663MSQ2_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 89(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6J0CN99-A0A6J0CN99_PERMB-AGRF1isoformX3-Peromyscus_maniculatus_bairdii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None A0A2I2Z089-A0A2I2Z089_GORGO-AGRG4-Gorilla_gorilla_gorilla\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 261, start = 299\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 64(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters. \n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: Getting alignment indices with: A0A2I2Z089-A0A2I2Z089_GORGO-AGRG4-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>NoneNone\n",
      "  A0A6P3VXV3-A0A6P3VXV3_CLUHA-AGRD1isoformX2-Clupea_harengus\n",
      "A0A663MSW3-A0A663MSW3_ATHCN-AGRL3-Athene_cunicularia\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 55(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P3VXV3-A0A6P3VXV3_CLUHA-AGRD1isoformX2-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663MSW3-A0A663MSW3_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 270, start = 517\n",
      "DEBUG: self.indices = [269, 270, 271]\n",
      "DEBUG: minus_one_residue = 325, start = 235\n",
      "DEBUG: self.indices = [324, 325, 326]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 307, start = 480A0A2I2ZDM2-A0A2I2ZDM2_GORGO-AGRL3-Gorilla_gorilla_gorilla\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)None \n",
      "A0A7K9BF27-A0A7K9BF27_DRONO-AGRA3protein-Dromaius_novaehollandiae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A2I2ZDM2-A0A2I2ZDM2_GORGO-AGRL3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)\n",
      "\n",
      " DEBUG: Getting alignment indices with: A0A7K9BF27-A0A7K9BF27_DRONO-AGRA3protein-Dromaius_novaehollandiae, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A663MSX8-A0A663MSX8_ATHCN-AGRL3-Athene_cunicularia\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: Getting alignment indices with: A0A663MSX8-A0A663MSX8_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 327, start = 233\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: self.indices = [326, 327, 328]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "  NoneDEBUG: self.indices = [306, 307, 308]None A0A2I2ZFY4-A0A2I2ZFY4_GORGO-AGRL2-Gorilla_gorilla_gorilla \n",
      "\n",
      "A0A7L0TQ58-A0A7L0TQ58_CHOAC-AGRA3protein-Chordeiles_acutipennis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZFY4-A0A2I2ZFY4_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A7L0TQ58-A0A7L0TQ58_CHOAC-AGRA3protein-Chordeiles_acutipennis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 306, start = 481DEBUG: minus_one_residue = 327, start = 233A0A663MTE6-A0A663MTE6_ATHCN-AGRL3-Athene_cunicularia\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [326, 327, 328]DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A663MTE6-A0A663MTE6_ATHCN-AGRL3-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A2I2ZG69-A0A2I2ZG69_GORGO-AGRB2-Gorilla_gorilla_gorilla\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A7L1HT18-A0A7L1HT18_9CHAR-AGRA3protein-Nycticryphes_semicollaris.Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)DEBUG: minus_one_residue = 294, start = 493\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]DEBUG: Getting alignment indices with: A0A7L1HT18-A0A7L1HT18_9CHAR-AGRA3protein-Nycticryphes_semicollaris., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A2I2ZG69-A0A2I2ZG69_GORGO-AGRB2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 322, start = 465 \n",
      "A0A663N8E4-A0A663N8E4_ATHCN-AGRV1-Athene_cuniculariaDEBUG: minus_one_residue = 327, start = 233\n",
      "\n",
      "DEBUG: self.indices = [321, 322, 323]\n",
      "DEBUG: self.indices = [326, 327, 328]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 71(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663N8E4-A0A663N8E4_ATHCN-AGRV1-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A2I2ZGX8-A0A2I2ZGX8_GORGO-AGRB2-Gorilla_gorilla_gorilla\n",
      "DEBUG: minus_one_residue = 268, start = 519\n",
      "DEBUG: self.indices = [267, 268, 269]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant) \n",
      "None A0A0A0MXC8-A0A0A0MXC8_PAPAN-AGRE1-Papio_anubis\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZGX8-A0A2I2ZGX8_GORGO-AGRB2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> None\n",
      " Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 66(new variant)A0A663N959-A0A663N959_ATHCN-CadherinEGFLAGseven-passG-typereceptor1-Athene_cunicularia\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A0A0MXC8-A0A0A0MXC8_PAPAN-AGRE1-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663N959-A0A663N959_ATHCN-CadherinEGFLAGseven-passG-typereceptor1-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 322, start = 465\n",
      "DEBUG: self.indices = [321, 322, 323]DEBUG: minus_one_residue = 260, start = 301\n",
      "\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 330, start = 457 None\n",
      " DEBUG: self.indices = [329, 330, 331]\n",
      "Initializing GainDomain Object.\n",
      "A0A2I2ZHS1-A0A2I2ZHS1_GORGO-AGRG6-Gorilla_gorilla_gorilla\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 59(new variant) \n",
      "A0A2I3LRM6-A0A2I3LRM6_PAPAN-AGRE1-Papio_anubisDEBUG: Getting alignment indices with: A0A2I2ZHS1-A0A2I2ZHS1_GORGO-AGRG6-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663NAR3-A0A663NAR3_ATHCN-AGRV1-Athene_cuniculariaFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 66(new variant)\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 71(new variant)DEBUG: Getting alignment indices with: A0A2I3LRM6-A0A2I3LRM6_PAPAN-AGRE1-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 260, start = 527\n",
      "\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A663NAR3-A0A663NAR3_ATHCN-AGRV1-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 260, start = 301 A0A2I2ZJJ8-A0A2I2ZJJ8_GORGO-AGRG6-Gorilla_gorilla_gorilla\n",
      "DEBUG: minus_one_residue = 269, start = 518\n",
      "DEBUG: self.indices = [268, 269, 270]\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 59(new variant)DEBUG: self.indices = [259, 260, 261]\n",
      "\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A2I2ZJJ8-A0A2I2ZJJ8_GORGO-AGRG6-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5L209-A0A2K5L209_CERAT-AGRD1-Cercocebus_atys\n",
      " A0A663NBU2-A0A663NBU2_ATHCN-CadherinEGFLAGseven-passG-typereceptor1-Athene_cunicularia\n",
      "DEBUG: minus_one_residue = 260, start = 527Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 77(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "DEBUG: Getting alignment indices with: A0A663NBU2-A0A663NBU2_ATHCN-CadherinEGFLAGseven-passG-typereceptor1-Athene_cunicularia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2K5L209-A0A2K5L209_CERAT-AGRD1-Cercocebus_atys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "NoneDEBUG: minus_one_residue = 329, start = 458 A0A2I2ZKW6-A0A2I2ZKW6_GORGO-AGRL2-Gorilla_gorilla_gorilla\n",
      "DEBUG: self.indices = [328, 329, 330]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZKW6-A0A2I2ZKW6_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 308, start = 253 None A0A665T6A0-A0A665T6A0_ECHNA-AGRV1-Echeneis_naucrates\n",
      "\n",
      "DEBUG: self.indices = [307, 308, 309]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 70(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481 \n",
      "DEBUG: self.indices = [305, 306, 307]None\n",
      " A0A452IQV5-A0A452IQV5_9SAUR-AGRB3-Gopherus_agassizii\n",
      "DEBUG: Getting alignment indices with: A0A665T6A0-A0A665T6A0_ECHNA-AGRV1-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 96(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A452IQV5-A0A452IQV5_9SAUR-AGRB3-Gopherus_agassizii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A2I2ZLL6-A0A2I2ZLL6_GORGO-AGRB2-Gorilla_gorilla_gorilla\n",
      "DEBUG: minus_one_residue = 263, start = 518\n",
      "DEBUG: minus_one_residue = 286, start = 275DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "DEBUG: self.indices = [285, 286, 287]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZLL6-A0A2I2ZLL6_GORGO-AGRB2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A665TER5-A0A665TER5_ECHNA-AGRV1-Echeneis_naucrates\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 70(new variant)None A0A4W3I9M0-A0A4W3I9M0_CALMI-Uncharacterizedprotein-Callorhinchus_milii\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665TER5-A0A665TER5_ECHNA-AGRV1-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 322, start = 465\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: self.indices = [321, 322, 323]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A4W3I9M0-A0A4W3I9M0_CALMI-Uncharacterizedprotein-Callorhinchus_milii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 263, start = 518 \n",
      "A0A2I2ZSZ8-A0A2I2ZSZ8_GORGO-AGRG4-Gorilla_gorilla_gorillaDEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "DEBUG: minus_one_residue = 320, start = 241\n",
      "DEBUG: self.indices = [319, 320, 321]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 63(new variant)\n",
      "A0A665UX19-A0A665UX19_ECHNA-AGRG2-like-Echeneis_naucratesWARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZSZ8-A0A2I2ZSZ8_GORGO-AGRG4-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 270, start = 517Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 58(new variant) \n",
      "None DEBUG: Getting alignment indices with: A0A665UX19-A0A665UX19_ECHNA-AGRG2-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A673TJ69-A0A673TJ69_SURSU-AGRD2-Suricata_suricatta\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A673TJ69-A0A673TJ69_SURSU-AGRD2-Suricata_suricatta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [269, 270, 271]\n",
      "DEBUG: minus_one_residue = 265, start = 522\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 279, start = 282\n",
      " \n",
      "NoneDEBUG: self.indices = [278, 279, 280]\n",
      " A0A2I2ZX12-A0A2I2ZX12_GORGO-AGRL2-Gorilla_gorilla_gorilla\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None  A0A665V0H9-A0A665V0H9_ECHNA-Uncharacterizedprotein-Echeneis_naucratesNone\n",
      " A0A7K7G3V6-A0A7K7G3V6_ERIRU-AGRA3protein-Erithacus_rubecula\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)DEBUG: Getting alignment indices with: A0A2I2ZX12-A0A2I2ZX12_GORGO-AGRL2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A665V0H9-A0A665V0H9_ECHNA-Uncharacterizedprotein-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7K7G3V6-A0A7K7G3V6_ERIRU-AGRA3protein-Erithacus_rubecula, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 261, start = 526DEBUG: minus_one_residue = 293, start = 494DEBUG: minus_one_residue = 328, start = 233\n",
      "\n",
      "DEBUG: self.indices = [260, 261, 262]DEBUG: self.indices = [327, 328, 329]\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  None\n",
      "Initializing GainDomain Object.\n",
      " NoneA0A2R9CP18-A0A2R9CP18_PANPA-AGRG2-Pan_paniscus \n",
      "None A0A2I2ZXA9-A0A2I2ZXA9_GORGO-AGRB3-Gorilla_gorilla_gorilla\n",
      " A0A665VRC6-A0A665VRC6_ECHNA-AGRL1a-Echeneis_naucratesFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 47(new variant)DEBUG: Getting alignment indices with: A0A2I2ZXA9-A0A2I2ZXA9_GORGO-AGRB3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2R9CP18-A0A2R9CP18_PANPA-AGRG2-Pan_paniscus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665VRC6-A0A665VRC6_ECHNA-AGRL1a-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 240, start = 322\n",
      "DEBUG: minus_one_residue = 287, start = 500DEBUG: self.indices = [239, 240, 241]\n",
      "\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None DEBUG: minus_one_residue = 303, start = 484\n",
      "None  DEBUG: self.indices = [302, 303, 304]A0A3B1KG58-A0A3B1KG58_ASTMX-AGRD2-Astyanax_mexicanus\n",
      "A0A2I3H7B3-A0A2I3H7B3_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B1KG58-A0A3B1KG58_ASTMX-AGRD2-Astyanax_mexicanus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)None\n",
      " DEBUG: Getting alignment indices with: A0A2I3H7B3-A0A2I3H7B3_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A665VRD9-A0A665VRD9_ECHNA-AGRL1a-Echeneis_naucrates\n",
      "DEBUG: minus_one_residue = 325, start = 237\n",
      "DEBUG: self.indices = [324, 325, 326]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665VRD9-A0A665VRD9_ECHNA-AGRL1a-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 303, start = 484\n",
      "DEBUG: self.indices = [302, 303, 304]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None  A0A2I3HQC7-A0A2I3HQC7_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 58(new variant)\n",
      "A0A3Q0CX88-A0A3Q0CX88_MESAU-AGRG2isoformX8-Mesocricetus_auratus\n",
      "DEBUG: Getting alignment indices with: A0A2I3HQC7-A0A2I3HQC7_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A665VSH6-A0A665VSH6_ECHNA-AGRL1a-Echeneis_naucrates\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 47(new variant)DEBUG: minus_one_residue = 260, start = 527\n",
      "\n",
      "DEBUG: self.indices = [259, 260, 261]DEBUG: Getting alignment indices with: A0A3Q0CX88-A0A3Q0CX88_MESAU-AGRG2isoformX8-Mesocricetus_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665VSH6-A0A665VSH6_ECHNA-AGRL1a-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 237, start = 325 \n",
      "DEBUG: self.indices = [236, 237, 238]None\n",
      " A0A2I3LF52-A0A2I3LF52_PAPAN-AGRL3-Papio_anubisDEBUG: minus_one_residue = 303, start = 484\n",
      "\n",
      "DEBUG: self.indices = [302, 303, 304]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant) \n",
      "DEBUG: Getting alignment indices with: A0A2I3LF52-A0A2I3LF52_PAPAN-AGRL3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "None A0A6J2QBP1-A0A6J2QBP1_COTGO-AGRD1isoformX5-Cottoperca_gobio\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 51(new variant)A0A665W8M1-A0A665W8M1_ECHNA-AGRL2-Echeneis_naucrates\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: Getting alignment indices with: A0A6J2QBP1-A0A6J2QBP1_COTGO-AGRD1isoformX5-Cottoperca_gobio, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665W8M1-A0A665W8M1_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found sequences object.\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 256\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneDEBUG: minus_one_residue = 292, start = 495None \n",
      " A0A2I3LGY1-A0A2I3LGY1_PAPAN-AGRL2-Papio_anubisDEBUG: self.indices = [291, 292, 293]\n",
      "A0A3P9IHJ8-A0A3P9IHJ8_ORYLA-Uncharacterizedprotein-Oryzias_latipes\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 78(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I3LGY1-A0A2I3LGY1_PAPAN-AGRL2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A3P9IHJ8-A0A3P9IHJ8_ORYLA-Uncharacterizedprotein-Oryzias_latipes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None  \n",
      "NoneA0A665W924-A0A665W924_ECHNA-AGRL2-Echeneis_naucrates \n",
      "A0A6P5J799-A0A6P5J799_PHACI-LOWQUALITYPROTEIN:AGRE1-Phascolarctos_cinereus\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "DEBUG: self.indices = [292, 293, 294]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 63(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665W924-A0A665W924_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 309, start = 422DEBUG: Getting alignment indices with: A0A6P5J799-A0A6P5J799_PHACI-LOWQUALITYPROTEIN:AGRE1-Phascolarctos_cinereus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I3LND2-A0A2I3LND2_PAPAN-AGRB2-Papio_anubis\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)DEBUG: self.indices = [308, 309, 310]\n",
      "\n",
      "\n",
      "DEBUG: minus_one_residue = 262, start = 300DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: Getting alignment indices with: A0A2I3LND2-A0A2I3LND2_PAPAN-AGRB2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " A0A3P9LVF3-A0A3P9LVF3_ORYLA-Uncharacterizedprotein-Oryzias_latipes\n",
      " None DEBUG: minus_one_residue = 322, start = 465Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 78(new variant)A0A665W937-A0A665W937_ECHNA-AGRL2-Echeneis_naucrates\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [321, 322, 323]DEBUG: Getting alignment indices with: A0A3P9LVF3-A0A3P9LVF3_ORYLA-Uncharacterizedprotein-Oryzias_latipes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A665W937-A0A665W937_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A6P7MDI3-A0A6P7MDI3_BETSP-AGRL2isoformX14-Betta_splendens\n",
      "DEBUG: minus_one_residue = 309, start = 422\n",
      "DEBUG: self.indices = [308, 309, 310]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 104(new variant) \n",
      "\n",
      "NoneDEBUG: Getting alignment indices with: A0A6P7MDI3-A0A6P7MDI3_BETSP-AGRL2isoformX14-Betta_splendens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "DEBUG: minus_one_residue = 305, start = 482A0A2I3LRM3-A0A2I3LRM3_PAPAN-AGRL2-Papio_anubis\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      " DEBUG: self.indices = [304, 305, 306]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3LRM3-A0A2I3LRM3_PAPAN-AGRL2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 306, start = 256\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "A0A5A9NF58-A0A5A9NF58_9TELE-AGRL2-Triplophysa_tibetana.DEBUG: minus_one_residue = 306, start = 481\n",
      "  NoneNone  A0A665W9E4-A0A665W9E4_ECHNA-AGRL2-Echeneis_naucratesA0A6P8F3M2-A0A6P8F3M2_CLUHA-AGRE1-likeisoformX2-Clupea_harengus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 98(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 14 helical_counts[maxk] = 50(new variant)DEBUG: Getting alignment indices with: A0A5A9NF58-A0A5A9NF58_9TELE-AGRL2-Triplophysa_tibetana., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665W9E4-A0A665W9E4_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P8F3M2-A0A6P8F3M2_CLUHA-AGRE1-likeisoformX2-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "DEBUG: minus_one_residue = 292, start = 439\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      " DEBUG: self.indices = [291, 292, 293]\n",
      "None \n",
      "DEBUG: minus_one_residue = 230, start = 332DEBUG: self.indices = [304, 305, 306]A0A2I3LSZ0-A0A2I3LSZ0_PAPAN-AGRG6-Papio_anubis\n",
      "\n",
      "DEBUG: self.indices = [229, 230, 231]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3LSZ0-A0A2I3LSZ0_PAPAN-AGRG6-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A671DP09-A0A671DP09_RHIFE-AGRA3-Rhinolophus_ferrumequinum\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "  NoneFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)None  \n",
      "A0A665W9V5-A0A665W9V5_ECHNA-AGRL2-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A671DP09-A0A671DP09_RHIFE-AGRA3-Rhinolophus_ferrumequinum, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "A0A6Q2XKP5-A0A6Q2XKP5_ESOLU-AGRA2-Esox_lucius\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 76(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6Q2XKP5-A0A6Q2XKP5_ESOLU-AGRA2-Esox_lucius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 107(new variant)\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665W9V5-A0A665W9V5_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 321, start = 410\n",
      " DEBUG: self.indices = [320, 321, 322]None\n",
      "DEBUG: minus_one_residue = 291, start = 271\n",
      "DEBUG: self.indices = [290, 291, 292]DEBUG: minus_one_residue = 305, start = 482 \n",
      "\n",
      "A0A2I3LVM0-A0A2I3LVM0_PAPAN-AGRL3-Papio_anubisDEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "Initializing GainDomain Object.\n",
      "  None\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None DEBUG: Getting alignment indices with: A0A2I3LVM0-A0A2I3LVM0_PAPAN-AGRL3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None \n",
      "A0A6J3RDY6-A0A6J3RDY6_TURTR-AGRA3isoformX2-Tursiops_truncatus \n",
      "A0A7J8GNF3-A0A7J8GNF3_MOLMO-AGRF1-Molossus_molossus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)A0A665WAL9-A0A665WAL9_ECHNA-AGRL2-Echeneis_naucratesFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6J3RDY6-A0A6J3RDY6_TURTR-AGRA3isoformX2-Tursiops_truncatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A7J8GNF3-A0A7J8GNF3_MOLMO-AGRF1-Molossus_molossus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WAL9-A0A665WAL9_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 321, start = 410\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 265, start = 297None\n",
      "DEBUG: minus_one_residue = 293, start = 494 DEBUG: self.indices = [264, 265, 266]A0A2I3LYL5-A0A2I3LYL5_PAPAN-AGRL2-Papio_anubis\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "DEBUG: self.indices = [292, 293, 294]A0A7J8AHP7-A0A7J8AHP7_MYOMY-AGRA3-Myotis_myotis\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7J8AHP7-A0A7J8AHP7_MYOMY-AGRA3-Myotis_myotis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A2I3LYL5-A0A2I3LYL5_PAPAN-AGRL2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " \n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "None  A0A7J8GP44-A0A7J8GP44_MOLMO-AGRF1-Molossus_molossusA0A665WAN1-A0A665WAN1_ECHNA-AGRL2-Echeneis_naucrates\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 107(new variant)DEBUG: Getting alignment indices with: A0A7J8GP44-A0A7J8GP44_MOLMO-AGRF1-Molossus_molossus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481DEBUG: Getting alignment indices with: A0A665WAN1-A0A665WAN1_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: minus_one_residue = 321, start = 410\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "DEBUG: minus_one_residue = 265, start = 297\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: self.indices = [264, 265, 266]NoneDEBUG: self.indices = [304, 305, 306]\n",
      " A0A2I3M001-A0A2I3M001_PAPAN-AGRL2-Papio_anubis\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J8JPP9-A0A7J8JPP9_MOLMO-AGRA3-Molossus_molossus\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      " \n",
      "NoneDEBUG: Getting alignment indices with: A0A2I3M001-A0A2I3M001_PAPAN-AGRL2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      " A0A7K8JFF6-A0A7K8JFF6_9GRUI-AGRA3protein-Lophotis_ruficrista.None\n",
      " A0A665WB43-A0A665WB43_ECHNA-AGRL2-Echeneis_naucratesFound Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7J8JPP9-A0A7J8JPP9_MOLMO-AGRA3-Molossus_molossus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8JFF6-A0A7K8JFF6_9GRUI-AGRA3protein-Lophotis_ruficrista., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A665WB43-A0A665WB43_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "DEBUG: minus_one_residue = 321, start = 410\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "DEBUG: minus_one_residue = 329, start = 233\n",
      "DEBUG: minus_one_residue = 305, start = 482DEBUG: self.indices = [328, 329, 330]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      " \n",
      "None\n",
      "Initializing GainDomain Object.\n",
      "  A0A2I3M0I5-A0A2I3M0I5_PAPAN-AGRG4-Papio_anubisNone\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "F1RX85-F1RX85_PIG-AGRA2-Sus_scrofa \n",
      "\n",
      "Initializing GainDomain Object.\n",
      "None  None Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "A0A665WBQ6-A0A665WBQ6_ECHNA-AGRL2-Echeneis_naucrates\n",
      "DEBUG: Getting alignment indices with: F1RX85-F1RX85_PIG-AGRA2-Sus_scrofa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A7K8L8I9-A0A7K8L8I9_9GRUI-AGRA3protein-Ardeotis_kori.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WBQ6-A0A665WBQ6_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 66(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I3M0I5-A0A2I3M0I5_PAPAN-AGRG4-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A7K8L8I9-A0A7K8L8I9_9GRUI-AGRA3protein-Ardeotis_kori., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 314, start = 417\n",
      "DEBUG: self.indices = [313, 314, 315]\n",
      "DEBUG: minus_one_residue = 303, start = 484\n",
      "DEBUG: self.indices = [302, 303, 304]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 329, start = 233DEBUG: minus_one_residue = 269, start = 518 \n",
      "H2M724-H2M724_ORYLA-AGRA2-Oryzias_latipesDEBUG: self.indices = [328, 329, 330]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: self.indices = [268, 269, 270]Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 78(new variant)None\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " DEBUG: Getting alignment indices with: H2M724-H2M724_ORYLA-AGRA2-Oryzias_latipes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A665WBV5-A0A665WBV5_ECHNA-AGRL2-Echeneis_naucrates\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      "A0A7L4LY69-A0A7L4LY69_GLAPT-AGRA3protein-Glareola_pratincola\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "None Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)DEBUG: Getting alignment indices with: A0A665WBV5-A0A665WBV5_ECHNA-AGRL2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A2I3M7N7-A0A2I3M7N7_PAPAN-AGRB2-Papio_anubis\n",
      "\n",
      "\n",
      "DEBUG: minus_one_residue = 309, start = 422DEBUG: Getting alignment indices with: A0A7L4LY69-A0A7L4LY69_GLAPT-AGRA3protein-Glareola_pratincola, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [308, 309, 310]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3M7N7-A0A2I3M7N7_PAPAN-AGRB2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 329, start = 233 DEBUG: minus_one_residue = 305, start = 482\n",
      "\n",
      "NoneDEBUG: self.indices = [328, 329, 330]DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "DEBUG: minus_one_residue = 322, start = 465\n",
      " DEBUG: self.indices = [321, 322, 323]A0A218UJ73-A0A218UJ73_9PASE-GR124-Lonchura_striata_domestica\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 79(new variant)None\n",
      "Initializing GainDomain Object.\n",
      "  A0A665WQS1-A0A665WQS1_ECHNA-AGRB2-Echeneis_naucratesNone\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      "  NoneA0A7L4MVK6-A0A7L4MVK6_9AVES-AGRF5protein-Ceyx_cyanopectusDEBUG: Getting alignment indices with: A0A218UJ73-A0A218UJ73_9PASE-GR124-Lonchura_striata_domestica, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "\n",
      "A0A2I3MB54-A0A2I3MB54_PAPAN-AGRG6-Papio_anubis\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 96(new variant)Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 89(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I3MB54-A0A2I3MB54_PAPAN-AGRG6-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7L4MVK6-A0A7L4MVK6_9AVES-AGRF5protein-Ceyx_cyanopectus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 317, start = 415\n",
      "\n",
      "DEBUG: self.indices = [316, 317, 318]DEBUG: Getting alignment indices with: A0A665WQS1-A0A665WQS1_ECHNA-AGRB2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: minus_one_residue = 255, start = 307None\n",
      "DEBUG: self.indices = [254, 255, 256] \n",
      "A0A2D0R1P2-A0A2D0R1P2_ICTPU-LOWQUALITYPROTEIN:AGRD2-Ictalurus_punctatus\n",
      "DEBUG: minus_one_residue = 275, start = 512DEBUG: self.indices = [257, 258, 259]\n",
      "DEBUG: self.indices = [274, 275, 276]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0R1P2-A0A2D0R1P2_ICTPU-LOWQUALITYPROTEIN:AGRD2-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None W5PYD4-W5PYD4_SHEEP-AGRG2-Ovis_aries\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None  A0A2I3MBV8-A0A2I3MBV8_PAPAN-AGRL3-Papio_anubis\n",
      "None Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 56(new variant)A0A665WQW8-A0A665WQW8_ECHNA-AGRB1a-Echeneis_naucrates\n",
      "\n",
      "DEBUG: minus_one_residue = 325, start = 407Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)DEBUG: Getting alignment indices with: W5PYD4-W5PYD4_SHEEP-AGRG2-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3MBV8-A0A2I3MBV8_PAPAN-AGRL3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [324, 325, 326]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665WQW8-A0A665WQW8_ECHNA-AGRB1a-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 307, start = 480None\n",
      " DEBUG: minus_one_residue = 349, start = 213DEBUG: self.indices = [306, 307, 308]A0A2K5X4J8-A0A2K5X4J8_MACFA-AGRL3-Macaca_fascicularis\n",
      "\n",
      "DEBUG: minus_one_residue = 285, start = 502DEBUG: self.indices = [348, 349, 350]\n",
      "\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  DEBUG: Getting alignment indices with: A0A2K5X4J8-A0A2K5X4J8_MACFA-AGRL3-Macaca_fascicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None \n",
      "A0A2I3MI30-A0A2I3MI30_PAPAN-AGRB2-Papio_anubis \n",
      "None A0A665WRF8-A0A665WRF8_ECHNA-AGRB1a-Echeneis_naucrates\n",
      "NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      " DEBUG: minus_one_residue = 307, start = 425A0A096M0M3-A0A096M0M3_POEFO-Uncharacterizedprotein-Poecilia_formosaFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 77(new variant)DEBUG: Getting alignment indices with: A0A2I3MI30-A0A2I3MI30_PAPAN-AGRB2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665WRF8-A0A665WRF8_ECHNA-AGRB1a-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: Getting alignment indices with: A0A096M0M3-A0A096M0M3_POEFO-Uncharacterizedprotein-Poecilia_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 322, start = 465\n",
      "DEBUG: self.indices = [321, 322, 323]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None DEBUG: minus_one_residue = 278, start = 509A0A2Y9DJQ7-A0A2Y9DJQ7_TRIMA-AGRD2-Trichechus_manatus_latirostris\n",
      "\n",
      "DEBUG: self.indices = [277, 278, 279]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I3MRN7-A0A2I3MRN7_PAPAN-AGRF5-Papio_anubis\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 85(new variant)\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A2Y9DJQ7-A0A2Y9DJQ7_TRIMA-AGRD2-Trichechus_manatus_latirostris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A665WRK1-A0A665WRK1_ECHNA-AGRB2-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A2I3MRN7-A0A2I3MRN7_PAPAN-AGRF5-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 356, start = 207\n",
      "\n",
      "DEBUG: self.indices = [355, 356, 357]\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WRK1-A0A665WRK1_ECHNA-AGRB2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 266, start = 521 \n",
      "A0A1D5P588_A0A2H4I7A1-A0A1D5P588_CHICK-GR133variantX2-Gallus_gallusDEBUG: minus_one_residue = 297, start = 435DEBUG: self.indices = [265, 266, 267]\n",
      "\n",
      "DEBUG: self.indices = [296, 297, 298]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 74(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A1D5P588_A0A2H4I7A1-A0A1D5P588_CHICK-GR133variantX2-Gallus_gallus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 275, start = 512\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [274, 275, 276]\n",
      " None DEBUG: minus_one_residue = 308, start = 255NoneA0A2I3MY58-A0A2I3MY58_PAPAN-AGRL3-Papio_anubis\n",
      "\n",
      "DEBUG: self.indices = [307, 308, 309]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "\n",
      " A0A337S149-A0A337S149_FELCA-AGRA3-Felis_catusDEBUG: Getting alignment indices with: A0A2I3MY58-A0A2I3MY58_PAPAN-AGRL3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None \n",
      " None\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant) \n",
      "A0A1S3W4F3-A0A1S3W4F3_ERIEU-AGRF1isoformX2-Erinaceus_europaeusA0A665WRL4-A0A665WRL4_ECHNA-AGRB2-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A337S149-A0A337S149_FELCA-AGRA3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 74(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 321, start = 411\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WRL4-A0A665WRL4_ECHNA-AGRB2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A1S3W4F3-A0A1S3W4F3_ERIEU-AGRF1isoformX2-Erinaceus_europaeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneNo GPS residue present in this Domain. Is this really a GAIN? \n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = None, start = 587 \n",
      "NoneA0A2I3N2T1-A0A2I3N2T1_PAPAN-AGRF5-Papio_anubisALTERNATIVE GPS DETECTION!\n",
      " DEBUG: minus_one_residue = 265, start = 298A0A3B3DBA4-A0A3B3DBA4_ORYME-AGRA2-Oryzias_melastigma \n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)DEBUG:last_sheets = [(771, 777), (780, 785)]\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "DEBUG: Getting alignment indices with: A0A2I3N2T1-A0A2I3N2T1_PAPAN-AGRF5-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [191, 192]\n",
      "Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 78(new variant)\n",
      "\n",
      "\n",
      "KeyNotFound: 310HelixDEBUG: Getting alignment indices with: A0A3B3DBA4-A0A3B3DBA4_ORYME-AGRA2-Oryzias_melastigma, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A665WSW0-A0A665WSW0_ECHNA-AGRB1b-Echeneis_naucrates\n",
      "DEBUG: minus_one_residue = 309, start = 423\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneDEBUG: self.indices = [308, 309, 310] \n",
      "A0A2I3N5E1-A0A2I3N5E1_PAPAN-AGRL3-Papio_anubis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant) \n",
      "None DEBUG: Getting alignment indices with: A0A665WSW0-A0A665WSW0_ECHNA-AGRB1b-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A1S3W4X8-A0A1S3W4X8_ERIEU-AGRF1isoformX1-Erinaceus_europaeus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      " \n",
      "NoneDEBUG: Getting alignment indices with: A0A2I3N5E1-A0A2I3N5E1_PAPAN-AGRL3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A3Q2ZWK1-A0A3Q2ZWK1_KRYMA-AGRL2-like-Kryptolebias_marmoratus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 98(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A3Q2ZWK1-A0A3Q2ZWK1_KRYMA-AGRL2-like-Kryptolebias_marmoratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 274, start = 513\n",
      "DEBUG: self.indices = [273, 274, 275]\n",
      "DEBUG: Getting alignment indices with: A0A1S3W4X8-A0A1S3W4X8_ERIEU-AGRF1isoformX1-Erinaceus_europaeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: minus_one_residue = 291, start = 441\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " \n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: self.indices = [290, 291, 292]\n",
      "\n",
      "NoneDEBUG: minus_one_residue = 265, start = 298 \n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "A0A665WT11-A0A665WT11_ECHNA-AGRB2-Echeneis_naucrates\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      " A0A667GS13-A0A667GS13_LYNCA-AGRA3-Lynx_canadensisNone\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 74(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WT11-A0A665WT11_ECHNA-AGRB2-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A2I3N692-A0A2I3N692_PAPAN-AGRB2-Papio_anubis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A667GS13-A0A667GS13_LYNCA-AGRA3-Lynx_canadensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3N692-A0A2I3N692_PAPAN-AGRB2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      " DEBUG: minus_one_residue = None, start = 587None\n",
      "ALTERNATIVE GPS DETECTION! A0A2I0M863-A0A2I0M863_COLLI-AGRD1-Columba_livia DEBUG: minus_one_residue = 321, start = 411\n",
      "DEBUG:last_sheets = [(771, 777), (780, 785)]\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "DEBUG: self.indices = [191, 192]\n",
      "DEBUG: minus_one_residue = 322, start = 465Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 77(new variant)KeyNotFound: 310Helix\n",
      "\n",
      "DEBUG: self.indices = [321, 322, 323]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2I0M863-A0A2I0M863_COLLI-AGRD1-Columba_livia, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A674HXI4-A0A674HXI4_TERCA-AGRA2-Terrapene_carolina_triunguis\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 78(new variant)NoneNone  A0A665WTK0-A0A665WTK0_ECHNA-AGRB1b-Echeneis_naucratesA0A2I3N7Q9-A0A2I3N7Q9_PAPAN-AGRL2-Papio_anubis\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A674HXI4-A0A674HXI4_TERCA-AGRA2-Terrapene_carolina_triunguis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 308, start = 255\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [307, 308, 309]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)DEBUG: Getting alignment indices with: A0A2I3N7Q9-A0A2I3N7Q9_PAPAN-AGRL2-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A665WTK0-A0A665WTK0_ECHNA-AGRB1b-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 314, start = 418\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [313, 314, 315]None\n",
      " A0A2R9CTG3-A0A2R9CTG3_PANPA-AGRG2-Pan_paniscus\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 274, start = 513 DEBUG: self.indices = [292, 293, 294]Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 47(new variant)None\n",
      "\n",
      "DEBUG: self.indices = [273, 274, 275]\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A2R9CTG3-A0A2R9CTG3_PANPA-AGRG2-Pan_paniscus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6J1YCW3-A0A6J1YCW3_ACIJB-AGRA3-Acinonyx_jubatusCompleted Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant) None\n",
      " A0A665WX53-A0A665WX53_ECHNA-AGRL3-like-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A6J1YCW3-A0A6J1YCW3_ACIJB-AGRA3-Acinonyx_jubatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 240, start = 323\n",
      "DEBUG: self.indices = [239, 240, 241]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665WX53-A0A665WX53_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 321, start = 411 \n",
      "A0A2U4CJB9-A0A2U4CJB9_TURTR-AGRF1-Tursiops_truncatus\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2U4CJB9-A0A2U4CJB9_TURTR-AGRF1-Tursiops_truncatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6J2VTK7-A0A6J2VTK7_CHACN-AGRA3-Chanos_chanos\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 265, start = 298 Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 81(new variant)\n",
      "\n",
      "A0A665WXI7-A0A665WXI7_ECHNA-AGRL3-like-Echeneis_naucratesDEBUG: self.indices = [264, 265, 266]\n",
      "DEBUG: Getting alignment indices with: A0A6J2VTK7-A0A6J2VTK7_CHACN-AGRA3-Chanos_chanos, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WXI7-A0A665WXI7_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q1LLW7-A0A3Q1LLW7_BOVIN-AGRA3-Bos_taurus\n",
      "DEBUG: minus_one_residue = 316, start = 416\n",
      "DEBUG: self.indices = [315, 316, 317]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 77(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3Q1LLW7-A0A3Q1LLW7_BOVIN-AGRA3-Bos_taurus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4WE88-A0A6P4WE88_PANPR-AGRA3isoformX1-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 319, start = 244\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [318, 319, 320]\n",
      "\n",
      " NoneDEBUG: Getting alignment indices with: A0A6P4WE88-A0A6P4WE88_PANPR-AGRA3isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A665WXV0-A0A665WXV0_ECHNA-AGRL3-like-Echeneis_naucrates\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A665WXV0-A0A665WXV0_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None A0A452ECV5-A0A452ECV5_CAPHI-AGRG2-Capra_hircus\n",
      "DEBUG: minus_one_residue = 321, start = 411\n",
      "DEBUG: self.indices = [320, 321, 322]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 51(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A452ECV5-A0A452ECV5_CAPHI-AGRG2-Capra_hircus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 384, start = 179A0A6P7LQB6-A0A6P7LQB6_BETSP-AGRF3-like-Betta_splendens\n",
      "\n",
      "DEBUG: self.indices = [383, 384, 385]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 80(new variant) \n",
      "A0A665WXX1-A0A665WXX1_ECHNA-AGRL3-like-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A6P7LQB6-A0A6P7LQB6_BETSP-AGRF3-like-Betta_splendens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WXX1-A0A665WXX1_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A5F4DAK5-A0A5F4DAK5_CANLF-AGRL4-Canis_lupus_familiaris\n",
      "DEBUG: minus_one_residue = 244, start = 488\n",
      "DEBUG: self.indices = [243, 244, 245]Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: Getting alignment indices with: A0A5F4DAK5-A0A5F4DAK5_CANLF-AGRL4-Canis_lupus_familiaris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5JC41-A0A7K5JC41_TOXRE-AGRA3protein-Toxostoma_redivivum\n",
      "DEBUG: minus_one_residue = 295, start = 268\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant) DEBUG: self.indices = [294, 295, 296]\n",
      "None \n",
      "A0A665WYK5-A0A665WYK5_ECHNA-AGRL3-like-Echeneis_naucratesDEBUG: Getting alignment indices with: A0A7K5JC41-A0A7K5JC41_TOXRE-AGRA3protein-Toxostoma_redivivum, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)None\n",
      " DEBUG: Getting alignment indices with: A0A665WYK5-A0A665WYK5_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A667Z4F7-A0A667Z4F7_9TELE-AGRA3-Myripristis_murdjan\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 77(new variant)DEBUG: minus_one_residue = 319, start = 413\n",
      "\n",
      "DEBUG: self.indices = [318, 319, 320]\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: Getting alignment indices with: A0A667Z4F7-A0A667Z4F7_9TELE-AGRA3-Myripristis_murdjan, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 304, start = 264 A0A7L1NT66-A0A7L1NT66_RHICY-AGRA3protein-Rhinopomastus_cyanomelas\n",
      "DEBUG: self.indices = [303, 304, 305]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A665WYN4-A0A665WYN4_ECHNA-AGRL3-like-Echeneis_naucratesFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      " NoneDEBUG: Getting alignment indices with: A0A7L1NT66-A0A7L1NT66_RHICY-AGRA3protein-Rhinopomastus_cyanomelas, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "A0A6P5R921-A0A6P5R921_MUSCR-AGRF1-Mus_caroli\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P5R921-A0A6P5R921_MUSCR-AGRF1-Mus_caroli, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WYN4-A0A665WYN4_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 319, start = 413\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: self.indices = [318, 319, 320]\n",
      "DEBUG: minus_one_residue = 264, start = 299\n",
      "DEBUG: self.indices = [263, 264, 265]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [306, 307, 308] \n",
      "None A0A834CLB7-A0A834CLB7_ORYME-GR124-Oryzias_melastigma\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 78(new variant)None\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A834CLB7-A0A834CLB7_ORYME-GR124-Oryzias_melastigma, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "  None A0A665WYX8-A0A665WYX8_ECHNA-AGRL3-like-Echeneis_naucrates\n",
      "A0A6P8ET89-A0A6P8ET89_CLUHA-AGRE1-likeisoformX1-Clupea_harengus\n",
      "Found Helix boundary with the following characteristics: maxk = 14 helical_counts[maxk] = 48(new variant)DEBUG: minus_one_residue = 309, start = 423\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8ET89-A0A6P8ET89_CLUHA-AGRE1-likeisoformX1-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A665WYX8-A0A665WYX8_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [308, 309, 310]\n",
      "\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 230, start = 333 \n",
      "NoneDEBUG: self.indices = [229, 230, 231] W5NRU2-W5NRU2_SHEEP-AGRA2-Ovis_aries\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: W5NRU2-W5NRU2_SHEEP-AGRA2-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "\n",
      " None A0A6P8F8F4-A0A6P8F8F4_CLUHA-AGRE1-likeisoformX2-Clupea_harengus\n",
      "Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 50(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8F8F4-A0A6P8F8F4_CLUHA-AGRE1-likeisoformX2-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A665WYZ1-A0A665WYZ1_ECHNA-AGRL3-like-Echeneis_naucratesDEBUG: minus_one_residue = 323, start = 409\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665WYZ1-A0A665WYZ1_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 230, start = 333\n",
      "DEBUG: self.indices = [229, 230, 231]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 307, start = 480DEBUG: self.indices = [322, 323, 324]None\n",
      " DEBUG: self.indices = [306, 307, 308]\n",
      "A0A7J7WU03-A0A7J7WU03_MYOMY-AGRF1-Myotis_myotis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "\n",
      "Initializing GainDomain Object.\n",
      "None DEBUG: Getting alignment indices with: A0A7J7WU03-A0A7J7WU03_MYOMY-AGRF1-Myotis_myotis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> None\n",
      "A0A665WZH4-A0A665WZH4_ECHNA-AGRL3-like-Echeneis_naucrates \n",
      "A0A287ASV0-A0A287ASV0_PIG-AGRA3-Sus_scrofa\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)DEBUG: Getting alignment indices with: A0A665WZH4-A0A665WZH4_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A287ASV0-A0A287ASV0_PIG-AGRA3-Sus_scrofa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 265, start = 298\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: minus_one_residue = 321, start = 412DEBUG: self.indices = [305, 306, 307] \n",
      "None\n",
      " DEBUG: self.indices = [320, 321, 322]A0A7J7XDG0-A0A7J7XDG0_MYOMY-AGRE1-Myotis_myotis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 65(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7J7XDG0-A0A7J7XDG0_MYOMY-AGRE1-Myotis_myotis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None None  A0A665WZI2-A0A665WZI2_ECHNA-AGRL3-like-Echeneis_naucratesA0A3B4DFA6-A0A3B4DFA6_PYGNA-AGRF3-like-Pygocentrus_nattereri\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 78(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B4DFA6-A0A3B4DFA6_PYGNA-AGRF3-like-Pygocentrus_nattereri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "DEBUG: minus_one_residue = 261, start = 302DEBUG: Getting alignment indices with: A0A665WZI2-A0A665WZI2_ECHNA-AGRL3-like-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 248, start = 485 DEBUG: minus_one_residue = 307, start = 480None\n",
      "\n",
      " DEBUG: self.indices = [247, 248, 249]A0A7J7YKQ0-A0A7J7YKQ0_PIPKU-AGRF1-Pipistrellus_kuhlii\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7J7YKQ0-A0A7J7YKQ0_PIPKU-AGRF1-Pipistrellus_kuhlii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None  NoneA0A452TV52-A0A452TV52_URSMA-AGRB2-Ursus_maritimus \n",
      "A0A665XDX0-A0A665XDX0_ECHNA-AGRL3-Echeneis_naucratesDEBUG: minus_one_residue = 265, start = 298\n",
      "DEBUG: self.indices = [264, 265, 266]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A452TV52-A0A452TV52_URSMA-AGRB2-Ursus_maritimus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665XDX0-A0A665XDX0_ECHNA-AGRL3-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J7YKQ3-A0A7J7YKQ3_PIPKU-AGRF1-Pipistrellus_kuhliiDEBUG: minus_one_residue = 322, start = 411\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)DEBUG: self.indices = [321, 322, 323]\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7J7YKQ3-A0A7J7YKQ3_PIPKU-AGRF1-Pipistrellus_kuhlii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      " A0A665XE28-A0A665XE28_ECHNA-AGRL3-Echeneis_naucratesNone \n",
      "A0A4X1V6V0-A0A4X1V6V0_PIG-AGRA3-Sus_scrofa\n",
      "DEBUG: minus_one_residue = 265, start = 298Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: self.indices = [264, 265, 266]DEBUG: Getting alignment indices with: A0A4X1V6V0-A0A4X1V6V0_PIG-AGRA3-Sus_scrofa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665XE28-A0A665XE28_ECHNA-AGRL3-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A803SS41-A0A803SS41_ANOCA-AGRB3-Anolis_carolinensis\n",
      "DEBUG: minus_one_residue = 321, start = 412\n",
      "DEBUG: self.indices = [320, 321, 322]DEBUG: minus_one_residue = 306, start = 481\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A803SS41-A0A803SS41_ANOCA-AGRB3-Anolis_carolinensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None\n",
      "Initializing GainDomain Object.\n",
      "  A0A6J0CFM3-A0A6J0CFM3_PERMB-AGRL1isoformX5-Peromyscus_maniculatus_bairdiiNone\n",
      " A0A665XEV3-A0A665XEV3_ECHNA-AGRL3-Echeneis_naucrates\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6J0CFM3-A0A6J0CFM3_PERMB-AGRL1isoformX5-Peromyscus_maniculatus_bairdii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 263, start = 300\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665XEV3-A0A665XEV3_ECHNA-AGRL3-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 308, start = 425\n",
      "NoneDEBUG: self.indices = [307, 308, 309] \n",
      "A0A803U194-A0A803U194_ANOCA-AGRB3-Anolis_carolinensis\n",
      "DEBUG: minus_one_residue = 293, start = 494Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "DEBUG: self.indices = [292, 293, 294]DEBUG: Getting alignment indices with: A0A803U194-A0A803U194_ANOCA-AGRB3-Anolis_carolinensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6J2V7Z2-A0A6J2V7Z2_CHACN-AGRA1isoformX1-Chanos_chanos\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 77(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: Getting alignment indices with: A0A6J2V7Z2-A0A6J2V7Z2_CHACN-AGRA1isoformX1-Chanos_chanos, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 300None\n",
      " DEBUG: self.indices = [262, 263, 264]A0A665XFI4-A0A665XFI4_ECHNA-AGRL3-Echeneis_naucrates\n",
      "\n",
      "DEBUG: minus_one_residue = 324, start = 409\n",
      "DEBUG: self.indices = [323, 324, 325]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A665XFI4-A0A665XFI4_ECHNA-AGRL3-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "None  NoneA0A833ZFH9-A0A833ZFH9_9CHIR-AGRG2-Phyllostomus_discolor \n",
      "A0A6P4YKK5-A0A6P4YKK5_BRABE-AGRG7-like-Branchiostoma_belcheriDEBUG: minus_one_residue = 306, start = 481\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 47(new variant)\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 69(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A833ZFH9-A0A833ZFH9_9CHIR-AGRG2-Phyllostomus_discolor, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P4YKK5-A0A6P4YKK5_BRABE-AGRG7-like-Branchiostoma_belcheri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " \n",
      "None A0A665XFW7-A0A665XFW7_ECHNA-AGRL3-Echeneis_naucrates\n",
      "DEBUG: minus_one_residue = 240, start = 323\n",
      "DEBUG: minus_one_residue = 291, start = 442Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)DEBUG: self.indices = [239, 240, 241]\n",
      "\n",
      "DEBUG: self.indices = [290, 291, 292]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      "F1P9I0-F1P9I0_CANLF-AGRL4-Canis_lupus_familiarisDEBUG: Getting alignment indices with: A0A665XFW7-A0A665XFW7_ECHNA-AGRL3-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      " None Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 95(new variant)A0A6P7HB06-A0A6P7HB06_9TELE-AGRG6isoformX6-Parambassis_ranga\n",
      "\n",
      "DEBUG: Getting alignment indices with: F1P9I0-F1P9I0_CANLF-AGRL4-Canis_lupus_familiaris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 64(new variant)\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: Getting alignment indices with: A0A6P7HB06-A0A6P7HB06_9TELE-AGRG6isoformX6-Parambassis_ranga, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 295, start = 268\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [294, 295, 296] \n",
      "None \n",
      "Initializing GainDomain Object.\n",
      "A0A667FK65-A0A667FK65_LYNCA-AGRL3-Lynx_canadensisDEBUG: minus_one_residue = 263, start = 470 \n",
      "\n",
      "None Q8VEC3_B0V2Q5_Q8BMV8-AGRF1_MOUSE-AGRF1-Mus_musculus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A667FK65-A0A667FK65_LYNCA-AGRL3-Lynx_canadensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: Q8VEC3_B0V2Q5_Q8BMV8-AGRF1_MOUSE-AGRF1-Mus_musculus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 264, start = 299None\n",
      "\n",
      " DEBUG: self.indices = [263, 264, 265]DEBUG: self.indices = [306, 307, 308]\n",
      "A0A6P7R5E3-A0A6P7R5E3_MUSCR-AGRL1isoformX4-Mus_caroli\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 104(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "  DEBUG: Getting alignment indices with: A0A6P7R5E3-A0A6P7R5E3_MUSCR-AGRL1isoformX4-Mus_caroli, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>NoneNone\n",
      "  A0A667FRI6-A0A667FRI6_LYNCA-AGRL3-Lynx_canadensisA0A096NIW6-A0A096NIW6_PAPAN-AGRF1-Papio_anubis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A096NIW6-A0A096NIW6_PAPAN-AGRF1-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: minus_one_residue = 308, start = 425DEBUG: Getting alignment indices with: A0A667FRI6-A0A667FRI6_LYNCA-AGRL3-Lynx_canadensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 266, start = 298\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "DEBUG: self.indices = [307, 308, 309]DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      "A0A0D9RJL6-A0A0D9RJL6_CHLSB-AGRF1-Chlorocebus_sabaeus \n",
      "None\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant) \n",
      " G3TN46-G3TN46_LOXAF-AGRA2-Loxodonta_africanaDEBUG: Getting alignment indices with: A0A0D9RJL6-A0A0D9RJL6_CHLSB-AGRF1-Chlorocebus_sabaeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: G3TN46-G3TN46_LOXAF-AGRA2-Loxodonta_africana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A667FRL8-A0A667FRL8_LYNCA-AGRL3-Lynx_canadensis\n",
      "DEBUG: minus_one_residue = 266, start = 298\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A667FRL8-A0A667FRL8_LYNCA-AGRL3-Lynx_canadensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A0G2KPF7-A0A0G2KPF7_DANRE-AGRF3b-Danio_rerio\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 78(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: Getting alignment indices with: A0A0G2KPF7-A0A0G2KPF7_DANRE-AGRF3b-Danio_rerio, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 327, start = 406\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: self.indices = [326, 327, 328]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A667FRM6-A0A667FRM6_LYNCA-AGRL3-Lynx_canadensis\n",
      "DEBUG: minus_one_residue = 242, start = 322\n",
      "DEBUG: self.indices = [241, 242, 243] Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)None\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A667FRM6-A0A667FRM6_LYNCA-AGRL3-Lynx_canadensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A096N1W0-A0A096N1W0_PAPAN-AGRF3-Papio_anubis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 307, start = 480DEBUG: Getting alignment indices with: A0A096N1W0-A0A096N1W0_PAPAN-AGRF3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A1U7QEP9-A0A1U7QEP9_MESAU-AGRF1isoformX4-Mesocricetus_auratus\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "73 of which have detected consensus GPS motifs.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: minus_one_residue = 272, start = 462\n",
      "DEBUG: self.indices = [271, 272, 273]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A096NZJ3-A0A096NZJ3_PAPAN-AGRA3-Papio_anubisDEBUG: Getting alignment indices with: A0A1U7QEP9-A0A1U7QEP9_MESAU-AGRF1isoformX4-Mesocricetus_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)DEBUG: minus_one_residue = 265, start = 299\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A096NZJ3-A0A096NZJ3_PAPAN-AGRA3-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [264, 265, 266]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5BXN5-A0A2K5BXN5_AOTNA-AGRF1-Aotus_nancymaae\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K5BXN5-A0A2K5BXN5_AOTNA-AGRF1-Aotus_nancymaae, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5NRC8-A0A2K5NRC8_CERAT-AGRA3-Cercocebus_atys\n",
      "DEBUG: minus_one_residue = 265, start = 299\n",
      "DEBUG: self.indices = [264, 265, 266]Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2K5NRC8-A0A2K5NRC8_CERAT-AGRA3-Cercocebus_atys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5I1B7-A0A2K5I1B7_COLAP-AGRF1-Colobus_angolensis_palliatus\n",
      "DEBUG: minus_one_residue = 321, start = 413Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K5I1B7-A0A2K5I1B7_COLAP-AGRF1-Colobus_angolensis_palliatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "DEBUG: minus_one_residue = 266, start = 298\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " A0A2K5RXR1-A0A2K5RXR1_CEBIM-AGRA3-Cebus_imitator None\n",
      " A0A2K5L3G2-A0A2K5L3G2_CERAT-AGRF1-Cercocebus_atys\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)DEBUG: Getting alignment indices with: A0A2K5RXR1-A0A2K5RXR1_CEBIM-AGRA3-Cebus_imitator, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A2K5L3G2-A0A2K5L3G2_CERAT-AGRF1-Cercocebus_atys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "DEBUG: minus_one_residue = 266, start = 298\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A2K5USY6-A0A2K5USY6_MACFA-AGRA3-Macaca_fascicularis\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      " DEBUG: Getting alignment indices with: A0A2K5USY6-A0A2K5USY6_MACFA-AGRA3-Macaca_fascicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A2K5UIJ0-A0A2K5UIJ0_MACFA-AGRF1-Macaca_fascicularis\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K5UIJ0-A0A2K5UIJ0_MACFA-AGRF1-Macaca_fascicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6BP17-A0A2K6BP17_MACNE-AGRA3-Macaca_nemestrinaDEBUG: minus_one_residue = 266, start = 298\n",
      "\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K6BP17-A0A2K6BP17_MACNE-AGRA3-Macaca_nemestrina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6BDM9-A0A2K6BDM9_MACNE-AGRF1-Macaca_nemestrina\n",
      "DEBUG: minus_one_residue = 321, start = 413Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "\n",
      "DEBUG: self.indices = [320, 321, 322]DEBUG: Getting alignment indices with: A0A2K6BDM9-A0A2K6BDM9_MACNE-AGRF1-Macaca_nemestrina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 266, start = 298 \n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "A0A369S286-A0A369S286_9METZ-AGRL3-Trichoplax_sp._H2.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 76(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A369S286-A0A369S286_9METZ-AGRL3-Trichoplax_sp._H2., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A452U9H0-A0A452U9H0_URSMA-AGRE1-Ursus_maritimus\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 56(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A452U9H0-A0A452U9H0_URSMA-AGRE1-Ursus_maritimus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 429\n",
      "DEBUG: minus_one_residue = 239, start = 325DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "DEBUG: self.indices = [238, 239, 240]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B4THH9-A0A3B4THH9_SERDU-AGRL2-Seriola_dumerili\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B4THH9-A0A3B4THH9_SERDU-AGRL2-Seriola_dumerili, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6J0TQW0-A0A6J0TQW0_9SAUR-AGRA1isoformX1-Pogona_vitticeps\n",
      "DEBUG: minus_one_residue = 291, start = 443Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 80(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6J0TQW0-A0A6J0TQW0_9SAUR-AGRA1isoformX1-Pogona_vitticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [290, 291, 292]\n",
      "DEBUG: minus_one_residue = 317, start = 247\n",
      "DEBUG: self.indices = [316, 317, 318]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A452RQP8-A0A452RQP8_URSAM-AGRA3-Ursus_americanus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)A0A6P3PVE7-A0A6P3PVE7_PTEVA-AGRF1-Pteropus_vampyrus\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A452RQP8-A0A452RQP8_URSAM-AGRA3-Ursus_americanus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P3PVE7-A0A6P3PVE7_PTEVA-AGRF1-Pteropus_vampyrus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "DEBUG: minus_one_residue = 265, start = 299\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      " None A0A4W5R2S6-A0A4W5R2S6_9TELE-Uncharacterizedprotein-Hucho_hucho\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8F5J1-A0A6P8F5J1_CLUHA-AGRE1-likeisoformX1-Clupea_harengusFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 70(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A4W5R2S6-A0A4W5R2S6_9TELE-Uncharacterizedprotein-Hucho_hucho, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 12 helical_counts[maxk] = 48(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8F5J1-A0A6P8F5J1_CLUHA-AGRE1-likeisoformX1-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 428\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 230, start = 334None\n",
      " DEBUG: self.indices = [229, 230, 231]A0A5F4VTP4-A0A5F4VTP4_CALJA-Uncharacterizedprotein-Callithrix_jacchus\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant) \n",
      "NoneDEBUG: Getting alignment indices with: A0A5F4VTP4-A0A5F4VTP4_CALJA-Uncharacterizedprotein-Callithrix_jacchus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " A0A6P8UHW7-A0A6P8UHW7_GYMAC-AGRE2-like-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 321, start = 413Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 54(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8UHW7-A0A6P8UHW7_GYMAC-AGRE2-like-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [320, 321, 322]DEBUG: minus_one_residue = 260, start = 304\n",
      "\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J8K4C1-A0A7J8K4C1_ROUAE-AGRF1-Rousettus_aegyptiacus\n",
      " A0A5F7ZLE4-A0A5F7ZLE4_MACMU-AGRA3-Macaca_mulatta\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7J8K4C1-A0A7J8K4C1_ROUAE-AGRF1-Rousettus_aegyptiacus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A5F7ZLE4-A0A5F7ZLE4_MACMU-AGRA3-Macaca_mulatta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 265, start = 299\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J8K4C5-A0A7J8K4C5_ROUAE-AGRF1-Rousettus_aegyptiacus\n",
      " None Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)A0A5N5NK28-A0A5N5NK28_PANHP-Uncharacterizedprotein-Pangasianodon_hypophthalmus\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7J8K4C5-A0A7J8K4C5_ROUAE-AGRF1-Rousettus_aegyptiacus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A5N5NK28-A0A5N5NK28_PANHP-Uncharacterizedprotein-Pangasianodon_hypophthalmus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 265, start = 299\n",
      "DEBUG: self.indices = [264, 265, 266]DEBUG: minus_one_residue = 325, start = 409\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: self.indices = [324, 325, 326]None\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "A0A7R5KYY4-A0A7R5KYY4_9PASS-AGRB2isoformX4-Pipra_filicauda\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 96(new variant)\n",
      " A0A663DKD0-A0A663DKD0_AQUCH-AGRA3-Aquila_chrysaetos_chrysaetos.\n",
      "DEBUG: Getting alignment indices with: A0A7R5KYY4-A0A7R5KYY4_9PASS-AGRB2isoformX4-Pipra_filicauda, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A663DKD0-A0A663DKD0_AQUCH-AGRA3-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 286, start = 278DEBUG: minus_one_residue = 319, start = 420\n",
      "\n",
      "DEBUG: self.indices = [285, 286, 287]DEBUG: self.indices = [318, 319, 320]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None  None A0A669ESE9-A0A669ESE9_ORENI-Uncharacterizedprotein-Oreochromis_niloticusF6U5Y3_A0A1D5RCA4-F6U5Y3_MACMU-AGRF1-Macaca_mulatta\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 78(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 90(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A669ESE9-A0A669ESE9_ORENI-Uncharacterizedprotein-Oreochromis_niloticus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: F6U5Y3_A0A1D5RCA4-F6U5Y3_MACMU-AGRF1-Macaca_mulatta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 310, start = 424\n",
      "DEBUG: self.indices = [309, 310, 311]\n",
      "DEBUG: minus_one_residue = 266, start = 298\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [265, 266, 267] None\n",
      " A0A670J6M3-A0A670J6M3_PODMU-Uncharacterizedprotein-Podarcis_muralis\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 81(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A670J6M3-A0A670J6M3_PODMU-Uncharacterizedprotein-Podarcis_muralis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None H3DJM8-H3DJM8_TETNG-AGRB1b-Tetraodon_nigroviridis\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: H3DJM8-H3DJM8_TETNG-AGRB1b-Tetraodon_nigroviridis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 319, start = 415\n",
      "DEBUG: self.indices = [318, 319, 320]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 288, start = 276 \n",
      "None DEBUG: self.indices = [287, 288, 289]A0A673HI84-A0A673HI84_9TELE-AGRL2-like-Sinocyclocheilus_rhinocerous.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 67(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A673HI84-A0A673HI84_9TELE-AGRL2-like-Sinocyclocheilus_rhinocerous., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A087YHC6-A0A087YHC6_POEFO-AGRD1-Poecilia_formosa\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 51(new variant)\n",
      "DEBUG: minus_one_residue = 253, start = 481DEBUG: Getting alignment indices with: A0A087YHC6-A0A087YHC6_POEFO-AGRD1-Poecilia_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [252, 253, 254]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 305, start = 260 \n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "None A0A673HIZ6-A0A673HIZ6_9TELE-AGRL2-like-Sinocyclocheilus_rhinocerous.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 67(new variant) A0A1U7TJT2-A0A1U7TJT2_CARSF-AGRF1isoformX1-Carlito_syrichta\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A673HIZ6-A0A673HIZ6_9TELE-AGRL2-like-Sinocyclocheilus_rhinocerous., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 89(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A1U7TJT2-A0A1U7TJT2_CARSF-AGRF1isoformX1-Carlito_syrichta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 266, start = 299\n",
      "DEBUG: self.indices = [265, 266, 267]DEBUG: minus_one_residue = 253, start = 481\n",
      "\n",
      "DEBUG: self.indices = [252, 253, 254]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I3TLH8-A0A2I3TLH8_PANTR-AGRF1-Pan_troglodytes\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant) \n",
      "DEBUG: Getting alignment indices with: A0A2I3TLH8-A0A2I3TLH8_PANTR-AGRF1-Pan_troglodytes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "A0A6G1R1B7-A0A6G1R1B7_9TELE-AGRA3-Channa_argusFound Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 78(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6G1R1B7-A0A6G1R1B7_9TELE-AGRA3-Channa_argus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 265, start = 300\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 327, start = 407\n",
      "DEBUG: self.indices = [326, 327, 328]\n",
      "None\n",
      "Initializing GainDomain Object.\n",
      "  None A0A6J3HH97-A0A6J3HH97_SAPAP-AGRA3-Sapajus_apella\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "A0A2K5RWZ3-A0A2K5RWZ3_CEBIM-AGRF1-Cebus_imitatorDEBUG: Getting alignment indices with: A0A6J3HH97-A0A6J3HH97_SAPAP-AGRA3-Sapajus_apella, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K5RWZ3-A0A2K5RWZ3_CEBIM-AGRF1-Cebus_imitator, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413DEBUG: minus_one_residue = 266, start = 299\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [265, 266, 267] \n",
      "None A0A6P9AX19-A0A6P9AX19_PANGU-AGRL1isoformX6-Pantherophis_guttatus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 105(new variant) None \n",
      "DEBUG: Getting alignment indices with: A0A6P9AX19-A0A6P9AX19_PANGU-AGRL1isoformX6-Pantherophis_guttatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A2K6V1S3-A0A2K6V1S3_SAIBB-AGRF1-Saimiri_boliviensis_boliviensis\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K6V1S3-A0A2K6V1S3_SAIBB-AGRF1-Saimiri_boliviensis_boliviensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 429\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "DEBUG: minus_one_residue = 266, start = 299\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834BE36-A0A834BE36_9CHIR-AGRA3-Phyllostomus_discolor\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834BE36-A0A834BE36_9CHIR-AGRA3-Phyllostomus_discolor, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None U3B0D2-U3B0D2_CALJA-PutativeGR125-Callithrix_jacchus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: U3B0D2-U3B0D2_CALJA-PutativeGR125-Callithrix_jacchus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None U3E886-U3E886_CALJA-PutativeGR125-Callithrix_jacchus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: U3E886-U3E886_CALJA-PutativeGR125-Callithrix_jacchus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 413\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I4C057-A0A2I4C057_9TELE-GR126isoformX5-Austrofundulus_limnaeus.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I4C057-A0A2I4C057_9TELE-GR126isoformX5-Austrofundulus_limnaeus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 479\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2Y9JM20-A0A2Y9JM20_ENHLU-AGRA3isoformX1-Enhydra_lutris_kenyoni.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2Y9JM20-A0A2Y9JM20_ENHLU-AGRA3isoformX1-Enhydra_lutris_kenyoni., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A452CC04-A0A452CC04_BALAS-AGRL2isoformX4-Balaenoptera_acutorostrata_scammoni\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A452CC04-A0A452CC04_BALAS-AGRL2isoformX4-Balaenoptera_acutorostrata_scammoni, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 220, start = 515\n",
      "DEBUG: self.indices = [219, 220, 221]\n",
      "KeyNotFound: 310Helix\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A452SEU5-A0A452SEU5_URSAM-AGRA2-Ursus_americanus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A452SEU5-A0A452SEU5_URSAM-AGRA2-Ursus_americanus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 327, start = 408\n",
      "DEBUG: self.indices = [326, 327, 328]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A452SEW5-A0A452SEW5_URSAM-AGRA2-Ursus_americanus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A452SEW5-A0A452SEW5_URSAM-AGRA2-Ursus_americanus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 327, start = 408\n",
      "DEBUG: self.indices = [326, 327, 328]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6D2Y2J1-A0A6D2Y2J1_PANTR-ADGRA3isoform1-Pan_troglodytes\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6D2Y2J1-A0A6D2Y2J1_PANTR-ADGRA3isoform1-Pan_troglodytes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6I8PDB5-A0A6I8PDB5_ORNAN-AGRA3-Ornithorhynchus_anatinus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6I8PDB5-A0A6I8PDB5_ORNAN-AGRA3-Ornithorhynchus_anatinus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6I9JXI8-A0A6I9JXI8_CHRAS-probableGR125-Chrysochloris_asiatica\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6I9JXI8-A0A6I9JXI8_CHRAS-probableGR125-Chrysochloris_asiatica, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J5ZIH0-A0A7J5ZIH0_AMEME-Uncharacterizedprotein-Ameiurus_melas\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7J5ZIH0-A0A7J5ZIH0_AMEME-Uncharacterizedprotein-Ameiurus_melas, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 325, start = 410\n",
      "DEBUG: self.indices = [324, 325, 326]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None F6WDX1-F6WDX1_ORNAN-AGRA3-Ornithorhynchus_anatinus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: F6WDX1-F6WDX1_ORNAN-AGRA3-Ornithorhynchus_anatinus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None G1S489-G1S489_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: G1S489-G1S489_NOMLE-Uncharacterizedprotein-Nomascus_leucogenys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "Initializing GainDomain Object.\n",
      " None H2QPA1_A0A2J8NW34-H2QPA1_PANTR-AGRA3-Pan_troglodytes\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: H2QPA1_A0A2J8NW34-H2QPA1_PANTR-AGRA3-Pan_troglodytes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 414\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A151MG56-A0A151MG56_ALLMI-AGRD2isoformA-Alligator_mississippiensis\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A151MG56-A0A151MG56_ALLMI-AGRD2isoformA-Alligator_mississippiensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 329, start = 407\n",
      "DEBUG: self.indices = [328, 329, 330]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A151MGI0-A0A151MGI0_ALLMI-AGRD2isoformB-Alligator_mississippiensis\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A151MGI0-A0A151MGI0_ALLMI-AGRD2isoformB-Alligator_mississippiensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 329, start = 407\n",
      "DEBUG: self.indices = [328, 329, 330]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I2ZBI9-A0A2I2ZBI9_GORGO-AGRA3-Gorilla_gorilla_gorilla\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I2ZBI9-A0A2I2ZBI9_GORGO-AGRA3-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 415\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "Found sequences object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A0G2JDA9-A0A0G2JDA9_MOUSE-AG-protein-coupledreceptorG5-Mus_musculus\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A1U8CSX2-A0A1U8CSX2_MESAU-AGRE1-like-Mesocricetus_auratus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A1U8CSX2-A0A1U8CSX2_MESAU-AGRE1-like-Mesocricetus_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 75, start = 18\n",
      "DEBUG: self.indices = [74, 75, 76]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B4DJ68-A0A3B4DJ68_PYGNA-AGRE3-like-Pygocentrus_nattereri\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3B4DJ68-A0A3B4DJ68_PYGNA-AGRE3-like-Pygocentrus_nattereri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 71, start = 22\n",
      "DEBUG: self.indices = [70, 71, 72]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6J8EQQ9-A0A6J8EQQ9_MYTCO-Coadhesin,Thrombospondin-1,Mucin-likeprotein,Hemicentin-1,AGRB3,Thrombospondin-2,Adisintegrinandmetalloproteinasewiththrombospondinmotifsadt-1-Mytilus_coruscus\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[10]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6J8EQQ9-A0A6J8EQQ9_MYTCO-Coadhesin,Thrombospondin-1,Mucin-likeprotein,Hemicentin-1,AGRB3,Thrombospondin-2,Adisintegrinandmetalloproteinasewiththrombospondinmotifsadt-1-Mytilus_coruscus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 25\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(46, 50), (89, 91)]\n",
      "DEBUG: self.indices = [26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q4I9B6-A0A3Q4I9B6_NEOBR-AGRA3-Neolamprologus_brichardi\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3Q4I9B6-A0A3Q4I9B6_NEOBR-AGRA3-Neolamprologus_brichardi, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 71, start = 23\n",
      "DEBUG: self.indices = [70, 71, 72]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A4W6BP11-A0A4W6BP11_LATCA-AGRA1-Lates_calcarifer\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A4W6BP11-A0A4W6BP11_LATCA-AGRA1-Lates_calcarifer, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 90, start = 4\n",
      "DEBUG: self.indices = [89, 90, 91]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None H3BSN7-H3BSN7_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[18]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BSN7-H3BSN7_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 30\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(86, 95), (100, 106)]\n",
      "DEBUG: self.indices = [66, 67, 68, 69]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2Y9M2G1-A0A2Y9M2G1_DELLE-AGRG3isoformX4-Delphinapterus_leucas\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2Y9M2G1-A0A2Y9M2G1_DELLE-AGRG3isoformX4-Delphinapterus_leucas, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 93, start = 2\n",
      "DEBUG: self.indices = [92, 93, 94]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None F1QX89-F1QX89_DANRE-AGRA3-Danio_rerio\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: F1QX89-F1QX89_DANRE-AGRA3-Danio_rerio, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 73, start = 22\n",
      "DEBUG: self.indices = [72, 73, 74]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A833ZDY5-A0A833ZDY5_9CHIR-AGRE1-Phyllostomus_discolor\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A833ZDY5-A0A833ZDY5_9CHIR-AGRE1-Phyllostomus_discolor, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 75, start = 21\n",
      "DEBUG: self.indices = [74, 75, 76]\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " None W5MFY5-W5MFY5_LEPOC-AGRG3-Lepisosteus_oculatus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: W5MFY5-W5MFY5_LEPOC-AGRG3-Lepisosteus_oculatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 86, start = 10\n",
      "DEBUG: self.indices = [85, 86, 87]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7N9IHT8-A0A7N9IHT8_MACFA-AGRG6-Macaca_fascicularis\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[11]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A7N9IHT8-A0A7N9IHT8_MACFA-AGRG6-Macaca_fascicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 41\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(98, 101), (106, 109)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None H3BV72-H3BV72_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[17]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BV72-H3BV72_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 109)]\n",
      "DEBUG: self.indices = [74, 75]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5JIL6-A0A2K5JIL6_COLAP-AGRG6-Colobus_angolensis_palliatus\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[12,  0,  0]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2K5JIL6-A0A2K5JIL6_COLAP-AGRG6-Colobus_angolensis_palliatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K5KXF8-A0A2K5KXF8_CERAT-AGRG6-Cercocebus_atys\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[13,  0,  0]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2K5KXF8-A0A2K5KXF8_CERAT-AGRG6-Cercocebus_atys, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6D7J8-A0A2K6D7J8_MACNE-AGRG6-Macaca_nemestrina\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[11]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2K6D7J8-A0A2K6D7J8_MACNE-AGRG6-Macaca_nemestrina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "\n",
      "DEBUG: minus_one_residue = None, start = 42ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6P9X8-A0A2K6P9X8_RHIRO-AGRG6-Rhinopithecus_roxellana\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[12]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2K6P9X8-A0A2K6P9X8_RHIRO-AGRG6-Rhinopithecus_roxellana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B3RSG6-A0A3B3RSG6_9TELE-PutativeAGRE4P-Paramormyrops_kingsleyae.\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3B3RSG6-A0A3B3RSG6_9TELE-PutativeAGRE4P-Paramormyrops_kingsleyae., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 96, start = 2\n",
      "DEBUG: self.indices = [95, 96, 97]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A672TR66-A0A672TR66_STRHB-AGRG3-Strigops_habroptila\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A672TR66-A0A672TR66_STRHB-AGRG3-Strigops_habroptila, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 27\n",
      "ALTERNATIVE GPS DETECTION!DEBUG:last_sheets = [(83, 88), (91, 96)] \n",
      "DEBUG: self.indices = [62, 63]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6J0Z3U0-A0A6J0Z3U0_ODOVR-AGRE1-like-Odocoileus_virginianus_texanus.\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6J0Z3U0-A0A6J0Z3U0_ODOVR-AGRE1-like-Odocoileus_virginianus_texanus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 75, start = 23\n",
      "DEBUG: self.indices = [74, 75, 76]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J7YQ76-A0A7J7YQ76_RHIFE-AGRG6-Rhinolophus_ferrumequinum\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[13]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A7J7YQ76-A0A7J7YQ76_RHIFE-AGRG6-Rhinolophus_ferrumequinum, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J8K4C9-A0A7J8K4C9_ROUAE-AGRG6-Rousettus_aegyptiacus\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[13]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A7J8K4C9-A0A7J8K4C9_ROUAE-AGRG6-Rousettus_aegyptiacus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7N9D1B0-A0A7N9D1B0_MACFA-AGRG6-Macaca_fascicularis\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[11]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A7N9D1B0-A0A7N9D1B0_MACFA-AGRG6-Macaca_fascicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 42\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [61, 62, 63, 64]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A803W608-A0A803W608_FICAL-AGRG6-Ficedula_albicollis\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 32(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A803W608-A0A803W608_FICAL-AGRG6-Ficedula_albicollis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 0\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(99, 102), (107, 110)]\n",
      "DEBUG: self.indices = [103, 104, 105, 106]\n",
      "KeyNotFound: 310Helix\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " None H3BRA1-H3BRA1_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[17]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BRA1-H3BRA1_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 109)]\n",
      "DEBUG: self.indices = [74, 75]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A087WQZ5-A0A087WQZ5_MOUSE-AGRB1-Mus_musculus\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A1D5RLH8-A0A1D5RLH8_MOUSE-AG-protein-coupledreceptorG1-Mus_musculus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 20(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A1D5RLH8-A0A1D5RLH8_MOUSE-AG-protein-coupledreceptorG1-Mus_musculus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 0\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 111)]\n",
      "DEBUG: self.indices = [106, 107]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A5C6NHP8-A0A5C6NHP8_9TELE-AGRG3GR97-Takifugu_flavidus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A5C6NHP8-A0A5C6NHP8_9TELE-AGRG3GR97-Takifugu_flavidus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 73, start = 26\n",
      "DEBUG: self.indices = [72, 73, 74]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q0QTH0-A0A3Q0QTH0_AMPCI-AGRG6-Amphilophus_citrinellus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3Q0QTH0-A0A3Q0QTH0_AMPCI-AGRG6-Amphilophus_citrinellus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 38, start = 62\n",
      "DEBUG: self.indices = [37, 38, 39]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q1LMT9-A0A3Q1LMT9_BOVIN-AGRB3-Bos_taurus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3Q1LMT9-A0A3Q1LMT9_BOVIN-AGRB3-Bos_taurus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 36, start = 64\n",
      "DEBUG: self.indices = [35, 36, 37]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6G1RAR6-A0A6G1RAR6_9GRUI-AGRB1-Gallirallus_okinawae.\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4YJ71-A0A6P4YJ71_BRABE-AGRG6-like-Branchiostoma_belcheri\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6P4YJ71-A0A6P4YJ71_BRABE-AGRG6-like-Branchiostoma_belcheri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 97, start = 3\n",
      "DEBUG: self.indices = [96, 97, 98]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None H3BT88-H3BT88_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[17]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BT88-H3BT88_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 111)]\n",
      "DEBUG: self.indices = [74, 75]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6G3W9-A0A2K6G3W9_PROCO-AGRL4-Propithecus_coquereli\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.DEBUG: Getting alignment indices with: A0A2K6G3W9-A0A2K6G3W9_PROCO-AGRL4-Propithecus_coquereli, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 70, start = 31\n",
      "DEBUG: self.indices = [69, 70, 71]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A4W2GQE0-A0A4W2GQE0_BOBOX-AGRB3-Bos_indicus_x_Bos_taurus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A4W2GQE0-A0A4W2GQE0_BOBOX-AGRB3-Bos_indicus_x_Bos_taurus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 89, start = 12\n",
      "DEBUG: self.indices = [88, 89, 90]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6G1RC21-A0A6G1RC21_9GRUI-AGRV1-Gallirallus_okinawae.\n",
      "WARNING: No boundaries detected!\n",
      "No boundaries detected.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6G1RC21-A0A6G1RC21_9GRUI-AGRV1-Gallirallus_okinawae., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 2\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(41, 46), (49, 57)]\n",
      "DEBUG: self.indices = [45, 46]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6G1RX43-A0A6G1RX43_9GRUI-AGRD2-Gallirallus_okinawae.\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6J1U2G8-A0A6J1U2G8_9SAUR-AGRG6-Notechis_scutatus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 30(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6J1U2G8-A0A6J1U2G8_9SAUR-AGRG6-Notechis_scutatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 0\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(103, 106), (111, 114)]\n",
      "DEBUG: self.indices = [107, 108, 109, 110]\n",
      "KeyNotFound: 310Helix\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B3HHW8-A0A3B3HHW8_ORYLA-AGRG3-Oryzias_latipes\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3B3HHW8-A0A3B3HHW8_ORYLA-AGRG3-Oryzias_latipes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 92, start = 16\n",
      "DEBUG: self.indices = [91, 92, 93]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None H3BSB8-H3BSB8_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[18]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BSB8-H3BSB8_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 113)]\n",
      "DEBUG: self.indices = [74, 75]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B4T4G7-A0A3B4T4G7_SERDU-AGRE17-Seriola_dumerili\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3B4T4G7-A0A3B4T4G7_SERDU-AGRE17-Seriola_dumerili, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 93, start = 11\n",
      "DEBUG: self.indices = [92, 93, 94]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      " A0A3N0Y3Y1-A0A3N0Y3Y1_ANAGA-AGRG7-Anabarilius_grahami\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3N0Y3Y1-A0A3N0Y3Y1_ANAGA-AGRG7-Anabarilius_grahami, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 94, start = 10\n",
      "DEBUG: self.indices = [93, 94, 95]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A590UJ41-A0A590UJ41_HUMAN-AGRL4-Homo_sapiens\n",
      "WARNING: No boundaries detected!\n",
      "No boundaries detected.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A590UJ41-A0A590UJ41_HUMAN-AGRL4-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = None, start = 6\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(6, 8), (23, 25)]\n",
      "DEBUG: self.indices = [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6Q2XMT6-A0A6Q2XMT6_ESOLU-AGRG3-Esox_lucius\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 29(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6Q2XMT6-A0A6Q2XMT6_ESOLU-AGRG3-Esox_lucius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "Found sequences object.DEBUG: minus_one_residue = None, start = 0\n",
      "\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(96, 101), (104, 109)]\n",
      "DEBUG: self.indices = [102, 103]\n",
      "KeyNotFound: 310Helix\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P3Q6X8-A0A6P3Q6X8_PTEVA-AGRL2isoformX1-Pteropus_vampyrus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)A0A7F8QD28-A0A7F8QD28_LEPWE-AGRB3-like-Leptonychotes_weddellii\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P3Q6X8-A0A6P3Q6X8_PTEVA-AGRL2isoformX1-Pteropus_vampyrus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A1B0GX62-A0A1B0GX62_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "DEBUG: minus_one_residue = 293, start = 494Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[18]])(new variant)\n",
      "\n",
      "No Subdomain boundaries detected. Possible Fragment found.DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A1B0GX62-A0A1B0GX62_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P3RML6-A0A6P3RML6_PTEVA-AGRL1isoformX2-Pteropus_vampyrus\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 113)]\n",
      "DEBUG: self.indices = [74, 75]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P3RML6-A0A6P3RML6_PTEVA-AGRL1isoformX2-Pteropus_vampyrus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A1D5RM68-A0A1D5RM68_MOUSE-AG-protein-coupledreceptorG1-Mus_musculus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 20(new variant)\n",
      "DEBUG: minus_one_residue = 308, start = 479\n",
      "DEBUG: Getting alignment indices with: A0A1D5RM68-A0A1D5RM68_MOUSE-AG-protein-coupledreceptorG1-Mus_musculus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [307, 308, 309]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P3RPA5-A0A6P3RPA5_PTEVA-LOWQUALITYPROTEIN:AGRG4-Pteropus_vampyrus\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 0\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 105), (108, 113)]\n",
      "DEBUG: self.indices = [106, 107]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 63(new variant)None \n",
      "A0A2R8Q5E9-A0A2R8Q5E9_DANRE-AGRG4b-Danio_rerioDEBUG: Getting alignment indices with: A0A6P3RPA5-A0A6P3RPA5_PTEVA-LOWQUALITYPROTEIN:AGRG4-Pteropus_vampyrus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A2R8Q5E9-A0A2R8Q5E9_DANRE-AGRG4b-Danio_rerio, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 270, start = 517DEBUG: minus_one_residue = 99, start = 6\n",
      "DEBUG: self.indices = [98, 99, 100]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [269, 270, 271] \n",
      "None A0A493U0Q9-A0A493U0Q9_ANAPP-AGRG5-Anas_platyrhynchos_platyrhynchos\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A493U0Q9-A0A493U0Q9_ANAPP-AGRG5-Anas_platyrhynchos_platyrhynchos, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P3TJL3-A0A6P3TJL3_SHEEP-AGRG6isoformX2-Ovis_aries\n",
      "DEBUG: minus_one_residue = 103, start = 2Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: self.indices = [102, 103, 104]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P3TJL3-A0A6P3TJL3_SHEEP-AGRG6isoformX2-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " NoneDEBUG: minus_one_residue = 258, start = 529\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneA0A672K819-A0A672K819_SINGR-AGRE16-Sinocyclocheilus_grahami A0A6P3VI87-A0A6P3VI87_CLUHA-AGRB3isoformX4-Clupea_harengus\n",
      "\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A672K819-A0A672K819_SINGR-AGRE16-Sinocyclocheilus_grahami, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)DEBUG: minus_one_residue = 102, start = 3\n",
      "DEBUG: self.indices = [101, 102, 103]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P3VI87-A0A6P3VI87_CLUHA-AGRB3isoformX4-Clupea_harengus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 284, start = 503\n",
      "DEBUG: self.indices = [283, 284, 285]\n",
      "H3BRI2-H3BRI2_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 22(new variant) None \n",
      "DEBUG: Getting alignment indices with: H3BRI2-H3BRI2_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P3YI82-A0A6P3YI82_SHEEP-AGRL3-Ovis_aries\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 0DEBUG: Getting alignment indices with: A0A6P3YI82-A0A6P3YI82_SHEEP-AGRL3-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(105, 110), (113, 116)]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [111, 112]DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      " None NoneA0A6P3YKG2-A0A6P3YKG2_SHEEP-AGRG6isoformX4-Ovis_aries\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 57(new variant) H3BS54-H3BS54_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "DEBUG: Getting alignment indices with: A0A6P3YKG2-A0A6P3YKG2_SHEEP-AGRG6isoformX4-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q3GYJ6-A0A3Q3GYJ6_9LABR-AGRL3-like-Labrus_bergylta\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[7]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.DEBUG: minus_one_residue = 258, start = 529\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A3Q3GYJ6-A0A3Q3GYJ6_9LABR-AGRL3-like-Labrus_bergylta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P3YMS2-A0A6P3YMS2_SHEEP-AGRG6isoformX5-Ovis_aries\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 57(new variant)DEBUG: minus_one_residue = 79, start = 27\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P3YMS2-A0A6P3YMS2_SHEEP-AGRG6isoformX5-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [78, 79, 80]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A668RNW9-A0A668RNW9_OREAU-AGRG1-Oreochromis_aureus\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[5]])(new variant)\n",
      "DEBUG: self.indices = [257, 258, 259]No Subdomain boundaries detected. Possible Fragment found.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A668RNW9-A0A668RNW9_OREAU-AGRG1-Oreochromis_aureus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "None A0A6P4SXM5-A0A6P4SXM5_PANPR-AGRL2isoformX4-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 35, start = 77\n",
      "DEBUG: self.indices = [34, 35, 36]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4SXM5-A0A6P4SXM5_PANPR-AGRL2isoformX4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 306, start = 481None\n",
      " DEBUG: self.indices = [305, 306, 307]A0A6G1R5X2-A0A6G1R5X2_9GRUI-AGRV1-Gallirallus_okinawae.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " This is not a GAIN domain.None\n",
      " No Subdomain boundaries detected. Possible Fragment found.A0A6P4SXN3-A0A6P4SXN3_PANPR-AGRL2isoformX9-Panthera_pardus\n",
      "DEBUG: Getting alignment indices with: A0A6G1R5X2-A0A6G1R5X2_9GRUI-AGRV1-Gallirallus_okinawae., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4SXN3-A0A6P4SXN3_PANPR-AGRL2isoformX9-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = 306, start = 481DEBUG: minus_one_residue = None, start = 16\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(84, 89), (111, 116)]\n",
      "DEBUG: self.indices = [74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4SXN7-A0A6P4SXN7_PANPR-AGRL2isoformX14-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)A0A7N8WWK2-A0A7N8WWK2_9TELE-AGRF6-Mastacembelus_armatus\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4SXN7-A0A6P4SXN7_PANPR-AGRL2isoformX14-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>This is not a GAIN domain.\n",
      "\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A7N8WWK2-A0A7N8WWK2_9TELE-AGRF6-Mastacembelus_armatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]DEBUG: minus_one_residue = 99, start = 7\n",
      "\n",
      "DEBUG: self.indices = [98, 99, 100]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4SXN9-A0A6P4SXN9_PANPR-AGRL2isoformX19-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)H3BMF8_A0A0A6YYR7_A0A0K0K195_H3BTS5-H3BMF8_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[17]])(new variant)DEBUG: Getting alignment indices with: A0A6P4SXN9-A0A6P4SXN9_PANPR-AGRL2isoformX19-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BMF8_A0A0A6YYR7_A0A0K0K195_H3BTS5-H3BMF8_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      "DEBUG: minus_one_residue = None, start = 32\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " ALTERNATIVE GPS DETECTION!None  A0A6P4SXP6-A0A6P4SXP6_PANPR-AGRL2isoformX29-Panthera_pardusDEBUG:last_sheets = [(100, 105), (108, 113)]\n",
      "\n",
      "DEBUG: self.indices = [74, 75]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4SXP6-A0A6P4SXP6_PANPR-AGRL2isoformX29-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6I9XJ41-A0A6I9XJ41_9SAUR-AGRE3-like-Thamnophis_sirtalis.\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[10]])(new variant)DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6I9XJ41-A0A6I9XJ41_9SAUR-AGRE3-like-Thamnophis_sirtalis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4SXQ1-A0A6P4SXQ1_PANPR-AGRL2isoformX33-Panthera_pardusDEBUG: minus_one_residue = 100, start = 8\n",
      "\n",
      "DEBUG: self.indices = [99, 100, 101]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4SXQ1-A0A6P4SXQ1_PANPR-AGRL2isoformX33-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None H3BUH2-H3BUH2_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "DEBUG: minus_one_residue = 306, start = 481Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[18]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: Getting alignment indices with: H3BUH2-H3BUH2_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "No GPS residue present in this Domain. Is this really a GAIN?\n",
      " DEBUG: minus_one_residue = None, start = 32None \n",
      "A0A6P4T0I7-A0A6P4T0I7_PANPR-AGRL2isoformX2-Panthera_pardusALTERNATIVE GPS DETECTION! \n",
      "DEBUG:last_sheets = [(100, 105), (108, 113)]\n",
      "DEBUG: self.indices = [74, 75]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4T0I7-A0A6P4T0I7_PANPR-AGRL2isoformX2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A1W2PQU5-A0A1W2PQU5_HUMAN-AG-protein-coupledreceptorV1-Homo_sapiens\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A7F8QF18-A0A7F8QF18_LEPWE-AGRB3-Leptonychotes_weddellii\n",
      "This is not a GAIN domain.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "No Subdomain boundaries detected. Possible Fragment found. None\n",
      " A0A6P4T0J1-A0A6P4T0J1_PANPR-AGRL2isoformX7-Panthera_pardus\n",
      "DEBUG: Getting alignment indices with: A0A7F8QF18-A0A7F8QF18_LEPWE-AGRB3-Leptonychotes_weddellii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4T0J1-A0A6P4T0J1_PANPR-AGRL2isoformX7-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 95, start = 15\n",
      "DEBUG: self.indices = [94, 95, 96]\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B3QTH8-A0A3B3QTH8_9TELE-AGRG4b-Paramormyrops_kingsleyae.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " This is not a GAIN domain.\n",
      "None No Subdomain boundaries detected. Possible Fragment found.A0A6P4T0J6-A0A6P4T0J6_PANPR-AGRL2isoformX12-Panthera_pardus\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A3B3QTH8-A0A3B3QTH8_9TELE-AGRG4b-Paramormyrops_kingsleyae., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A6P4T0J6-A0A6P4T0J6_PANPR-AGRL2isoformX12-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 102, start = 9\n",
      "DEBUG: self.indices = [101, 102, 103]\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6I9XQZ7-A0A6I9XQZ7_9SAUR-AGRD1-like-Thamnophis_sirtalis.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "No Strands detected. This is not a GAIN domain. None\n",
      " A0A6P4T0J9-A0A6P4T0J9_PANPR-AGRL2isoformX17-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      " A0A6Q2ZCP4-A0A6Q2ZCP4_ESOLU-AGRG3-Esox_luciusDEBUG: Getting alignment indices with: A0A6P4T0J9-A0A6P4T0J9_PANPR-AGRL2isoformX17-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 25(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6Q2ZCP4-A0A6Q2ZCP4_ESOLU-AGRG3-Esox_lucius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: minus_one_residue = 111, start = 0\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4T0K5-A0A6P4T0K5_PANPR-AGRL2isoformX22-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4T0K5-A0A6P4T0K5_PANPR-AGRL2isoformX22-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [110, 111, 112]DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "KeyNotFound: 310HelixDEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4T3Z4-A0A6P4T3Z4_PANPR-AGRB3isoformX2-Panthera_pardus\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant) A0A3B2W7X0-A0A3B2W7X0_MOUSE-AGRE1-Mus_musculus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3B2W7X0-A0A3B2W7X0_MOUSE-AGRE1-Mus_musculus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4T3Z4-A0A6P4T3Z4_PANPR-AGRB3isoformX2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: self.indices = [77, 78, 79, 80, 81, 82, 83, 84]\n",
      "DEBUG: minus_one_residue = None, start = 25\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(100, 101), (110, 112)]\n",
      "None A0A6P4T5X5-A0A6P4T5X5_PANPR-AGRL2isoformX1-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4T5X5-A0A6P4T5X5_PANPR-AGRL2isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 306, start = 481 \n",
      "A0A3Q1MAN8-A0A3Q1MAN8_BOVIN-AGRE3-Bos_taurus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A3Q1MAN8-A0A3Q1MAN8_BOVIN-AGRE3-Bos_taurus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "DEBUG: minus_one_residue = 101, start = 11\n",
      "DEBUG: self.indices = [100, 101, 102]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4T5X8-A0A6P4T5X8_PANPR-AGRL2isoformX6-Panthera_pardus\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)A0A452GB47-A0A452GB47_CAPHI-AGRE3-Capra_hircus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A452GB47-A0A452GB47_CAPHI-AGRE3-Capra_hircus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4T5X8-A0A6P4T5X8_PANPR-AGRL2isoformX6-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 100, start = 12\n",
      "\n",
      "DEBUG: self.indices = [99, 100, 101]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 306, start = 481 A0A6G1RYL3-A0A6G1RYL3_9GRUI-AGRA2-Gallirallus_okinawae.\n",
      "\n",
      "No Strands detected. This is not a GAIN domain.\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  None A0A6P4T5Y6-A0A6P4T5Y6_PANPR-AGRL2isoformX16-Panthera_pardus\n",
      "NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      " DEBUG: Getting alignment indices with: A0A6P4T5Y6-A0A6P4T5Y6_PANPR-AGRL2isoformX16-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A6P6KHI3-A0A6P6KHI3_CARAU-AGRG6-like-Carassius_auratus\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6P6KHI3-A0A6P6KHI3_CARAU-AGRG6-like-Carassius_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "DEBUG: minus_one_residue = 106, start = 6\n",
      "DEBUG: self.indices = [105, 106, 107]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone  A0A6P4T5Z7-A0A6P4T5Z7_PANPR-AGRL2isoformX31-Panthera_pardusE7FD86-E7FD86_DANRE-AGRG3-Danio_rerio\n",
      "\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: E7FD86-E7FD86_DANRE-AGRG3-Danio_rerio, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4T5Z7-A0A6P4T5Z7_PANPR-AGRL2isoformX31-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 109, start = 3\n",
      "DEBUG: self.indices = [108, 109, 110]DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4T5Z9-A0A6P4T5Z9_PANPR-AGRL2isoformX35-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4T5Z9-A0A6P4T5Z9_PANPR-AGRL2isoformX35-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "NoneDEBUG: minus_one_residue = 293, start = 494 \n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "H3BM73-H3BM73_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 22(new variant)\n",
      "DEBUG: Getting alignment indices with: H3BM73-H3BM73_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "No GPS residue present in this Domain. Is this really a GAIN? None \n",
      "A0A6P4TEA6-A0A6P4TEA6_PANPR-AGRL2isoformX5-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TEA6-A0A6P4TEA6_PANPR-AGRL2isoformX5-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = None, start = 0DEBUG: minus_one_residue = 293, start = 494\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(105, 110), (113, 118)]\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [111, 112]\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None None A0A6P4TEA8-A0A6P4TEA8_PANPR-AGRL2isoformX10-Panthera_pardus\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)H3BSF7-H3BSF7_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A6P4TEA8-A0A6P4TEA8_PANPR-AGRL2isoformX10-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[18]])(new variant)\n",
      "\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: H3BSF7-H3BSF7_HUMAN-AG-protein-coupledreceptorG1-Homo_sapiens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "No GPS residue present in this Domain. Is this really a GAIN?DEBUG: minus_one_residue = 306, start = 481\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "DEBUG: minus_one_residue = None, start = 38\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "ALTERNATIVE GPS DETECTION! DEBUG:last_sheets = [(93, 98), (107, 110)] \n",
      "DEBUG: self.indices = [61, 62, 63, 64, 65, 66, 67, 68]None\n",
      " A0A6P4TEA9-A0A6P4TEA9_PANPR-AGRL2isoformX15-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      " NoneDEBUG: Getting alignment indices with: A0A6P4TEA9-A0A6P4TEA9_PANPR-AGRL2isoformX15-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "A0A5N4CKU2-A0A5N4CKU2_CAMDR-AGRE3-Camelus_dromedarius\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A5N4CKU2-A0A5N4CKU2_CAMDR-AGRE3-Camelus_dromedarius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 87, start = 26A0A6P4TEB0-A0A6P4TEB0_PANPR-AGRL2isoformX20-Panthera_pardus\n",
      "\n",
      "DEBUG: self.indices = [86, 87, 88]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TEB0-A0A6P4TEB0_PANPR-AGRL2isoformX20-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      " DEBUG: self.indices = [305, 306, 307]A0A5N4CKV8-A0A5N4CKV8_CAMDR-AGRE3-Camelus_dromedarius\n",
      "\n",
      "\n",
      "This is not a GAIN domain.\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A5N4CKV8-A0A5N4CKV8_CAMDR-AGRE3-Camelus_dromedarius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P4TEB4-A0A6P4TEB4_PANPR-AGRL2isoformX30-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TEB4-A0A6P4TEB4_PANPR-AGRL2isoformX30-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 87, start = 26\n",
      "\n",
      "DEBUG: self.indices = [86, 87, 88]\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 306, start = 481 \n",
      "DEBUG: self.indices = [305, 306, 307]None\n",
      " A0A670XMQ6-A0A670XMQ6_PSETE-AGRE3-like-Pseudonaja_textilis\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[9]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A670XMQ6-A0A670XMQ6_PSETE-AGRE3-like-Pseudonaja_textilis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 100, start = 13A0A6P4TEB5-A0A6P4TEB5_PANPR-AGRL2isoformX34-Panthera_pardus\n",
      "\n",
      "DEBUG: self.indices = [99, 100, 101]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "66 valid proteins were found.\n",
      "8 of which have detected Subdomains.\n",
      "36 of which have detected consensus GPS motifs.\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TEB5-A0A6P4TEB5_PANPR-AGRL2isoformX34-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4TH00-A0A6P4TH00_PANPR-AGRB3isoformX3-Panthera_pardus\n",
      "Found sequences object.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TH00-A0A6P4TH00_PANPR-AGRB3isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8P9W5-A0A6P8P9W5_GEOSA-AGRL2isoformX3-Geotrypetes_seraphini\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)DEBUG: minus_one_residue = 287, start = 500\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8P9W5-A0A6P8P9W5_GEOSA-AGRL2isoformX3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4TJG7-A0A6P4TJG7_PANPR-AGRL2isoformX3-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 305, start = 482Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJG7-A0A6P4TJG7_PANPR-AGRL2isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 306, start = 481\n",
      " A0A6P8P9X7-A0A6P8P9X7_GEOSA-AGRL2isoformX12-Geotrypetes_seraphiniDEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P8P9X7-A0A6P8P9X7_GEOSA-AGRL2isoformX12-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None A0A6P4TJG9-A0A6P4TJG9_PANPR-AGRL2isoformX8-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJG9-A0A6P4TJG9_PANPR-AGRL2isoformX8-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 306, start = 481 \n",
      "A0A6P8PS95-A0A6P8PS95_GEOSA-AGRB1isoformX5-Geotrypetes_seraphiniDEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8PS95-A0A6P8PS95_GEOSA-AGRB1isoformX5-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4TJH0-A0A6P4TJH0_PANPR-AGRL2isoformX13-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)DEBUG: minus_one_residue = 288, start = 499\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJH0-A0A6P4TJH0_PANPR-AGRL2isoformX13-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None  NoneA0A6P8PZ02-A0A6P8PZ02_GEOSA-AGRB3isoformX3-Geotrypetes_seraphini \n",
      "A0A6P4TJH1-A0A6P4TJH1_PANPR-AGRL2isoformX18-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8PZ02-A0A6P8PZ02_GEOSA-AGRB3isoformX3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJH1-A0A6P4TJH1_PANPR-AGRL2isoformX18-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " None A0A6P8Q022-A0A6P8Q022_GEOSA-AGRL1-Geotrypetes_seraphini\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant) \n",
      "A0A6P4TJH2-A0A6P4TJH2_PANPR-AGRL2isoformX23-Panthera_pardusDEBUG: Getting alignment indices with: A0A6P8Q022-A0A6P8Q022_GEOSA-AGRL1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJH2-A0A6P4TJH2_PANPR-AGRL2isoformX23-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "DEBUG: minus_one_residue = 293, start = 494\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [292, 293, 294] \n",
      "None A0A6P8Q4E9-A0A6P8Q4E9_GEOSA-AGRB1isoformX2-Geotrypetes_seraphini\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)None \n",
      "A0A6P4TJH4-A0A6P4TJH4_PANPR-AGRL2isoformX32-Panthera_pardus\n",
      "DEBUG: Getting alignment indices with: A0A6P8Q4E9-A0A6P8Q4E9_GEOSA-AGRB1isoformX2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TJH4-A0A6P4TJH4_PANPR-AGRL2isoformX32-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 288, start = 499DEBUG: self.indices = [287, 288, 289]\n",
      "DEBUG: minus_one_residue = 306, start = 481\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: self.indices = [305, 306, 307] \n",
      "None A0A6P8Q5N5-A0A6P8Q5N5_GEOSA-cadherinEGFLAGseven-passG-typereceptor3-Geotrypetes_seraphini\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 97(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8Q5N5-A0A6P8Q5N5_GEOSA-cadherinEGFLAGseven-passG-typereceptor3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P4TLL3-A0A6P4TLL3_PANPR-AGRB3isoformX1-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 337, start = 450Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4TLL3-A0A6P4TLL3_PANPR-AGRB3isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [336, 337, 338]DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4TTB2-A0A6P4TTB2_PANPR-AGRG4-Panthera_pardus\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 67(new variant) A0A6P8Q6Y3-A0A6P8Q6Y3_GEOSA-AGRB1isoformX3-Geotrypetes_seraphini\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4TTB2-A0A6P4TTB2_PANPR-AGRG4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8Q6Y3-A0A6P8Q6Y3_GEOSA-AGRB1isoformX3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 270, start = 517\n",
      "\n",
      "DEBUG: minus_one_residue = 288, start = 499DEBUG: self.indices = [269, 270, 271]\n",
      "\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " A0A6P4U2D1-A0A6P4U2D1_PANPR-AGRG6isoformX4-Panthera_pardus None A0A6P8QBQ7-A0A6P8QBQ7_GEOSA-AGRB3isoformX1-Geotrypetes_seraphini\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4U2D1-A0A6P4U2D1_PANPR-AGRG6isoformX4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A6P8QBQ7-A0A6P8QBQ7_GEOSA-AGRB3isoformX1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 260, start = 527DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8QBR2-A0A6P8QBR2_GEOSA-AGRB3isoformX5-Geotrypetes_seraphini\n",
      "\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8QBR2-A0A6P8QBR2_GEOSA-AGRB3isoformX5-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4U3F3-A0A6P4U3F3_PANPR-AGRG6isoformX3-Panthera_pardusDEBUG: minus_one_residue = 286, start = 501\n",
      "\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 58(new variant)\n",
      " NoneDEBUG: Getting alignment indices with: A0A6P4U3F3-A0A6P4U3F3_PANPR-AGRG6isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A6P8QCI0-A0A6P8QCI0_GEOSA-LOWQUALITYPROTEIN:AGRG6-Geotrypetes_seraphini\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8QCI0-A0A6P8QCI0_GEOSA-LOWQUALITYPROTEIN:AGRG6-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 257, start = 530\n",
      "DEBUG: self.indices = [256, 257, 258]DEBUG: minus_one_residue = 256, start = 531\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [255, 256, 257] \n",
      "None A0A6P4U4F7-A0A6P4U4F7_PANPR-AGRG6isoformX2-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None DEBUG: Getting alignment indices with: A0A6P4U4F7-A0A6P4U4F7_PANPR-AGRG6isoformX2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P8QEP0-A0A6P8QEP0_GEOSA-AGRF5isoformX2-Geotrypetes_seraphini\n",
      "\n",
      "DEBUG: minus_one_residue = 258, start = 529Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "Found sequences object.\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P8QEP0-A0A6P8QEP0_GEOSA-AGRF5isoformX2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P4U6A7-A0A6P4U6A7_PANPR-AGRG6isoformX5-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)DEBUG: minus_one_residue = 264, start = 523\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A060WXI2-A0A060WXI2_ONCMY-AGRL4-Oncorhynchus_mykiss\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4U6A7-A0A6P4U6A7_PANPR-AGRG6isoformX5-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [263, 264, 265]DEBUG: Getting alignment indices with: A0A060WXI2-A0A060WXI2_ONCMY-AGRL4-Oncorhynchus_mykiss, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "DEBUG: minus_one_residue = 294, start = 167\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295] \n",
      "None A0A6P8QU58-A0A6P8QU58_GEOSA-AGRB1isoformX1-Geotrypetes_seraphini\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A6P8QU58-A0A6P8QU58_GEOSA-AGRB1isoformX1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [257, 258, 259]None \n",
      "A0A2D0PL37-A0A2D0PL37_ICTPU-AGRG2-likeisoformX4-Ictalurus_punctatus\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 37(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A2D0PL37-A0A2D0PL37_ICTPU-AGRG2-likeisoformX4-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P4U826-A0A6P4U826_PANPR-AGRL1isoformX4-Panthera_pardusDEBUG: minus_one_residue = 288, start = 499\n",
      "\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "DEBUG: minus_one_residue = 461, start = 0\n",
      "DEBUG: self.indices = [460, 461, 462]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4U826-A0A6P4U826_PANPR-AGRL1isoformX4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8QX19-A0A6P8QX19_GEOSA-AGRB1isoformX4-Geotrypetes_seraphini\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 308, start = 479\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)DEBUG: self.indices = [307, 308, 309]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8QX19-A0A6P8QX19_GEOSA-AGRB1isoformX4-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4U8Z7-A0A6P4U8Z7_PANPR-AGRL1isoformX3-Panthera_pardus\n",
      " DEBUG: minus_one_residue = 288, start = 499A0A2D0Q457-A0A2D0Q457_ICTPU-AGRG2-likeisoformX6-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4U8Z7-A0A6P4U8Z7_PANPR-AGRL1isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 44(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: Getting alignment indices with: A0A2D0Q457-A0A2D0Q457_ICTPU-AGRG2-likeisoformX6-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A6P8QZX4-A0A6P8QZX4_GEOSA-AGRB3isoformX4-Geotrypetes_seraphini\n",
      "DEBUG: minus_one_residue = 308, start = 479\n",
      "DEBUG: self.indices = [307, 308, 309]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8QZX4-A0A6P8QZX4_GEOSA-AGRB3isoformX4-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 219, start = 242\n",
      " DEBUG: self.indices = [218, 219, 220]None A0A6P4UAE0-A0A6P4UAE0_PANPR-AGRG6isoformX1-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 286, start = 501\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "\n",
      "KeyNotFound: 310HelixDEBUG: Getting alignment indices with: A0A6P4UAE0-A0A6P4UAE0_PANPR-AGRG6isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " \n",
      "None A0A6P8R2W0-A0A6P8R2W0_GEOSA-AGRF5isoformX1-Geotrypetes_seraphini\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8R2W0-A0A6P8R2W0_GEOSA-AGRF5isoformX1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: minus_one_residue = 258, start = 529 \n",
      "NoneDEBUG: self.indices = [257, 258, 259] \n",
      "A0A2U9C2L2-A0A2U9C2L2_SCOMX-PutativeEGFlatrophilinandseventransmembranedomain-containingprotein1isoform3-Scophthalmus_maximus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 264, start = 523DEBUG: Getting alignment indices with: A0A2U9C2L2-A0A2U9C2L2_SCOMX-PutativeEGFlatrophilinandseventransmembranedomain-containingprotein1isoform3-Scophthalmus_maximus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> None\n",
      " \n",
      "DEBUG: self.indices = [263, 264, 265]\n",
      "A0A6P4UHT9-A0A6P4UHT9_PANPR-AGRL1isoformX1-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 105(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P4UHT9-A0A6P4UHT9_PANPR-AGRL1isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 297, start = 164\n",
      "\n",
      " DEBUG: self.indices = [296, 297, 298]None\n",
      " A0A6P8R3I4-A0A6P8R3I4_GEOSA-AGRB3isoformX2-Geotrypetes_seraphini\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 308, start = 479 \n",
      "DEBUG: self.indices = [307, 308, 309]\n",
      "NoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant) \n",
      "A0A2U9C538-A0A2U9C538_SCOMX-AGRL4-Scophthalmus_maximusDEBUG: Getting alignment indices with: A0A6P8R3I4-A0A6P8R3I4_GEOSA-AGRB3isoformX2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4UKS8-A0A6P4UKS8_PANPR-AGRB2-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 97(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)DEBUG: Getting alignment indices with: A0A6P4UKS8-A0A6P4UKS8_PANPR-AGRB2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 286, start = 501DEBUG: Getting alignment indices with: A0A2U9C538-A0A2U9C538_SCOMX-AGRL4-Scophthalmus_maximus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 323, start = 464None\n",
      "DEBUG: self.indices = [322, 323, 324] \n",
      "A0A6P8R7K5-A0A6P8R7K5_GEOSA-AGRG2-Geotrypetes_seraphini\n",
      "DEBUG: minus_one_residue = 297, start = 164\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None  A0A6P4VVZ1-A0A6P4VVZ1_PANPR-cadherinEGFLAGseven-passG-typereceptor1-Panthera_pardusNoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 55(new variant) \n",
      "A0A3B4TIE2-A0A3B4TIE2_SERDU-AGRL4-Seriola_dumerili\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8R7K5-A0A6P8R7K5_GEOSA-AGRG2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 94(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B4TIE2-A0A3B4TIE2_SERDU-AGRL4-Seriola_dumerili, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 260, start = 527\n",
      "DEBUG: self.indices = [259, 260, 261]\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)DEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P4VVZ1-A0A6P4VVZ1_PANPR-cadherinEGFLAGseven-passG-typereceptor1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " None  A0A6P8S659-A0A6P8S659_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX2-Geotrypetes_seraphiniDEBUG: minus_one_residue = 341, start = 446A0A3B4WHG1-A0A3B4WHG1_SERLL-AGRL4-Seriola_lalandi_dorsalis.\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [340, 341, 342]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B4WHG1-A0A3B4WHG1_SERLL-AGRL4-Seriola_lalandi_dorsalis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A6P8S659-A0A6P8S659_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P4WD96-A0A6P4WD96_PANPR-AGRF5isoformX4-Panthera_pardus\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 297, start = 164Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "DEBUG: Getting alignment indices with: A0A6P4WD96-A0A6P4WD96_PANPR-AGRF5isoformX4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 343, start = 444\n",
      "DEBUG: self.indices = [342, 343, 344]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q1IF79-A0A3Q1IF79_ANATE-AGRL4-Anabas_testudineusDEBUG: minus_one_residue = 266, start = 521\n",
      "\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 98(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: Getting alignment indices with: A0A3Q1IF79-A0A3Q1IF79_ANATE-AGRL4-Anabas_testudineus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "None A0A6P8S6R9-A0A6P8S6R9_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX1-Geotrypetes_seraphini\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant) A0A6P4WMJ6-A0A6P4WMJ6_PANPR-AGRF5isoformX3-Panthera_pardus\n",
      "\n",
      "DEBUG: minus_one_residue = 297, start = 164DEBUG: Getting alignment indices with: A0A6P8S6R9-A0A6P8S6R9_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4WMJ6-A0A6P4WMJ6_PANPR-AGRF5isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [296, 297, 298]\n",
      "\n",
      "DEBUG: minus_one_residue = 343, start = 444\n",
      "DEBUG: self.indices = [342, 343, 344]\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "NoneDEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      " A0A3Q2DSN5-A0A3Q2DSN5_CYPVA-AGRL4-Cyprinodon_variegatus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 95(new variant) \n",
      "DEBUG: Getting alignment indices with: A0A3Q2DSN5-A0A3Q2DSN5_CYPVA-AGRL4-Cyprinodon_variegatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P8SCB9-A0A6P8SCB9_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX3-Geotrypetes_seraphini\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)A0A6P4WP54-A0A6P4WP54_PANPR-AGRF5isoformX2-Panthera_pardus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8SCB9-A0A6P8SCB9_GEOSA-cadherinEGFLAGseven-passG-typereceptor1isoformX3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P4WP54-A0A6P4WP54_PANPR-AGRF5isoformX2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]DEBUG: minus_one_residue = 343, start = 444\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [342, 343, 344] \n",
      "None A0A3Q3AAI0-A0A3Q3AAI0_KRYMA-AGRL4-Kryptolebias_marmoratus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4X720-A0A6P4X720_PANPR-AGRL3isoformX6-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant) \n",
      "None DEBUG: Getting alignment indices with: A0A3Q3AAI0-A0A3Q3AAI0_KRYMA-AGRL4-Kryptolebias_marmoratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P8SHA8-A0A6P8SHA8_GEOSA-AGRD2isoformX3-Geotrypetes_seraphiniFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4X720-A0A6P4X720_PANPR-AGRL3isoformX6-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8SHA8-A0A6P8SHA8_GEOSA-AGRD2isoformX3-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 294, start = 493DEBUG: minus_one_residue = 294, start = 167\n",
      "DEBUG: minus_one_residue = 336, start = 451\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]DEBUG: self.indices = [293, 294, 295]\n",
      "DEBUG: self.indices = [335, 336, 337]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None None  \n",
      "A0A3Q3NJJ6-A0A3Q3NJJ6_9TELE-AGRL4-Mastacembelus_armatusA0A6P4X723-A0A6P4X723_PANPR-AGRL3isoformX7-Panthera_pardus\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant) None\n",
      " DEBUG: Getting alignment indices with: A0A6P4X723-A0A6P4X723_PANPR-AGRL3isoformX7-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A3Q3NJJ6-A0A3Q3NJJ6_9TELE-AGRL4-Mastacembelus_armatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P8SI06-A0A6P8SI06_GEOSA-AGRD2isoformX2-Geotrypetes_seraphini\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8SI06-A0A6P8SI06_GEOSA-AGRD2isoformX2-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 294, start = 167\n",
      "DEBUG: self.indices = [293, 294, 295]DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: minus_one_residue = 329, start = 458\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: self.indices = [306, 307, 308]NoneDEBUG: self.indices = [328, 329, 330]\n",
      " A0A484D1J5-A0A484D1J5_PERFV-Uncharacterizedprotein-Perca_flavescens\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: Getting alignment indices with: A0A484D1J5-A0A484D1J5_PERFV-Uncharacterizedprotein-Perca_flavescens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "  None A0A6P8SI09-A0A6P8SI09_GEOSA-AGRD2isoformX5-Geotrypetes_seraphini\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: minus_one_residue = 297, start = 164DEBUG: Getting alignment indices with: A0A6P8SI09-A0A6P8SI09_GEOSA-AGRD2isoformX5-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "A0A6P4X733-A0A6P4X733_PANPR-AGRL3isoformX9-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4X733-A0A6P4X733_PANPR-AGRL3isoformX9-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 336, start = 451\n",
      "DEBUG: self.indices = [335, 336, 337]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "NoneDEBUG: self.indices = [306, 307, 308] \n",
      "A0A4U5UKC8-A0A4U5UKC8_COLLU-AGRL4-Collichthys_lucidus\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A4U5UKC8-A0A4U5UKC8_COLLU-AGRL4-Collichthys_lucidus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "None NoneA0A6P4X767-A0A6P4X767_PANPR-AGRL3isoformX16-Panthera_pardus A0A6P8SQ23-A0A6P8SQ23_GEOSA-AGRD2isoformX1-Geotrypetes_seraphini\n",
      "\n",
      "DEBUG: minus_one_residue = 294, start = 167Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A6P8SQ23-A0A6P8SQ23_GEOSA-AGRD2isoformX1-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P4X767-A0A6P4X767_PANPR-AGRL3isoformX16-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A4W5KUJ1-A0A4W5KUJ1_9TELE-Uncharacterizedprotein-Hucho_huchoDEBUG: minus_one_residue = 336, start = 451\n",
      "\n",
      "DEBUG: self.indices = [335, 336, 337]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 94(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A4W5KUJ1-A0A4W5KUJ1_9TELE-Uncharacterizedprotein-Hucho_hucho, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8SQ26-A0A6P8SQ26_GEOSA-AGRD2isoformX4-Geotrypetes_seraphiniDEBUG: minus_one_residue = 292, start = 169DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "DEBUG: Getting alignment indices with: A0A6P8SQ26-A0A6P8SQ26_GEOSA-AGRD2isoformX4-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  None A0A4W5KUP7-A0A4W5KUP7_9TELE-Uncharacterizedprotein-Hucho_hucho\n",
      "DEBUG: minus_one_residue = 336, start = 451\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 94(new variant)NoneDEBUG: self.indices = [335, 336, 337]\n",
      "DEBUG: Getting alignment indices with: A0A4W5KUP7-A0A4W5KUP7_9TELE-Uncharacterizedprotein-Hucho_hucho, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A6P4X772-A0A6P4X772_PANPR-AGRL3isoformX17-Panthera_pardus\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 292, start = 169 None\n",
      " DEBUG: Getting alignment indices with: A0A6P4X772-A0A6P4X772_PANPR-AGRL3isoformX17-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [291, 292, 293]A0A6P8SUS7-A0A6P8SUS7_GEOSA-AGRD2isoformX6-Geotrypetes_seraphini\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 101(new variant)\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: A0A6P8SUS7-A0A6P8SUS7_GEOSA-AGRD2isoformX6-Geotrypetes_seraphini, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      " None A0A4Z2B7C0-A0A4Z2B7C0_9TELE-Uncharacterizedprotein-Takifugu_bimaculatus.\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None DEBUG: minus_one_residue = 336, start = 451A0A6P4X790-A0A6P4X790_PANPR-AGRL3isoformX19-Panthera_pardus\n",
      "DEBUG: self.indices = [335, 336, 337]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4X790-A0A6P4X790_PANPR-AGRL3isoformX19-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A4Z2B7C0-A0A4Z2B7C0_9TELE-Uncharacterizedprotein-Takifugu_bimaculatus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 307, start = 480A0A6P8SZD4-A0A6P8SZD4_GYMAC-AGRG2-like-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 294, start = 167\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 55(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None DEBUG: Getting alignment indices with: A0A6P8SZD4-A0A6P8SZD4_GYMAC-AGRG2-like-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P4X7A8-A0A6P4X7A8_PANPR-AGRL3isoformX22-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A5C6MTE7-A0A5C6MTE7_9TELE-AGRL4-Takifugu_flavidus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4X7A8-A0A6P4X7A8_PANPR-AGRL3isoformX22-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A5C6MTE7-A0A5C6MTE7_9TELE-AGRL4-Takifugu_flavidus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 268, start = 519\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [267, 268, 269]DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "DEBUG: minus_one_residue = 294, start = 167\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone \n",
      "Initializing GainDomain Object.\n",
      "  NoneA0A6P4X944-A0A6P4X944_PANPR-AGRF5isoformX1-Panthera_pardus\n",
      "A0A6P8T2Q4-A0A6P8T2Q4_GYMAC-AGRB3isoformX1-Gymnodraco_acuticeps A0A5F5XVK9-A0A5F5XVK9_FELCA-AGRA3-Felis_catus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4X944-A0A6P4X944_PANPR-AGRF5isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T2Q4-A0A6P8T2Q4_GYMAC-AGRB3isoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A5F5XVK9-A0A5F5XVK9_FELCA-AGRA3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 321, start = 140\n",
      "DEBUG: minus_one_residue = 287, start = 500 DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "A0A6P4X948-A0A6P4X948_PANPR-AGRF5isoformX5-Panthera_pardusDEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P4X948-A0A6P4X948_PANPR-AGRF5isoformX5-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None  NoneA0A665UXT5-A0A665UXT5_ECHNA-AGRL4-Echeneis_naucrates\n",
      "\n",
      " A0A6P8T2Q8-A0A6P8T2Q8_GYMAC-AGRB3isoformX5-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: minus_one_residue = 266, start = 521DEBUG: Getting alignment indices with: A0A6P8T2Q8-A0A6P8T2Q8_GYMAC-AGRB3isoformX5-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A665UXT5-A0A665UXT5_ECHNA-AGRL4-Echeneis_naucrates, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "DEBUG: minus_one_residue = 297, start = 164\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P4XIY2-A0A6P4XIY2_PANPR-AGRL3isoformX8-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " A0A6P8T2R6-A0A6P8T2R6_GYMAC-AGRB3isoformX10-Gymnodraco_acuticeps Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "None\n",
      "DEBUG: Getting alignment indices with: A0A6P4XIY2-A0A6P4XIY2_PANPR-AGRL3isoformX8-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "A0A668A7D1-A0A668A7D1_9TELE-AGRL4-Myripristis_murdjan\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T2R6-A0A6P8T2R6_GYMAC-AGRB3isoformX10-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A668A7D1-A0A668A7D1_9TELE-AGRL4-Myripristis_murdjan, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]DEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [293, 294, 295] \n",
      "None A0A6P4XJ15-A0A6P4XJ15_PANPR-AGRL3isoformX13-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)None\n",
      " DEBUG: Getting alignment indices with: A0A6P4XJ15-A0A6P4XJ15_PANPR-AGRL3isoformX13-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A668W2K4-A0A668W2K4_OREAU-AGRL4-Oreochromis_aureus\n",
      "\n",
      "A0A6P8T2R9-A0A6P8T2R9_GYMAC-AGRB3isoformX4-Gymnodraco_acuticepsFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A668W2K4-A0A668W2K4_OREAU-AGRL4-Oreochromis_aureus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: A0A6P8T2R9-A0A6P8T2R9_GYMAC-AGRB3isoformX4-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "  NoneDEBUG: self.indices = [286, 287, 288]None \n",
      "A0A668W4X2-A0A668W4X2_OREAU-AGRL4-Oreochromis_aureus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A668W4X2-A0A668W4X2_OREAU-AGRL4-Oreochromis_aureus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8T2S4-A0A6P8T2S4_GYMAC-AGRB3isoformX9-Gymnodraco_acuticeps\n",
      " A0A6P4XM51-A0A6P4XM51_PANPR-AGRL3isoformX4-Panthera_pardusDEBUG: minus_one_residue = 294, start = 167\n",
      "DEBUG: self.indices = [293, 294, 295]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4XM51-A0A6P4XM51_PANPR-AGRL3isoformX4-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P8T2S4-A0A6P8T2S4_GYMAC-AGRB3isoformX9-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " \n",
      "None A0A669BGG9-A0A669BGG9_ORENI-Uncharacterizedprotein-Oreochromis_niloticusDEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: minus_one_residue = 286, start = 501\n",
      "\n",
      "\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 94(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A669BGG9-A0A669BGG9_ORENI-Uncharacterizedprotein-Oreochromis_niloticus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8T304-A0A6P8T304_GYMAC-AGRB3isoformX7-Gymnodraco_acuticeps\n",
      " None DEBUG: minus_one_residue = 294, start = 167A0A6P4XMC4-A0A6P4XMC4_PANPR-AGRL3isoformX14-Panthera_pardus\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P4XMC4-A0A6P4XMC4_PANPR-AGRL3isoformX14-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: Getting alignment indices with: A0A6P8T304-A0A6P8T304_GYMAC-AGRB3isoformX7-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A671Z2G8-A0A671Z2G8_SPAAU-AGRL4-Sparus_aurata\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A671Z2G8-A0A671Z2G8_SPAAU-AGRL4-Sparus_aurata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None None DEBUG: minus_one_residue = 297, start = 164A0A6P8T309-A0A6P8T309_GYMAC-AGRB3isoformX12-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: self.indices = [296, 297, 298] \n",
      "A0A6P4XMD0-A0A6P4XMD0_PANPR-AGRL3isoformX15-Panthera_pardus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T309-A0A6P8T309_GYMAC-AGRB3isoformX12-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P4XMD0-A0A6P4XMD0_PANPR-AGRL3isoformX15-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A673CG72-A0A673CG72_9TELE-Uncharacterizedprotein-Sphaeramia_orbicularis\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: Getting alignment indices with: A0A673CG72-A0A673CG72_9TELE-Uncharacterizedprotein-Sphaeramia_orbicularis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4XZZ8-A0A6P4XZZ8_PANPR-AGRL3isoformX5-Panthera_pardus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 297, start = 164A0A6P8T678-A0A6P8T678_GYMAC-AGRL3isoformX3-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T678-A0A6P8T678_GYMAC-AGRL3isoformX3-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4XZZ8-A0A6P4XZZ8_PANPR-AGRL3isoformX5-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " A0A674F6Q2-A0A674F6Q2_SALTR-AGRL4-Salmo_trutta\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 94(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A674F6Q2-A0A674F6Q2_SALTR-AGRL4-Salmo_trutta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "NoneDEBUG: minus_one_residue = 294, start = 167\n",
      " DEBUG: self.indices = [293, 294, 295]\n",
      "A0A6P8T679-A0A6P8T679_GYMAC-AGRL3isoformX1-Gymnodraco_acuticeps\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A0A6P4Y047-A0A6P4Y047_PANPR-AGRL3isoformX11-Panthera_pardusDEBUG: Getting alignment indices with: A0A6P8T679-A0A6P8T679_GYMAC-AGRL3isoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant) A0A674NR19-A0A674NR19_TAKRU-Uncharacterizedprotein-Takifugu_rubripes\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y047-A0A6P4Y047_PANPR-AGRL3isoformX11-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 307, start = 480\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: Getting alignment indices with: A0A674NR19-A0A674NR19_TAKRU-Uncharacterizedprotein-Takifugu_rubripes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 294, start = 167 A0A6P8T683-A0A6P8T683_GYMAC-AGRL3isoformX8-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: self.indices = [293, 294, 295]\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P8T683-A0A6P8T683_GYMAC-AGRL3isoformX8-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone A0A6A5F2A1-A0A6A5F2A1_PERFL-Uncharacterizedprotein-Perca_fluviatilis\n",
      " Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "A0A6P4Y051-A0A6P4Y051_PANPR-AGRL3isoformX12-Panthera_pardus\n",
      "DEBUG: Getting alignment indices with: A0A6A5F2A1-A0A6A5F2A1_PERFL-Uncharacterizedprotein-Perca_fluviatilis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y051-A0A6P4Y051_PANPR-AGRL3isoformX12-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> \n",
      "NoneDEBUG: minus_one_residue = 297, start = 164 A0A6P8T684-A0A6P8T684_GYMAC-AGRL3isoformX6-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: self.indices = [296, 297, 298]\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T684-A0A6P8T684_GYMAC-AGRL3isoformX6-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "Initializing GainDomain Object.\n",
      "None\n",
      "  NoneA0A6G0IRY7-A0A6G0IRY7_LARCR-AGRL4-Larimichthys_crocea A0A6P4Y078-A0A6P4Y078_PANPR-AGRL3isoformX18-Panthera_pardus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6G0IRY7-A0A6G0IRY7_LARCR-AGRL4-Larimichthys_crocea, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 307, start = 480Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]DEBUG: Getting alignment indices with: A0A6P4Y078-A0A6P4Y078_PANPR-AGRL3isoformX18-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8T685-A0A6P8T685_GYMAC-AGRL3isoformX4-Gymnodraco_acuticepsDEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480DEBUG: self.indices = [293, 294, 295]\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T685-A0A6P8T685_GYMAC-AGRL3isoformX4-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P4Y0A4-A0A6P4Y0A4_PANPR-AGRL3isoformX21-Panthera_pardus\n",
      " A0A6P6KF14-A0A6P6KF14_CARAU-AGRE2-like-Carassius_auratus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 61(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P6KF14-A0A6P6KF14_CARAU-AGRE2-like-Carassius_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y0A4-A0A6P4Y0A4_PANPR-AGRL3isoformX21-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: minus_one_residue = 244, start = 217\n",
      "DEBUG: self.indices = [243, 244, 245]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8T688-A0A6P8T688_GYMAC-AGRL3isoformX10-Gymnodraco_acuticepsDEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [306, 307, 308]None\n",
      " A0A6P7I6U1-A0A6P7I6U1_9TELE-AGRL4-Parambassis_rangaFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8T688-A0A6P8T688_GYMAC-AGRL3isoformX10-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A6P7I6U1-A0A6P7I6U1_9TELE-AGRL4-Parambassis_ranga, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A6P4Y4W8-A0A6P4Y4W8_PANPR-AGRL3isoformX1-Panthera_pardus\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y4W8-A0A6P4Y4W8_PANPR-AGRL3isoformX1-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 297, start = 164\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8T689-A0A6P8T689_GYMAC-AGRL3isoformX9-Gymnodraco_acuticeps\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 307, start = 480NoneFound Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308] DEBUG: Getting alignment indices with: A0A6P8T689-A0A6P8T689_GYMAC-AGRL3isoformX9-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A7K5S8U5-A0A7K5S8U5_LANLU-AGRL4protein-Lanius_ludovicianus\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)None\n",
      " DEBUG: Getting alignment indices with: A0A7K5S8U5-A0A7K5S8U5_LANLU-AGRL4protein-Lanius_ludovicianus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P4Y4X7-A0A6P4Y4X7_PANPR-AGRL3isoformX2-Panthera_pardus\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 104(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y4X7-A0A6P4Y4X7_PANPR-AGRL3isoformX2-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 285, start = 176 None\n",
      " DEBUG: self.indices = [284, 285, 286]A0A6P8TJ93-A0A6P8TJ93_GYMAC-AGRB3isoformX2-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 97(new variant)\n",
      "Initializing GainDomain Object.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " NoneDEBUG: Getting alignment indices with: A0A6P8TJ93-A0A6P8TJ93_GYMAC-AGRB3isoformX2-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A7K6CB54-A0A7K6CB54_PTIVI-AGRL4protein-Ptilonorhynchus_violaceus\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant) A0A6P4Y512-A0A6P4Y512_PANPR-AGRL3isoformX3-Panthera_pardus\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7K6CB54-A0A7K6CB54_PTIVI-AGRL4protein-Ptilonorhynchus_violaceus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: Getting alignment indices with: A0A6P4Y512-A0A6P4Y512_PANPR-AGRL3isoformX3-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "DEBUG: minus_one_residue = 287, start = 174\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8TJ97-A0A6P8TJ97_GYMAC-AGRB3isoformX6-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "NoneFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant) A0A7K7LSY4-A0A7K7LSY4_9PASS-AGRL4protein-Brachypodius_atriceps\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8TJ97-A0A6P8TJ97_GYMAC-AGRB3isoformX6-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K7LSY4-A0A7K7LSY4_9PASS-AGRL4protein-Brachypodius_atriceps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "DEBUG: minus_one_residue = 285, start = 176\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8TJA2-A0A6P8TJA2_GYMAC-AGRB3isoformX11-Gymnodraco_acuticeps\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8TJA2-A0A6P8TJA2_GYMAC-AGRB3isoformX11-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A7K7Y237-A0A7K7Y237_THRLU-AGRL4protein-Thryothorus_ludovicianus\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K7Y237-A0A7K7Y237_THRLU-AGRL4protein-Thryothorus_ludovicianus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 289, start = 498\n",
      "\n",
      "DEBUG: self.indices = [288, 289, 290]\n",
      "DEBUG: minus_one_residue = 285, start = 176\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8TK71-A0A6P8TK71_GYMAC-AGRB3isoformX3-Gymnodraco_acuticeps\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8TK71-A0A6P8TK71_GYMAC-AGRB3isoformX3-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8CRC8-A0A7K8CRC8_9CORV-AGRL4protein-Eulacestoma_nigropectus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8CRC8-A0A7K8CRC8_9CORV-AGRL4protein-Eulacestoma_nigropectus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 286, start = 501\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "DEBUG: minus_one_residue = 285, start = 176\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [284, 285, 286] \n",
      "None A0A6P8TK80-A0A6P8TK80_GYMAC-AGRB3isoformX8-Gymnodraco_acuticeps\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1BK03-A0A7L1BK03_GYMTI-AGRL4protein-Gymnorhina_tibicen\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8TK80-A0A6P8TK80_GYMAC-AGRB3isoformX8-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1BK03-A0A7L1BK03_GYMTI-AGRL4protein-Gymnorhina_tibicen, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 285, start = 176\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "DEBUG: minus_one_residue = 287, start = 500\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "  NoneNone A0A6P8TML6-A0A6P8TML6_GYMAC-AGRL3isoformX5-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      " A0A7L2C7C8-A0A7L2C7C8_9PASS-AGRL4protein-Alaudala_cheleensis\n",
      "DEBUG: Getting alignment indices with: A0A6P8TML6-A0A6P8TML6_GYMAC-AGRL3isoformX5-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2C7C8-A0A7L2C7C8_9PASS-AGRL4protein-Alaudala_cheleensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 285, start = 176\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "DEBUG: minus_one_residue = 307, start = 480\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneDEBUG: self.indices = [306, 307, 308] \n",
      "A0A7L2CU31-A0A7L2CU31_CATFU-AGRL4protein-Catharus_fuscescens\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A6P8TS20-A0A6P8TS20_GYMAC-AGRL3isoformX2-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8TS20-A0A6P8TS20_GYMAC-AGRL3isoformX2-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7L2CU31-A0A7L2CU31_CATFU-AGRL4protein-Catharus_fuscescens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 306, start = 481DEBUG: minus_one_residue = 285, start = 176\n",
      "\n",
      "DEBUG: self.indices = [284, 285, 286]DEBUG: self.indices = [305, 306, 307]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2DRR3-A0A7L2DRR3_ANTMN-AGRL4protein-Anthoscopus_minutus\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " NoneFound Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2DRR3-A0A7L2DRR3_ANTMN-AGRL4protein-Anthoscopus_minutus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " A0A6P8TS26-A0A6P8TS26_GYMAC-AGRL3isoformX7-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "DEBUG: minus_one_residue = 285, start = 176DEBUG: Getting alignment indices with: A0A6P8TS26-A0A6P8TS26_GYMAC-AGRL3isoformX7-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 307, start = 480 \n",
      "A0A7L2M488-A0A7L2M488_9SYLV-AGRL4protein-Hippolais_icterinaDEBUG: self.indices = [306, 307, 308]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 90(new variant)\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "DEBUG: Getting alignment indices with: A0A7L2M488-A0A7L2M488_9SYLV-AGRL4protein-Hippolais_icterina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " DEBUG: minus_one_residue = 287, start = 174A0A6P8TS30-A0A6P8TS30_GYMAC-AGRL3isoformX11-Gymnodraco_acuticeps\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: self.indices = [286, 287, 288]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 103(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8TS30-A0A6P8TS30_GYMAC-AGRL3isoformX11-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2MP94-A0A7L2MP94_9PASS-AGRL4protein-Rhadina_sibilatrix.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)DEBUG: minus_one_residue = 307, start = 480\n",
      "\n",
      "DEBUG: self.indices = [306, 307, 308]\n",
      "DEBUG: Getting alignment indices with: A0A7L2MP94-A0A7L2MP94_9PASS-AGRL4protein-Rhadina_sibilatrix., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 285, start = 176 \n",
      "NoneDEBUG: self.indices = [284, 285, 286]\n",
      " \n",
      "Initializing GainDomain Object.\n",
      " A0A6P8U5I5-A0A6P8U5I5_GYMAC-AGRB1-like-Gymnodraco_acuticeps\n",
      "None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)A0A7L2NK71-A0A7L2NK71_PYCJO-AGRL4protein-Pycnonotus_jocosus\n",
      "DEBUG: Getting alignment indices with: A0A6P8U5I5-A0A6P8U5I5_GYMAC-AGRB1-like-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: minus_one_residue = 283, start = 504DEBUG: Getting alignment indices with: A0A7L2NK71-A0A7L2NK71_PYCJO-AGRL4protein-Pycnonotus_jocosus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 285, start = 176\n",
      "DEBUG: self.indices = [284, 285, 286]None \n",
      "A0A6P8UGJ7-A0A6P8UGJ7_GYMAC-AGRL1isoformX1-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A6P8UGJ7-A0A6P8UGJ7_GYMAC-AGRL1isoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>None\n",
      " A0A834EYR7-A0A834EYR7_ORYME-EGF,latrophilinseventransmembranedomain-containingprotein1-Oryzias_melastigma\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834EYR7-A0A834EYR7_ORYME-EGF,latrophilinseventransmembranedomain-containingprotein1-Oryzias_melastigma, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "DEBUG: minus_one_residue = 297, start = 164\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [296, 297, 298] \n",
      "None A0A6P8UMK2-A0A6P8UMK2_GYMAC-AGRF5-likeisoformX1-Gymnodraco_acuticeps\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant)None C0H9C9-C0H9C9_SALSA-EGF,latrophilinandseventransmembranedomain-containingprotein1isoformX1-Salmo_salar\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: C0H9C9-C0H9C9_SALSA-EGF,latrophilinandseventransmembranedomain-containingprotein1isoformX1-Salmo_salar, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8UMK2-A0A6P8UMK2_GYMAC-AGRF5-likeisoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 294, start = 167\n",
      "\n",
      "DEBUG: self.indices = [293, 294, 295]\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [261, 262, 263]None\n",
      " F6YLD2-F6YLD2_MACMU-AGRE2-Macaca_mulatta\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 60(new variant)\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: Getting alignment indices with: F6YLD2-F6YLD2_MACMU-AGRE2-Macaca_mulatta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P8UPZ3-A0A6P8UPZ3_GYMAC-AGRF5-likeisoformX3-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 242, start = 219\n",
      "DEBUG: self.indices = [241, 242, 243]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant) None\n",
      "DEBUG: Getting alignment indices with: A0A6P8UPZ3-A0A6P8UPZ3_GYMAC-AGRF5-likeisoformX3-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " A0A2K6FS58-A0A2K6FS58_PROCO-AGRE5-Propithecus_coquereli\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K6FS58-A0A2K6FS58_PROCO-AGRE5-Propithecus_coquereli, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "DEBUG: minus_one_residue = 255, start = 207\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None\n",
      "Initializing GainDomain Object.\n",
      " A0A6P8UWY9-A0A6P8UWY9_GYMAC-AGRB1isoformX8-Gymnodraco_acuticeps None\n",
      " A0A6P8IZ47-A0A6P8IZ47_ACTTE-AGRD1-like-Actinia_tenebrosaFound Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8UWY9-A0A6P8UWY9_GYMAC-AGRB1isoformX8-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 76(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8IZ47-A0A6P8IZ47_ACTTE-AGRD1-like-Actinia_tenebrosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 499\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "DEBUG: minus_one_residue = 264, start = 198\n",
      "DEBUG: self.indices = [263, 264, 265]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      " A0A6P8UZB3-A0A6P8UZB3_GYMAC-AGRB1isoformX1-Gymnodraco_acuticepsNone \n",
      "A0A6P9AX77-A0A6P9AX77_PANGU-AGRE5isoformX2-Pantherophis_guttatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 57(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8UZB3-A0A6P8UZB3_GYMAC-AGRB1isoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P9AX77-A0A6P9AX77_PANGU-AGRE5isoformX2-Pantherophis_guttatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 252, start = 210\n",
      "DEBUG: minus_one_residue = 283, start = 504\n",
      "DEBUG: self.indices = [251, 252, 253]DEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8UZB6-A0A6P8UZB6_GYMAC-AGRB1isoformX5-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8UZB6-A0A6P8UZB6_GYMAC-AGRB1isoformX5-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " None A0A7K5BD83-A0A7K5BD83_9FURN-AGRL4protein-Furnarius_figulus.\n",
      "DEBUG: minus_one_residue = 283, start = 504\n",
      "DEBUG: self.indices = [282, 283, 284]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 94(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7K5BD83-A0A7K5BD83_9FURN-AGRL4protein-Furnarius_figulus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 288, start = 174 \n",
      "A0A6P8UZB8-A0A6P8UZB8_GYMAC-AGRB1isoformX10-Gymnodraco_acuticepsDEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8UZB8-A0A6P8UZB8_GYMAC-AGRB1isoformX10-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 283, start = 504 \n",
      "NoneDEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " A0A7K8X385-A0A7K8X385_9PICI-AGRL4protein-Eubucco_bourcierii\n",
      "None\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      " DEBUG: Getting alignment indices with: A0A7K8X385-A0A7K8X385_9PICI-AGRL4protein-Eubucco_bourcierii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'> A0A6P8V0L1-A0A6P8V0L1_GYMAC-AGRF5-likeisoformX2-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 288, start = 174\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V0L1-A0A6P8V0L1_GYMAC-AGRF5-likeisoformX2-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 262, start = 525\n",
      "None A0A7K9U4E6-A0A7K9U4E6_9AVES-AGRL4protein-Chloroceryle_aenea\n",
      "DEBUG: self.indices = [261, 262, 263]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9U4E6-A0A7K9U4E6_9AVES-AGRL4protein-Chloroceryle_aenea, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 288, start = 174\n",
      "DEBUG: self.indices = [287, 288, 289] \n",
      "None \n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2B4L1-A0A7L2B4L1_9GRUI-AGRL4protein-Heliornis_fulicaA0A6P8V150-A0A6P8V150_GYMAC-AGRG6-like-Gymnodraco_acuticeps\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 62(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2B4L1-A0A7L2B4L1_9GRUI-AGRL4protein-Heliornis_fulica, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V150-A0A6P8V150_GYMAC-AGRG6-like-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 174\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851N9X7-A0A851N9X7_9DEND-AGRL4protein-Campylorhamphus_procurvoides.\n",
      "DEBUG: minus_one_residue = 261, start = 526\n",
      "DEBUG: self.indices = [260, 261, 262]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: Getting alignment indices with: A0A851N9X7-A0A851N9X7_9DEND-AGRL4protein-Campylorhamphus_procurvoides., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "NoneDEBUG: minus_one_residue = 288, start = 174 \n",
      "A0A6P8V518-A0A6P8V518_GYMAC-AGRB1isoformX3-Gymnodraco_acuticepsDEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V518-A0A6P8V518_GYMAC-AGRB1isoformX3-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A2D0PNQ4-A0A2D0PNQ4_ICTPU-AGRG1-likeisoformX3-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 48(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0PNQ4-A0A2D0PNQ4_ICTPU-AGRG1-likeisoformX3-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 283, start = 504\n",
      "\n",
      "DEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8V527-A0A6P8V527_GYMAC-AGRB1isoformX7-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 463, start = 0\n",
      "DEBUG: self.indices = [462, 463, 464]Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V527-A0A6P8V527_GYMAC-AGRB1isoformX7-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 285, start = 502\n",
      "Initializing GainDomain Object.\n",
      "\n",
      " DEBUG: self.indices = [284, 285, 286]\n",
      "None \n",
      "Initializing GainDomain Object.\n",
      "A0A2I2ZFB4-A0A2I2ZFB4_GORGO-AGRE2-Gorilla_gorilla_gorilla\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 60(new variant) \n",
      "None DEBUG: Getting alignment indices with: A0A2I2ZFB4-A0A2I2ZFB4_GORGO-AGRE2-Gorilla_gorilla_gorilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "A0A6P8V6B0-A0A6P8V6B0_GYMAC-AGRL2isoformX1-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V6B0-A0A6P8V6B0_GYMAC-AGRL2isoformX1-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 248, start = 215\n",
      "DEBUG: self.indices = [247, 248, 249]\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "NoneDEBUG: self.indices = [304, 305, 306] \n",
      "A0A4W6CM79-A0A4W6CM79_LATCA-AGRF3b-Lates_calcarifer\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 78(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A4W6CM79-A0A4W6CM79_LATCA-AGRF3b-Lates_calcarifer, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8V6B4-A0A6P8V6B4_GYMAC-AGRL2isoformX6-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 240, start = 223Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: self.indices = [239, 240, 241]DEBUG: Getting alignment indices with: A0A6P8V6B4-A0A6P8V6B4_GYMAC-AGRL2isoformX6-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A663E283-A0A663E283_AQUCH-AGRG2-Aquila_chrysaetos_chrysaetos.DEBUG: minus_one_residue = 305, start = 482\n",
      "\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 48(new variant) \n",
      "None DEBUG: Getting alignment indices with: A0A663E283-A0A663E283_AQUCH-AGRG2-Aquila_chrysaetos_chrysaetos., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>A0A6P8V6B8-A0A6P8V6B8_GYMAC-AGRL2isoformX9-Gymnodraco_acuticeps\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V6B8-A0A6P8V6B8_GYMAC-AGRL2isoformX9-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 242, start = 221\n",
      "DEBUG: self.indices = [241, 242, 243]\n",
      "DEBUG: minus_one_residue = 292, start = 495\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6J2VP53-A0A6J2VP53_CHACN-AGRE3-like-Chanos_chanos\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 33(new variant)A0A6P8V6C2-A0A6P8V6C2_GYMAC-AGRL2isoformX13-Gymnodraco_acuticeps\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6J2VP53-A0A6J2VP53_CHACN-AGRE3-like-Chanos_chanos, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P8V6C2-A0A6P8V6C2_GYMAC-AGRL2isoformX13-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 203, start = 260\n",
      "DEBUG: self.indices = [202, 203, 204]\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "Initializing GainDomain Object.\n",
      "A0A6P4UHQ4-A0A6P4UHQ4_PANPR-AGRE2-likeisoformX6-Panthera_pardus \n",
      "None A0A6P8V6D2-A0A6P8V6D2_GYMAC-AGRL2isoformX20-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 60(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)DEBUG: Getting alignment indices with: A0A6P4UHQ4-A0A6P4UHQ4_PANPR-AGRE2-likeisoformX6-Panthera_pardus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A6P8V6D2-A0A6P8V6D2_GYMAC-AGRL2isoformX20-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 292, start = 495\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: minus_one_residue = 202, start = 261 \n",
      "None A0A6P8V786-A0A6P8V786_GYMAC-AGRL2isoformX3-Gymnodraco_acuticeps\n",
      "DEBUG: self.indices = [201, 202, 203]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 98(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V786-A0A6P8V786_GYMAC-AGRL2isoformX3-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None DEBUG: minus_one_residue = 292, start = 495\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "A0A6P7HJ43-A0A6P7HJ43_9TELE-AGRF1-likeisoformX2-Parambassis_ranga\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P7HJ43-A0A6P7HJ43_9TELE-AGRF1-likeisoformX2-Parambassis_ranga, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      " None A0A6P8V788-A0A6P8V788_GYMAC-AGRL2isoformX8-Gymnodraco_acuticepsDEBUG: minus_one_residue = 253, start = 210\n",
      "DEBUG: self.indices = [252, 253, 254]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None \n",
      "A0A7K4UND8-A0A7K4UND8_9SYLV-AGRL4protein-Sinosuthora_webbiana.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V788-A0A6P8V788_GYMAC-AGRL2isoformX8-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A7K4UND8-A0A7K4UND8_9SYLV-AGRL4protein-Sinosuthora_webbiana., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: minus_one_residue = 305, start = 482DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None  A0A6P8V794-A0A6P8V794_GYMAC-AGRL2isoformX11-Gymnodraco_acuticepsNone\n",
      " A0A7K5CNJ8-A0A7K5CNJ8_MOTAL-AGRL4protein-Motacilla_alba\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5CNJ8-A0A7K5CNJ8_MOTAL-AGRL4protein-Motacilla_alba, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)DEBUG: minus_one_residue = 287, start = 176\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V794-A0A6P8V794_GYMAC-AGRL2isoformX11-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneDEBUG: minus_one_residue = 305, start = 482 A0A7K5EAK5-A0A7K5EAK5_POLCE-AGRL4protein-Polioptila_caerulea\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5EAK5-A0A7K5EAK5_POLCE-AGRL4protein-Polioptila_caerulea, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P8V7A2-A0A6P8V7A2_GYMAC-AGRL2isoformX17-Gymnodraco_acuticeps\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 106(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V7A2-A0A6P8V7A2_GYMAC-AGRL2isoformX17-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5JDV9-A0A7K5JDV9_TOXRE-AGRL4protein-Toxostoma_redivivum\n",
      "DEBUG: minus_one_residue = 305, start = 482Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7K5JDV9-A0A7K5JDV9_TOXRE-AGRL4protein-Toxostoma_redivivum, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " DEBUG: minus_one_residue = 287, start = 176None\n",
      " DEBUG: self.indices = [286, 287, 288]A0A6P8V9T5-A0A6P8V9T5_GYMAC-AGRB1isoformX4-Gymnodraco_acuticeps\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5MX91-A0A7K5MX91_CARCD-AGRL4protein-Cardinalis_cardinalis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5MX91-A0A7K5MX91_CARCD-AGRL4protein-Cardinalis_cardinalis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5P561-A0A7K5P561_9CORV-AGRL4protein-Erythrocercus_mccallii.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5P561-A0A7K5P561_9CORV-AGRL4protein-Erythrocercus_mccallii., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5UGC6-A0A7K5UGC6_9PASS-AGRL4protein-Cephalopterus_ornatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5UGC6-A0A7K5UGC6_9PASS-AGRL4protein-Cephalopterus_ornatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 286, start = 177\n",
      "DEBUG: self.indices = [285, 286, 287]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5VPL1-A0A7K5VPL1_9CORV-AGRL4protein-Platysteira_castanea.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5VPL1-A0A7K5VPL1_9CORV-AGRL4protein-Platysteira_castanea., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: Getting alignment indices with: A0A6P8V9T5-A0A6P8V9T5_GYMAC-AGRB1isoformX4-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "DEBUG: minus_one_residue = 283, start = 504\n",
      "DEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5WDL9-A0A7K5WDL9_9SYLV-AGRL4protein-Hylia_prasina\n",
      "Initializing GainDomain Object.\n",
      " \n",
      "None A0A6P8V9U0-A0A6P8V9U0_GYMAC-AGRB1isoformX9-Gymnodraco_acuticeps\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5WDL9-A0A7K5WDL9_9SYLV-AGRL4protein-Hylia_prasina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A6P8V9U0-A0A6P8V9U0_GYMAC-AGRB1isoformX9-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "DEBUG: minus_one_residue = 283, start = 504\n",
      "DEBUG: self.indices = [282, 283, 284]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K6FN74-A0A7K6FN74_9CORV-AGRL4protein-Daphoenositta_chrysoptera\n",
      "\n",
      "Initializing GainDomain Object.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant) None A0A6P8VEY3-A0A6P8VEY3_GYMAC-AGRL1isoformX2-Gymnodraco_acuticeps\n",
      "DEBUG: Getting alignment indices with: A0A7K6FN74-A0A7K6FN74_9CORV-AGRL4protein-Daphoenositta_chrysoptera, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 105(new variant)DEBUG: minus_one_residue = 287, start = 176\n",
      "\n",
      "DEBUG: self.indices = [286, 287, 288]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A6P8VEY3-A0A6P8VEY3_GYMAC-AGRL1isoformX2-Gymnodraco_acuticeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K6I9W7-A0A7K6I9W7_9PASS-AGRL4protein-Dasyornis_broadbenti\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 88(new variant)Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A7K6I9W7-A0A7K6I9W7_9PASS-AGRL4protein-Dasyornis_broadbenti, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K6LH61-A0A7K6LH61_9CORV-AGRL4protein-Falcunculus_frontatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K6LH61-A0A7K6LH61_9CORV-AGRL4protein-Falcunculus_frontatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K7EBW0-A0A7K7EBW0_9SYLV-AGRL4protein-Sylvia_atricapilla\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K7EBW0-A0A7K7EBW0_9SYLV-AGRL4protein-Sylvia_atricapilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8A1L9-A0A7K8A1L9_9PASE-AGRL4protein-Melanocharis_versteri\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8A1L9-A0A7K8A1L9_9PASE-AGRL4protein-Melanocharis_versteri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "Found sequences object.\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834FM00-A0A834FM00_ORYME-AGRG4-Oryzias_melastigma\n",
      "Found Helix boundary with the following characteristics: maxk = 10 helical_counts[maxk] = 69(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834FM00-A0A834FM00_ORYME-AGRG4-Oryzias_melastigma, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 281, start = 506\n",
      "DEBUG: self.indices = [280, 281, 282]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QA86-A0A834QA86_MARMO-AGRG4-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 64(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A834QA86-A0A834QA86_MARMO-AGRG4-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 270, start = 517\n",
      "DEBUG: self.indices = [269, 270, 271]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QB65-A0A834QB65_MARMO-AGRG4-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 64(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A834QB65-A0A834QB65_MARMO-AGRG4-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 270, start = 517\n",
      "DEBUG: self.indices = [269, 270, 271]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QG82-A0A834QG82_MARMO-AGRF5-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834QG82-A0A834QG82_MARMO-AGRF5-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QH99-A0A834QH99_MARMO-AGRF5-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834QH99-A0A834QH99_MARMO-AGRF5-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 266, start = 521\n",
      "DEBUG: self.indices = [265, 266, 267]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QNG7-A0A834QNG7_MARMO-GR126-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834QNG7-A0A834QNG7_MARMO-GR126-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QPE0-A0A834QPE0_MARMO-GR126-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834QPE0-A0A834QPE0_MARMO-GR126-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A834QQX1-A0A834QQX1_MARMO-GR126-Marmota_monax\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A834QQX1-A0A834QQX1_MARMO-GR126-Marmota_monax, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 258, start = 529\n",
      "DEBUG: self.indices = [257, 258, 259]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A835P1F5-A0A835P1F5_9PASS-Uncharacterizedprotein-Lamprotornis_superbus.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A835P1F5-A0A835P1F5_9PASS-Uncharacterizedprotein-Lamprotornis_superbus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A835P2Z6-A0A835P2Z6_9PASS-Uncharacterizedprotein-Lamprotornis_superbus.\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A835P2Z6-A0A835P2Z6_9PASS-Uncharacterizedprotein-Lamprotornis_superbus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A835ZST6-A0A835ZST6_SHEEP-Uncharacterizedprotein-Ovis_aries\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A835ZST6-A0A835ZST6_SHEEP-Uncharacterizedprotein-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A836AKN9-A0A836AKN9_SHEEP-Uncharacterizedprotein-Ovis_aries\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 94(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A836AKN9-A0A836AKN9_SHEEP-Uncharacterizedprotein-Ovis_aries, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 341, start = 446\n",
      "DEBUG: self.indices = [340, 341, 342]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850TDF8-A0A850TDF8_GRUAM-AGRF5protein-Grus_americana\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850TDF8-A0A850TDF8_GRUAM-AGRF5protein-Grus_americana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850U8Q1-A0A850U8Q1_GRUAM-CELR1protein-Grus_americana\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850U8Q1-A0A850U8Q1_GRUAM-CELR1protein-Grus_americana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 341, start = 446\n",
      "DEBUG: self.indices = [340, 341, 342]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850UFA2-A0A850UFA2_9CORV-AGRF5protein-Chloropsis_hardwickii.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850UFA2-A0A850UFA2_9CORV-AGRF5protein-Chloropsis_hardwickii., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850UH95-A0A850UH95_9CORV-AGRG6protein-Chloropsis_hardwickii.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850UH95-A0A850UH95_9CORV-AGRG6protein-Chloropsis_hardwickii., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850W6V8-A0A850W6V8_FREMA-AGRF5protein-Fregata_magnificens\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 85(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850W6V8-A0A850W6V8_FREMA-AGRF5protein-Fregata_magnificens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 261, start = 526\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850WPA7-A0A850WPA7_PIACA-AGRG6protein-Piaya_cayana\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850WPA7-A0A850WPA7_PIACA-AGRG6protein-Piaya_cayana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850X012-A0A850X012_PIACA-CELR1protein-Piaya_cayana\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850X012-A0A850X012_PIACA-CELR1protein-Piaya_cayana, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850Y2F1-A0A850Y2F1_AEGCA-AGRG6protein-Aegithalos_caudatus\n",
      "Found Helix boundary with the following characteristics: maxk = 8 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850Y2F1-A0A850Y2F1_AEGCA-AGRG6protein-Aegithalos_caudatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850YBY7-A0A850YBY7_AEGCA-CELR3protein-Aegithalos_caudatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850YBY7-A0A850YBY7_AEGCA-CELR3protein-Aegithalos_caudatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 337, start = 450\n",
      "DEBUG: self.indices = [336, 337, 338]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850YQ24-A0A850YQ24_AEGCA-AGRF5protein-Aegithalos_caudatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850YQ24-A0A850YQ24_AEGCA-AGRF5protein-Aegithalos_caudatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850ZD09-A0A850ZD09_9PASS-AGRF5protein-Tichodroma_muraria.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850ZD09-A0A850ZD09_9PASS-AGRF5protein-Tichodroma_muraria., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850ZX85-A0A850ZX85_SULDA-AGRF5protein-Sula_dactylatra\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A850ZX85-A0A850ZX85_SULDA-AGRF5protein-Sula_dactylatra, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851A2G4-A0A851A2G4_SULDA-CELR3protein-Sula_dactylatra\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 81(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851A2G4-A0A851A2G4_SULDA-CELR3protein-Sula_dactylatra, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851AXP6-A0A851AXP6_PICGY-AGRF5protein-Picathartes_gymnocephalus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851AXP6-A0A851AXP6_PICGY-AGRF5protein-Picathartes_gymnocephalus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851CIT8-A0A851CIT8_CALVR-AGRF5protein-Calyptomena_viridis\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 88(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851CIT8-A0A851CIT8_CALVR-AGRF5protein-Calyptomena_viridis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851DZ98-A0A851DZ98_9CORV-AGRF5protein-Dryoscopus_gambensis.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A851DZ98-A0A851DZ98_9CORV-AGRF5protein-Dryoscopus_gambensis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851EZG1-A0A851EZG1_PITSO-AGRG6protein-Pitta_sordida\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851EZG1-A0A851EZG1_PITSO-AGRG6protein-Pitta_sordida, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851F0J4-A0A851F0J4_PITSO-AGRF5protein-Pitta_sordida\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851F0J4-A0A851F0J4_PITSO-AGRF5protein-Pitta_sordida, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851IRR0-A0A851IRR0_9PASS-AGRF5protein-Donacobius_atricapilla.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851IRR0-A0A851IRR0_9PASS-AGRF5protein-Donacobius_atricapilla., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 265, start = 522\n",
      "DEBUG: self.indices = [264, 265, 266]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851J4Z5-A0A851J4Z5_9PASS-CELR3protein-Donacobius_atricapilla.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851J4Z5-A0A851J4Z5_9PASS-CELR3protein-Donacobius_atricapilla., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851JI70-A0A851JI70_9PASS-AGRG6protein-Donacobius_atricapilla.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851JI70-A0A851JI70_9PASS-AGRG6protein-Donacobius_atricapilla., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851KEW8-A0A851KEW8_VIDCH-AGRG6protein-Vidua_chalybeata\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851KEW8-A0A851KEW8_VIDCH-AGRG6protein-Vidua_chalybeata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851KZ85-A0A851KZ85_VIDCH-AGRF5protein-Vidua_chalybeata\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851KZ85-A0A851KZ85_VIDCH-AGRF5protein-Vidua_chalybeata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851LDZ4-A0A851LDZ4_CORCR-AGRF5protein-Corythaeola_cristata\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851LDZ4-A0A851LDZ4_CORCR-AGRF5protein-Corythaeola_cristata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851LGI8-A0A851LGI8_CORCR-AGRG6protein-Corythaeola_cristata\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851LGI8-A0A851LGI8_CORCR-AGRG6protein-Corythaeola_cristata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 248, start = 539\n",
      "DEBUG: self.indices = [247, 248, 249]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851M7A4-A0A851M7A4_9DEND-AGRG6protein-Campylorhamphus_procurvoides.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851M7A4-A0A851M7A4_9DEND-AGRG6protein-Campylorhamphus_procurvoides., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851MBP8-A0A851MBP8_9DEND-AGRF5protein-Campylorhamphus_procurvoides.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851MBP8-A0A851MBP8_9DEND-AGRF5protein-Campylorhamphus_procurvoides., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851NIP4-A0A851NIP4_9GALL-CELR1protein-Penelope_pileata.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851NIP4-A0A851NIP4_9GALL-CELR1protein-Penelope_pileata., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851NTV1-A0A851NTV1_9GALL-CELR3protein-Penelope_pileata.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851NTV1-A0A851NTV1_9GALL-CELR3protein-Penelope_pileata., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851P8M2-A0A851P8M2_9GALL-AGRG6protein-Penelope_pileata.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 59(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851P8M2-A0A851P8M2_9GALL-AGRG6protein-Penelope_pileata., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851PG00-A0A851PG00_9GALL-AGRF5protein-Penelope_pileata.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 83(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851PG00-A0A851PG00_9GALL-AGRF5protein-Penelope_pileata., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 259, start = 528\n",
      "DEBUG: self.indices = [258, 259, 260]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851PHL8-A0A851PHL8_ANHAN-AGRF5protein-Anhinga_anhinga\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A851PHL8-A0A851PHL8_ANHAN-AGRF5protein-Anhinga_anhinga, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851QPJ0-A0A851QPJ0_TYCCO-CELR1protein-Tychaedon_coryphoeus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851QPJ0-A0A851QPJ0_TYCCO-CELR1protein-Tychaedon_coryphoeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851R3M5-A0A851R3M5_TYCCO-AGRG6protein-Tychaedon_coryphoeus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851R3M5-A0A851R3M5_TYCCO-AGRG6protein-Tychaedon_coryphoeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851RII4-A0A851RII4_TYCCO-AGRF5protein-Tychaedon_coryphoeus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851RII4-A0A851RII4_TYCCO-AGRF5protein-Tychaedon_coryphoeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851RT08-A0A851RT08_CERFA-CELR3protein-Certhia_familiaris\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851RT08-A0A851RT08_CERFA-CELR3protein-Certhia_familiaris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 337, start = 450\n",
      "DEBUG: self.indices = [336, 337, 338]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851RYI8-A0A851RYI8_CERFA-AGRF5protein-Certhia_familiaris\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 92(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851RYI8-A0A851RYI8_CERFA-AGRF5protein-Certhia_familiaris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 525\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851RZP7-A0A851RZP7_CERFA-AGRG6protein-Certhia_familiaris\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851RZP7-A0A851RZP7_CERFA-AGRG6protein-Certhia_familiaris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 254, start = 533\n",
      "DEBUG: self.indices = [253, 254, 255]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851T473-A0A851T473_9AVES-CELR1protein-Nothocercus_nigrocapillus.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851T473-A0A851T473_9AVES-CELR1protein-Nothocercus_nigrocapillus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 341, start = 446DEBUG: self.indices = [340, 341, 342]\n",
      "\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851T7Z6-A0A851T7Z6_9AVES-AGRF5protein-Nothocercus_nigrocapillus.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 90(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851T7Z6-A0A851T7Z6_9AVES-AGRF5protein-Nothocercus_nigrocapillus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 261, start = 526\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851TEW6-A0A851TEW6_9AVES-AGRG6protein-Nothocercus_nigrocapillus.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851TEW6-A0A851TEW6_9AVES-AGRG6protein-Nothocercus_nigrocapillus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851TYU6-A0A851TYU6_9PASS-CELR3protein-Elachura_formosa\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851TYU6-A0A851TYU6_9PASS-CELR3protein-Elachura_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851U5V8-A0A851U5V8_9PASS-AGRG6protein-Elachura_formosa\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851U5V8-A0A851U5V8_9PASS-AGRG6protein-Elachura_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851UHJ6-A0A851UHJ6_9PASS-AGRF5protein-Elachura_formosa\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "\n",
      "DEBUG: Getting alignment indices with: A0A851UHJ6-A0A851UHJ6_9PASS-AGRF5protein-Elachura_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851V3R9-A0A851V3R9_9PASS-AGRG6protein-Copsychus_sechellarum.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 58(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851V3R9-A0A851V3R9_9PASS-AGRG6protein-Copsychus_sechellarum., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851VB96-A0A851VB96_9PASS-AGRF5protein-Copsychus_sechellarum.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851VB96-A0A851VB96_9PASS-AGRF5protein-Copsychus_sechellarum., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851VHI9-A0A851VHI9_9PASS-CELR3protein-Copsychus_sechellarum.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851VHI9-A0A851VHI9_9PASS-CELR3protein-Copsychus_sechellarum., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851WCE8-A0A851WCE8_9CORV-AGRL2protein-Corvus_moneduloides\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 106(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851WCE8-A0A851WCE8_9CORV-AGRL2protein-Corvus_moneduloides, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 482\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " None A0A851WXA4-A0A851WXA4_9CORV-AGRF5protein-Corvus_moneduloides\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 92(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851WXA4-A0A851WXA4_9CORV-AGRF5protein-Corvus_moneduloides, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851XH41-A0A851XH41_EOLRO-CELR3protein-Eolophus_roseicapilla\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851XH41-A0A851XH41_EOLRO-CELR3protein-Eolophus_roseicapilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 337, start = 450\n",
      "DEBUG: self.indices = [336, 337, 338]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851Y3G4-A0A851Y3G4_EOLRO-AGRF5protein-Eolophus_roseicapilla\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851Y3G4-A0A851Y3G4_EOLRO-AGRF5protein-Eolophus_roseicapilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851YXY7-A0A851YXY7_9AVES-AGRG6protein-Halcyon_senegalensis.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851YXY7-A0A851YXY7_9AVES-AGRG6protein-Halcyon_senegalensis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 256, start = 531\n",
      "DEBUG: self.indices = [255, 256, 257]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851ZL69-A0A851ZL69_CALOR-AGRG6protein-Calcarius_ornatus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851ZL69-A0A851ZL69_CALOR-AGRG6protein-Calcarius_ornatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851ZPA9-A0A851ZPA9_9AVES-AGRF5protein-Halcyon_senegalensis.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A851ZPA9-A0A851ZPA9_9AVES-AGRF5protein-Halcyon_senegalensis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851ZY22-A0A851ZY22_CALOR-CELR1protein-Calcarius_ornatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851ZY22-A0A851ZY22_CALOR-CELR1protein-Calcarius_ornatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852A874-A0A852A874_CALOR-AGRF5protein-Calcarius_ornatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852A874-A0A852A874_CALOR-AGRF5protein-Calcarius_ornatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 264, start = 523\n",
      "DEBUG: self.indices = [263, 264, 265]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852AKH6-A0A852AKH6_9CORV-AGRF5protein-Chloropsis_cyanopogon.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852AKH6-A0A852AKH6_9CORV-AGRF5protein-Chloropsis_cyanopogon., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852BDR8-A0A852BDR8_9CORV-CELR3protein-Chloropsis_cyanopogon.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A852BDR8-A0A852BDR8_9CORV-CELR3protein-Chloropsis_cyanopogon., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 340, start = 447\n",
      "DEBUG: self.indices = [339, 340, 341]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852C7L5-A0A852C7L5_9PICI-AGRF5protein-Ramphastos_sulfuratus.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852C7L5-A0A852C7L5_9PICI-AGRF5protein-Ramphastos_sulfuratus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852D7W9-A0A852D7W9_PASAF-AGRF5protein-Passerina_amoena\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852D7W9-A0A852D7W9_PASAF-AGRF5protein-Passerina_amoena, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852EM53-A0A852EM53_VIDMA-AGRF5protein-Vidua_macroura\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 93(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852EM53-A0A852EM53_VIDMA-AGRF5protein-Vidua_macroura, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852F9H8-A0A852F9H8_PEUTA-AGRG6protein-Peucedramus_taeniatus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A852F9H8-A0A852F9H8_PEUTA-AGRG6protein-Peucedramus_taeniatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 255, start = 532\n",
      "DEBUG: self.indices = [254, 255, 256]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852FWP4-A0A852FWP4_PEUTA-AGRF5protein-Peucedramus_taeniatus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 91(new variant)\n",
      "WARNING: There are still secondary-structure associated residues within the Subdomain connecting loop. Please retry with different parameters.\n",
      "DEBUG: Getting alignment indices with: A0A852FWP4-A0A852FWP4_PEUTA-AGRF5protein-Peucedramus_taeniatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 524\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "75 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "Found sequences object.\n",
      "\n",
      "Initializing GainDomain Object.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " None A0A7K8AJM3-A0A7K8AJM3_9CORV-AGRL4protein-Cnemophilus_loriae\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8AJM3-A0A7K8AJM3_9CORV-AGRL4protein-Cnemophilus_loriae, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8BGP2-A0A7K8BGP2_9CORV-AGRL4protein-Rhagologus_leucostigma.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8BGP2-A0A7K8BGP2_9CORV-AGRL4protein-Rhagologus_leucostigma., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8EZQ0-A0A7K8EZQ0_9CORV-AGRL4protein-Struthidea_cinerea.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8EZQ0-A0A7K8EZQ0_9CORV-AGRL4protein-Struthidea_cinerea., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 175\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8G519-A0A7K8G519_ORTSP-AGRL4protein-Orthonyx_spaldingii\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8G519-A0A7K8G519_ORTSP-AGRL4protein-Orthonyx_spaldingii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " NoneA0A7K8IGR5-A0A7K8IGR5_9PASS-AGRL4protein-Chaetorhynchus_papuensis \n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8IGR5-A0A7K8IGR5_9PASS-AGRL4protein-Chaetorhynchus_papuensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8RCS1-A0A7K8RCS1_9PASS-AGRL4protein-Rhodinocichla_rosea.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8RCS1-A0A7K8RCS1_9PASS-AGRL4protein-Rhodinocichla_rosea., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9AL66-A0A7K9AL66_DRONO-AGRL4protein-Dromaius_novaehollandiae\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9AL66-A0A7K9AL66_DRONO-AGRL4protein-Dromaius_novaehollandiae, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 175\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9GQL9-A0A7K9GQL9_LOXLE-AGRL4protein-Loxia_leucoptera\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9GQL9-A0A7K9GQL9_LOXLE-AGRL4protein-Loxia_leucoptera, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9IPZ4-A0A7K9IPZ4_9CORV-AGRL4protein-Myiagra_hebetior.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9IPZ4-A0A7K9IPZ4_9CORV-AGRL4protein-Myiagra_hebetior., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9LR57-A0A7K9LR57_9PASS-AGRL4protein-Rhabdornis_inornatus.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9LR57-A0A7K9LR57_9PASS-AGRL4protein-Rhabdornis_inornatus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9QJU3-A0A7K9QJU3_IRECY-AGRL4protein-Irena_cyanogastra\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9QJU3-A0A7K9QJU3_IRECY-AGRL4protein-Irena_cyanogastra, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9W2Z0-A0A7K9W2Z0_9PASS-AGRL4protein-Rhipidura_dahli.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9W2Z0-A0A7K9W2Z0_9PASS-AGRL4protein-Rhipidura_dahli., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L0A8V7-A0A7L0A8V7_9CORV-AGRL4protein-Dicrurus_megarhynchus.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L0A8V7-A0A7L0A8V7_9CORV-AGRL4protein-Dicrurus_megarhynchus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L0QRX7-A0A7L0QRX7_SETKR-AGRL4protein-Setophaga_kirtlandii\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L0QRX7-A0A7L0QRX7_SETKR-AGRL4protein-Setophaga_kirtlandii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L0YVP7-A0A7L0YVP7_9PASE-AGRL4protein-Ploceus_nigricollis.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L0YVP7-A0A7L0YVP7_9PASE-AGRL4protein-Ploceus_nigricollis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1CB44-A0A7L1CB44_9PASS-AGRL4protein-Illadopsis_cleaveri\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1CB44-A0A7L1CB44_9PASS-AGRL4protein-Illadopsis_cleaveri, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1F9N8-A0A7L1F9N8_SYLBO-AGRL4protein-Sylvia_borin\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1F9N8-A0A7L1F9N8_SYLBO-AGRL4protein-Sylvia_borin, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1GPX1-A0A7L1GPX1_9PICI-AGRL4protein-Indicator_maculatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1GPX1-A0A7L1GPX1_9PICI-AGRL4protein-Indicator_maculatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 175\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1K6E5-A0A7L1K6E5_RYNNI-AGRL4protein-Rynchops_niger\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A7L1K6E5-A0A7L1K6E5_RYNNI-AGRL4protein-Rynchops_niger, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 175\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1Q462-A0A7L1Q462_ORIOR-AGRL4protein-Oriolus_oriolus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1Q462-A0A7L1Q462_ORIOR-AGRL4protein-Oriolus_oriolus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1QR70-A0A7L1QR70_9PASS-AGRL4protein-Cisticola_juncidis.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 86(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1QR70-A0A7L1QR70_9PASS-AGRL4protein-Cisticola_juncidis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 285, start = 178\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1RNH5-A0A7L1RNH5_9PASS-AGRL4protein-Locustella_ochotensis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1RNH5-A0A7L1RNH5_9PASS-AGRL4protein-Locustella_ochotensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1VSH5-A0A7L1VSH5_SITEU-AGRL4protein-Sitta_europaea\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1VSH5-A0A7L1VSH5_SITEU-AGRL4protein-Sitta_europaea, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1WDK9-A0A7L1WDK9_9PASS-AGRL4protein-Certhia_brachydactyla\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1WDK9-A0A7L1WDK9_9PASS-AGRL4protein-Certhia_brachydactyla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2FMZ2-A0A7L2FMZ2_QUIME-AGRL4protein-Quiscalus_mexicanus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2FMZ2-A0A7L2FMZ2_QUIME-AGRL4protein-Quiscalus_mexicanus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2TP43-A0A7L2TP43_POMRU-AGRL4protein-Pomatostomus_ruficeps\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2TP43-A0A7L2TP43_POMRU-AGRL4protein-Pomatostomus_ruficeps, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L3KTA8-A0A7L3KTA8_9PASS-AGRL4protein-Drymodes_brunneopygia.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L3KTA8-A0A7L3KTA8_9PASS-AGRL4protein-Drymodes_brunneopygia., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A803K0F5-A0A803K0F5_XENTR-AGRE1-Xenopus_tropicalis\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 69(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A803K0F5-A0A803K0F5_XENTR-AGRE1-Xenopus_tropicalis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 242, start = 221\n",
      "DEBUG: self.indices = [241, 242, 243]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850UH47-A0A850UH47_9CORV-AGRL4protein-Chloropsis_hardwickii.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850UH47-A0A850UH47_9CORV-AGRL4protein-Chloropsis_hardwickii., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A850Y764-A0A850Y764_AEGCA-AGRL4protein-Aegithalos_caudatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A850Y764-A0A850Y764_AEGCA-AGRL4protein-Aegithalos_caudatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851DY38-A0A851DY38_9CORV-AGRL4protein-Dryoscopus_gambensis.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851DY38-A0A851DY38_9CORV-AGRL4protein-Dryoscopus_gambensis., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851JGS8-A0A851JGS8_9PASS-AGRL4protein-Donacobius_atricapilla.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 89(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851JGS8-A0A851JGS8_9PASS-AGRL4protein-Donacobius_atricapilla., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851TT27-A0A851TT27_9PASS-AGRL4protein-Elachura_formosa\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 93(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851TT27-A0A851TT27_9PASS-AGRL4protein-Elachura_formosa, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 292, start = 171\n",
      "DEBUG: self.indices = [291, 292, 293]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851XWV7-A0A851XWV7_EOLRO-AGRL4protein-Eolophus_roseicapilla\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 78(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851XWV7-A0A851XWV7_EOLRO-AGRL4protein-Eolophus_roseicapilla, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 275, start = 188\n",
      "DEBUG: self.indices = [274, 275, 276]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A852IKM5-A0A852IKM5_9PASS-AGRL4protein-Nicator_chloris.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 84(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A852IKM5-A0A852IKM5_9PASS-AGRL4protein-Nicator_chloris., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 176\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0Q495-A0A2D0Q495_ICTPU-AGRE2-like-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 57(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0Q495-A0A2D0Q495_ICTPU-AGRE2-like-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 244, start = 220\n",
      "DEBUG: self.indices = [243, 244, 245]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I3T9E7_A0A2J8LUG3-A0A2I3T9E7_PANTR-AGRE2-Pan_troglodytes\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3T9E7_A0A2J8LUG3-A0A2I3T9E7_PANTR-AGRE2-Pan_troglodytes, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: minus_one_residue = 247, start = 217\n",
      "DEBUG: self.indices = [246, 247, 248]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q3SZ76-A0A3Q3SZ76_9TELE-AGRL4-Mastacembelus_armatus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3Q3SZ76-A0A3Q3SZ76_9TELE-AGRL4-Mastacembelus_armatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 297, start = 167\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q3WH64-A0A3Q3WH64_MOLML-AGRL4-Mola_mola\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3Q3WH64-A0A3Q3WH64_MOLML-AGRL4-Mola_mola, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 297, start = 167\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5I098-A0A7K5I098_CROSL-AGRL4protein-Crotophaga_sulcirostris\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5I098-A0A7K5I098_CROSL-AGRL4protein-Crotophaga_sulcirostris, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K6YWZ1-A0A7K6YWZ1_ALCTO-AGRL4protein-Alca_torda\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K6YWZ1-A0A7K6YWZ1_ALCTO-AGRL4protein-Alca_torda, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9HG95-A0A7K9HG95_9AVES-AGRL4protein-Bucco_capensis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9HG95-A0A7K9HG95_9AVES-AGRL4protein-Bucco_capensis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9SU85-A0A7K9SU85_9AVES-AGRL4protein-Galbula_dea.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9SU85-A0A7K9SU85_9AVES-AGRL4protein-Galbula_dea., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 293, start = 171\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K9VI91-A0A7K9VI91_ANSSE-AGRL4protein-Anseranas_semipalmata\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K9VI91-A0A7K9VI91_ANSSE-AGRL4protein-Anseranas_semipalmata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 293, start = 171\n",
      "DEBUG: self.indices = [292, 293, 294]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L0AWR4-A0A7L0AWR4_9AVES-AGRL4protein-Ciconia_maguari.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L0AWR4-A0A7L0AWR4_9AVES-AGRL4protein-Ciconia_maguari., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L1T6E1-A0A7L1T6E1_ARAGA-AGRL4protein-Aramus_guarauna\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 92(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L1T6E1-A0A7L1T6E1_ARAGA-AGRL4protein-Aramus_guarauna, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 291, start = 173\n",
      "DEBUG: self.indices = [290, 291, 292]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L2HKC4-A0A7L2HKC4_SAGSE-AGRL4protein-Sagittarius_serpentarius\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L2HKC4-A0A7L2HKC4_SAGSE-AGRL4protein-Sagittarius_serpentarius, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L3FT52-A0A7L3FT52_9GRUI-AGRL4protein-Zapornia_atra.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L3FT52-A0A7L3FT52_9GRUI-AGRL4protein-Zapornia_atra., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L4BYG6-A0A7L4BYG6_9AVES-AGRL4protein-Nyctiprogne_leucopyga.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 90(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L4BYG6-A0A7L4BYG6_9AVES-AGRL4protein-Nyctiprogne_leucopyga., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7N6A197-A0A7N6A197_ANATE-AGRL4-Anabas_testudineus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 101(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7N6A197-A0A7N6A197_ANATE-AGRL4-Anabas_testudineus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 300, start = 164\n",
      "DEBUG: self.indices = [299, 300, 301]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851NQM1-A0A851NQM1_9GALL-AGRL4protein-Penelope_pileata.\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 91(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851NQM1-A0A851NQM1_9GALL-AGRL4protein-Penelope_pileata., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 176\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None M4ATZ8-M4ATZ8_XIPMA-Uncharacterizedprotein-Xiphophorus_maculatus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 95(new variant)\n",
      "DEBUG: Getting alignment indices with: M4ATZ8-M4ATZ8_XIPMA-Uncharacterizedprotein-Xiphophorus_maculatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 296, start = 168\n",
      "DEBUG: self.indices = [295, 296, 297]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None W5MQ29-W5MQ29_LEPOC-AGRE5b,duplicate3-Lepisosteus_oculatus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 53(new variant)\n",
      "DEBUG: Getting alignment indices with: W5MQ29-W5MQ29_LEPOC-AGRE5b,duplicate3-Lepisosteus_oculatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 233, start = 231\n",
      "DEBUG: self.indices = [232, 233, 234]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2I3LWG3-A0A2I3LWG3_PAPAN-AGRE1-Papio_anubis\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 68(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2I3LWG3-A0A2I3LWG3_PAPAN-AGRE1-Papio_anubis, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 261, start = 204\n",
      "DEBUG: self.indices = [260, 261, 262]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2K6B5D3-A0A2K6B5D3_MACNE-AGRE5-Macaca_nemestrina\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2K6B5D3-A0A2K6B5D3_MACNE-AGRE5-Macaca_nemestrina, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 263, start = 202\n",
      "DEBUG: self.indices = [262, 263, 264]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B3QZ56-A0A3B3QZ56_9TELE-AGRG2a-Paramormyrops_kingsleyae.\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 53(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: A0A3B3QZ56-A0A3B3QZ56_9TELE-AGRG2a-Paramormyrops_kingsleyae., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 259, start = 206\n",
      "DEBUG: self.indices = [258, 259, 260]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3Q2L5M0-A0A3Q2L5M0_HORSE-AGRD1-Equus_caballus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 76(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3Q2L5M0-A0A3Q2L5M0_HORSE-AGRD1-Equus_caballus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 308, start = 157\n",
      "DEBUG: self.indices = [307, 308, 309]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A5F5XWQ3-A0A5F5XWQ3_FELCA-AGRA3-Felis_catus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 82(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A5F5XWQ3-A0A5F5XWQ3_FELCA-AGRA3-Felis_catus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 321, start = 144\n",
      "DEBUG: self.indices = [320, 321, 322]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A671Z2H0-A0A671Z2H0_SPAAU-AGRL4-Sparus_aurata\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 97(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A671Z2H0-A0A671Z2H0_SPAAU-AGRL4-Sparus_aurata, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 297, start = 168\n",
      "DEBUG: self.indices = [296, 297, 298]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A673ZNV3-A0A673ZNV3_SALTR-AGRD1-Salmo_trutta\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 34(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A673ZNV3-A0A673ZNV3_SALTR-AGRD1-Salmo_trutta, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 276, start = 189\n",
      "DEBUG: self.indices = [275, 276, 277]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P5C2P3-A0A6P5C2P3_BOSIN-AGRE3-likeisoformX3-Bos_indicus\n",
      "Found Helix boundary with the following characteristics: maxk = 6 helical_counts[maxk] = 37(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P5C2P3-A0A6P5C2P3_BOSIN-AGRE3-likeisoformX3-Bos_indicus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 213, start = 252\n",
      "DEBUG: self.indices = [212, 213, 214]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P6PCS1-A0A6P6PCS1_CARAU-AGRG2-like-Carassius_auratus\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[ 8,  2,  0,  0, 12,  5,  1,  2, 19,  2,  1,  4, 10]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6P6PCS1-A0A6P6PCS1_CARAU-AGRG2-like-Carassius_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 390, start = 75\n",
      "DEBUG: self.indices = [389, 390, 391]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P7M5T4-A0A6P7M5T4_BETSP-AGRD2-likeisoformX10-Betta_splendens\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A6P7M5T4-A0A6P7M5T4_BETSP-AGRD2-likeisoformX10-Betta_splendens, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 305, start = 160\n",
      "DEBUG: self.indices = [304, 305, 306]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K5D686-A0A7K5D686_9TYRA-AGRL4protein-Pachyramphus_minor.\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 98(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K5D686-A0A7K5D686_9TYRA-AGRL4protein-Pachyramphus_minor., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 311, start = 154\n",
      "DEBUG: self.indices = [310, 311, 312]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7K8E6K8-A0A7K8E6K8_LEURO-AGRL4protein-Leucopsar_rothschildi\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 85(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7K8E6K8-A0A7K8E6K8_LEURO-AGRL4protein-Leucopsar_rothschildi, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 285, start = 180\n",
      "DEBUG: self.indices = [284, 285, 286]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L0I694-A0A7L0I694_AREIN-AGRL4protein-Arenaria_interpres\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 87(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L0I694-A0A7L0I694_AREIN-AGRL4protein-Arenaria_interpres, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 288, start = 177\n",
      "DEBUG: self.indices = [287, 288, 289]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A3B4AB58-A0A3B4AB58_9GOBI-Uncharacterizedprotein-Periophthalmus_magnuspinnatus.\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 100(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A3B4AB58-A0A3B4AB58_9GOBI-Uncharacterizedprotein-Periophthalmus_magnuspinnatus., /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 302, start = 164\n",
      "DEBUG: self.indices = [301, 302, 303]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A6P6NP28-A0A6P6NP28_CARAU-AGRG2-like-Carassius_auratus\n",
      "Found Helix boundary with the following characteristics: maxk = None helical_counts[maxk] = array([[11,  5,  0,  2, 19,  2,  1,  4, 10]])(new variant)\n",
      "No Subdomain boundaries detected. Possible Fragment found.\n",
      "DEBUG: Getting alignment indices with: A0A6P6NP28-A0A6P6NP28_CARAU-AGRG2-like-Carassius_auratus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 390, start = 76\n",
      "DEBUG: self.indices = [389, 390, 391]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7J8BNF9-A0A7J8BNF9_ROUAE-AGRE1-Rousettus_aegyptiacus\n",
      "Found Helix boundary with the following characteristics: maxk = 4 helical_counts[maxk] = 67(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7J8BNF9-A0A7J8BNF9_ROUAE-AGRE1-Rousettus_aegyptiacus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 262, start = 204\n",
      "DEBUG: self.indices = [261, 262, 263]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A7L4F3M7-A0A7L4F3M7_HIRRU-AGRL4protein-Hirundo_rustica\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 88(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A7L4F3M7-A0A7L4F3M7_HIRRU-AGRL4protein-Hirundo_rustica, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 287, start = 179\n",
      "DEBUG: self.indices = [286, 287, 288]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A851QVI8-A0A851QVI8_TYCCO-AGRL4protein-Tychaedon_coryphoeus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 96(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A851QVI8-A0A851QVI8_TYCCO-AGRL4protein-Tychaedon_coryphoeus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 311, start = 155\n",
      "DEBUG: self.indices = [310, 311, 312]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None E9QMJ5-E9QMJ5_MOUSE-AGRE5-Mus_musculus\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 53(new variant)\n",
      "DEBUG: Getting alignment indices with: E9QMJ5-E9QMJ5_MOUSE-AGRE5-Mus_musculus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 242, start = 224\n",
      "DEBUG: self.indices = [241, 242, 243]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None G1M4E9-G1M4E9_AILME-AGRE2-Ailuropoda_melanoleuca\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 60(new variant)\n",
      "DEBUG: Getting alignment indices with: G1M4E9-G1M4E9_AILME-AGRE2-Ailuropoda_melanoleuca, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 248, start = 218\n",
      "DEBUG: self.indices = [247, 248, 249]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None G3VTH8-G3VTH8_SARHA-AGRE5-Sarcophilus_harrisii\n",
      "Found Helix boundary with the following characteristics: maxk = 2 helical_counts[maxk] = 54(new variant)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG: Getting alignment indices with: G3VTH8-G3VTH8_SARHA-AGRE5-Sarcophilus_harrisii, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 247, start = 219\n",
      "DEBUG: self.indices = [246, 247, 248]\n",
      "\n",
      "Initializing GainDomain Object.\n",
      " None A0A2D0Q6N2-A0A2D0Q6N2_ICTPU-AGRG7isoformX2-Ictalurus_punctatus\n",
      "Found Helix boundary with the following characteristics: maxk = 0 helical_counts[maxk] = 73(new variant)\n",
      "DEBUG: Getting alignment indices with: A0A2D0Q6N2-A0A2D0Q6N2_ICTPU-AGRG7isoformX2-Ictalurus_punctatus, /home/flo/projects/agpcr_nom/big_mafft.fa <class 'dict'>\n",
      "DEBUG: minus_one_residue = 290, start = 177\n",
      "DEBUG: self.indices = [289, 290, 291]\n",
      "Completed Collection initialitazion of 75 sequences.\n",
      "75 valid proteins were found.\n",
      "73 of which have detected Subdomains.\n",
      "75 of which have detected consensus GPS motifs.\n",
      "\n",
      "Completed mutithreaded filtering.\n"
     ]
    }
   ],
   "source": [
    "#batch_filter_seqs(test_arg[0])\n",
    "run_mp_collection(arg_list, n_threads=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
